<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Asciidoctor 2.0.15.dev">
<title>Frequently Asked Questions</title>
<link rel="stylesheet" href="./index.css">
</head>
<body class="article">
<div id="header">
</div>
<div id="content">
<div class="sidebarblock navig">
<div class="content">
<div class="ulist">
<div class="title">General</div>
<ul>
<li>
<p><a href="index.html">Main page</a></p>
</li>
<li>
<p><a href="../bcftools.html">Manual page</a></p>
</li>
<li>
<p><a href="install.html">Installation</a></p>
</li>
<li>
<p><a href="publications.html">Publications</a></p>
</li>
</ul>
</div>
<div class="ulist">
<div class="title">Calling</div>
<ul>
<li>
<p><a href="cnv-calling.html">CNV calling</a></p>
</li>
<li>
<p><a href="csq-calling.html">Consequence calling</a></p>
</li>
<li>
<p><a href="consensus-sequence.html">Consensus calling</a></p>
</li>
<li>
<p><a href="roh-calling.html">ROH calling</a></p>
</li>
<li>
<p><a href="variant-calling.html">Variant calling and filtering</a></p>
</li>
</ul>
</div>
<div class="ulist">
<div class="title">Tips and Tricks</div>
<ul>
<li>
<p><a href="convert.html">Converting formats</a></p>
</li>
<li>
<p><a href="annotate.html">Adding annotation</a></p>
</li>
<li>
<p><a href="query.html">Extracting information</a></p>
</li>
<li>
<p><a href="filtering.html">Filtering expressions</a></p>
</li>
<li>
<p><a href="scaling.html">Performance and Scaling</a></p>
</li>
<li>
<p><a href="plugins.html">Plugins</a></p>
</li>
<li>
<p><a href="FAQ.html">FAQ</a></p>
</li>
</ul>
</div>
</div>
</div>
<div id="main">
<div class="sect1">
<h2 id="_frequently_asked_questions">Frequently Asked Questions</h2>
<div class="sectionbody">
<div id="incorrect-nfields" class="paragraph">
<div class="title"><strong>Incorrect number of fields at chr1:1234567</strong></div>
<p>This error is triggered when the number of values in the data line does not match
its definition in the header. For example, one may see an error like</p>
</div>
<div class="listingblock">
<div class="content">
<pre>[W::bcf_calc_ac] Incorrect number of AC fields at Chrxx:xxxxx. (This message is printed only once.)</pre>
</div>
</div>
<div class="paragraph">
<p>In this example, the VCF specification defines the tag AC as <code>Number=A</code></p>
</div>
<div class="listingblock">
<div class="content">
<pre>##INFO=&lt;ID=AC,Number=A,Type=Integer,Description="Allele count in genotypes"&gt;</pre>
</div>
</div>
<div class="paragraph">
<p>and expects a value for each ALT allele, for example</p>
</div>
<div class="listingblock">
<div class="content">
<pre>chr1  64334  .  A  C,T  .  .  AC=1,1  GT  0/1  0/1</pre>
</div>
</div>
<div class="paragraph">
<p>The error above is printed when different number of values is encountered, for example <code>AC=1</code> or <code>AC=1,1,1</code> in the example above.</p>
</div>
<div class="paragraph">
<p>Other such definitions are <code>Number=R</code> (there must be as many values as there are REF+ALT alleles in total),
and <code>Number=G</code> (this is more complicated, see the section 1.4.2 of the <a href="http://samtools.github.io/hts-specs/VCFv4.3.pdf">VCF specification</a>).</p>
</div>
<div class="paragraph">
<p><em>How to verify:</em><br>
Look up the tag definition in the header (<code>bcftools view -h file.vcf.gz | grep TAG</code>) to check the expected number
of values and then check the number of alleles and values in the data line (<code>bcftools view -H file.vcf.gz -r chr1:1234567</code>).
Note that the program only works with ploidy 1 or 2, so if defined as <code>Number=G</code> and the ploidy is bigger,
the program is not ready for cases like that.</p>
</div>
<div class="paragraph">
<p><em>How to fix:</em><br>
If the tag is not important for your analysis, a quick and dirty workaround is to remove the
tag from the VCF completely</p>
</div>
<div class="listingblock">
<div class="content">
<pre>bcftools annotate -x TAG</pre>
</div>
</div>
<div class="paragraph">
<p>If the tag must remain in the VCF, change the definition of the tag in the header to <code>Number=.</code></p>
</div>
<div class="listingblock">
<div class="content">
<pre>bcftools view -h old.vcf &gt; hdr.txt
# edit hdr.txt and change the tag definition to Number=.
bcftools reheader -h hdr.txt old.vcf &gt; new.vcf</pre>
</div>
</div>
<div id="region-subset" class="paragraph">
<div class="title"><strong>The -R option pulls in sites from outside of the regions file</strong></div>
<p>As described in the <a href="../bcftools.html#common_options">manual page</a>, the <code>-R</code> option takes into account overlapping records.
If a strict subset by position is required, add (or replace with) the <code>-T</code> option.</p>
</div>
<div id="query-vs-view-filtering" class="paragraph">
<div class="title"><strong>Filtering expressions seem to work differently for <code>view</code> and <code>query</code>. What is going on?</strong></div>
<p>Say you want to print a list of samples with non-reference genotypes. This can be done using the following command</p>
</div>
<div class="listingblock">
<div class="content">
<pre>$ bcftools query -f'[%CHROM:%POS %SAMPLE %GT\n]' -i 'GT="alt"' file.vcf
1:67893 sample3 0/1</pre>
</div>
</div>
<div class="paragraph">
<p>However, you may also want to print genotypes of ALL samples at variant sites with at least one
non-reference genotype. In order for this to work, first select the desired rows
with the <code>view</code> command, then let <code>query</code> format the output</p>
</div>
<div class="listingblock">
<div class="content">
<pre>$ bcftools view -i 'GT="alt"' file.vcf -Ou | bcftools query -f'[%CHROM:%POS %SAMPLE %GT\n]'
1:67893 sample1 0/0
1:67893 sample2 0/0
1:67893 sample3 0/1</pre>
</div>
</div>
<div class="paragraph">
<div class="title"><strong>How to cite BCFtools?</strong></div>
<p>Please see <a href="publications.html">here</a>.</p>
</div>
<div class="sect2">
<h3 id="_feedback">Feedback</h3>
<div class="paragraph">
<p>We welcome your feedback, please help us improve this page by 
either opening an <a href="https://github.com/samtools/bcftools/issues">issue on github</a> or <a href="https://github.com/samtools/bcftools/tree/gh-pages/howtos">editing it directly</a> and sending
a pull request.</p>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="footer">
<div id="footer-text">
</div>
</div>
</body>
</html>