<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Asciidoctor 2.0.15.dev">
<title>Plugin fixref</title>
<link rel="stylesheet" href="./index.css">
</head>
<body class="article">
<div id="header">
</div>
<div id="content">
<div class="sidebarblock navig">
<div class="content">
<div class="ulist">
<div class="title">General</div>
<ul>
<li>
<p><a href="index.html">Main page</a></p>
</li>
<li>
<p><a href="../bcftools.html">Manual page</a></p>
</li>
<li>
<p><a href="install.html">Installation</a></p>
</li>
<li>
<p><a href="publications.html">Publications</a></p>
</li>
</ul>
</div>
<div class="ulist">
<div class="title">Calling</div>
<ul>
<li>
<p><a href="cnv-calling.html">CNV calling</a></p>
</li>
<li>
<p><a href="csq-calling.html">Consequence calling</a></p>
</li>
<li>
<p><a href="consensus-sequence.html">Consensus calling</a></p>
</li>
<li>
<p><a href="roh-calling.html">ROH calling</a></p>
</li>
<li>
<p><a href="variant-calling.html">Variant calling and filtering</a></p>
</li>
</ul>
</div>
<div class="ulist">
<div class="title">Tips and Tricks</div>
<ul>
<li>
<p><a href="convert.html">Converting formats</a></p>
</li>
<li>
<p><a href="annotate.html">Adding annotation</a></p>
</li>
<li>
<p><a href="query.html">Extracting information</a></p>
</li>
<li>
<p><a href="filtering.html">Filtering expressions</a></p>
</li>
<li>
<p><a href="scaling.html">Performance and Scaling</a></p>
</li>
<li>
<p><a href="plugins.html">Plugins</a></p>
</li>
<li>
<p><a href="FAQ.html">FAQ</a></p>
</li>
</ul>
</div>
</div>
</div>
<div id="main">
<div class="sect1">
<h2 id="_plugin_fixref">Plugin fixref</h2>
<div class="sectionbody">
<div class="admonitionblock warning">
<table>
<tr>
<td class="icon">
<div class="title">Warning</div>
</td>
<td class="content">
<strong>Do not use the program blindly, make an effort to understand what
strand convention your data uses! Make sure the reason for mismatching REF
alleles is not a different reference build!! Also do NOT use <code>bcftools norm --check-ref s</code> for this purpose,
as it will result in nonsense genotypes!!!</strong>
</td>
</tr>
</table>
</div>
<div class="paragraph">
<p>This tool helps to determine and fix strand orientation.
Currently it can collect and print numbers useful in determining
the strand convention (the <code>stats</code> mode), swap REF/ALT alleles based on the
SNP reference ID (the <code>id</code> mode),
flip or swap non-ambiguous SNPs (the <code>flip</code> mode),
or convert from the
<a href="http://www.illumina.com/documents/products/technotes/technote_topbot.pdf">Illumina TOP strand</a>
convention to the forward strand (the <code>top</code> mode).</p>
</div>
<div class="paragraph">
<p>Run the stats to learn the number of REF allele mismatches and the number of
non-biallelic sites:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>bcftools +fixref test.bcf -- -f ref.fa</pre>
</div>
</div>
<div class="paragraph">
<p>Another tool for checking the reference allele mismatches:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>bcftools norm --check-ref e -f /path/to/reference.fasta input.vcf.gz -Ou -o /dev/null</pre>
</div>
</div>
<div class="paragraph">
<p>If there are no REF mismatches and the number of multi-allelic sites is small, we are done.
If the output shows that the VCF is TOP-compatible, the following command can be
used to fix the strand:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>bcftools +fixref test.bcf -Ob -o output.bcf -- -f ref.fa -m top</pre>
</div>
</div>
<div class="paragraph">
<p>If the file contains dbSNP reference identificators (rsXXX in the ID column), the following
commands can be used to swap the reference and alternate alleles:</p>
</div>
<div class="listingblock">
<div class="content">
<pre># Get the dbSNP annotation file. Make sure the correct reference build is used (e.g. b37)
#   https://www.ncbi.nlm.nih.gov/variation/docs/human_variation_vcf/
wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b146_GRCh37p13/VCF/All_20151104.vcf.gz
wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b146_GRCh37p13/VCF/All_20151104.vcf.gz.tbi

# Swap the alleles
bcftools +fixref broken.bcf -Ob -o fixref.bcf -- -d -f /path/to/reference.fasta -i All_20151104.vcf.gz

# The above command might have changed the coordinates, we must sort the VCF.
bcftools sort fixref.bcf -Ob -o fixref.sorted.bcf</pre>
</div>
</div>
<div class="paragraph">
<p>In the most extreme case when nothing else is working, one can simply force
the unambiguous alleles onto the forward strand and drop the ambiguous genotypes.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>bcftools +fixref test.bcf -Ob -o output.bcf -- -f ref.fa -m flip -d</pre>
</div>
</div>
<div class="paragraph">
<p><strong>Note that this is an extremely unsafe operation and will most likely result in nonsense
genotypes.</strong> If you decide to use it anyway, make sure to <strong>check the sanity of the
result</strong> with the <a href="plugin.af-dist.html">af-dist plugin</a>!!</p>
</div>
<div class="admonitionblock warning">
<table>
<tr>
<td class="icon">
<div class="title">Warning</div>
</td>
<td class="content">
<strong>Do not use the program blindly, make an effort to understand what
strand convention your data uses! Make sure the reason for mismatching REF
alleles is not a different reference build!! Also do NOT use <code>bcftools norm --check-ref s</code> for this purpose,
as it will result in nonsense genotypes!!!</strong>
</td>
</tr>
</table>
</div>
<div class="sect2">
<h3 id="_feedback">Feedback</h3>
<div class="paragraph">
<p>We welcome your feedback, please help us improve this page by 
either opening an <a href="https://github.com/samtools/bcftools/issues">issue on github</a> or <a href="https://github.com/samtools/bcftools/tree/gh-pages/howtos">editing it directly</a> and sending
a pull request.</p>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="footer">
<div id="footer-text">
</div>
</div>
</body>
</html>