<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<!--[if IE]><meta http-equiv="X-UA-Compatible" content="IE=edge"><![endif]-->
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Asciidoctor 1.5.2">
<title>Plugin af-dist</title>
<link rel="stylesheet" href="./index.css">
</head>
<body class="article">
<div id="header">
</div>
<div id="content">
<div class="sidebarblock navig">
<div class="content">
<div class="ulist">
<div class="title">General</div>
<ul>
<li>
<p><a href="index.html">Main page</a></p>
</li>
<li>
<p><a href="../bcftools.html">Manual page</a></p>
</li>
<li>
<p><a href="install.html">Installation</a></p>
</li>
</ul>
</div>
<div class="ulist">
<div class="title">Calling</div>
<ul>
<li>
<p><a href="cnv-calling.html">CNV calling</a></p>
</li>
<li>
<p><a href="consensus-sequence.html">Consensus calling</a></p>
</li>
<li>
<p><a href="roh-calling.html">ROH calling</a></p>
</li>
<li>
<p><a href="variant-calling.html">Variant calling</a></p>
</li>
</ul>
</div>
<div class="ulist">
<div class="title">Tips and Tricks</div>
<ul>
<li>
<p><a href="query.html">Extracting information</a></p>
</li>
<li>
<p><a href="filtering.html">Filtering</a></p>
</li>
<li>
<p><a href="plugins.html">Plugins</a></p>
</li>
</ul>
</div>
</div>
</div>
<div id="main">
<div class="sect1">
<h2 id="_plugin_af_dist">Plugin af-dist</h2>
<div class="sectionbody">
<div class="paragraph">
<p>This plugin allows to detect possible strand issues by checking
genotype frequencies against population allele frequencies.</p>
</div>
<div class="paragraph">
<p>If working with human data, first download the 1000 Genomes allele frequency annotations</p>
</div>
<div class="listingblock">
<div class="content">
<pre>wget -O af.vcf.gz ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.wgs.phase3_shapeit2_mvncall_integrated_v5b.20130502.sites.vcf.gz

bcftools index af.vcf.gz</pre>
</div>
</div>
<div class="paragraph">
<p>Then annotate your data file and stream the result through the <code>af-dist</code> plugin to create
the genotype frequency distribution</p>
</div>
<div class="listingblock">
<div class="content">
<pre>bcftools annotate -c INFO/AF -a af.vcf.gz data.vcf.gz | bcftools +af-dist | grep ^PROB &gt; data.dist.txt</pre>
</div>
</div>
<div class="paragraph">
<p>The output should something like this</p>
</div>
<div class="listingblock">
<div class="content">
<pre>PROB_DIST   0.000000    0.100000    100618
PROB_DIST   0.100000    0.200000    144103
PROB_DIST   0.200000    0.300000    214923
PROB_DIST   0.300000    0.400000    320721
PROB_DIST   0.400000    0.500000    817965
PROB_DIST   0.500000    0.600000    84027
PROB_DIST   0.600000    0.700000    86531
PROB_DIST   0.700000    0.800000    97986
PROB_DIST   0.800000    0.900000    108776
PROB_DIST   0.900000    1.000000    176755</pre>
</div>
</div>
<div class="paragraph">
<p>Finally plot the distribution to check whether there are only few unlikely genotypes.</p>
</div>
<div class="exampleblock">
<div class="content">
<a href="plugins/af-dist.png"><img src="plugins/af-dist.png" width="500px"></a>
<div class="paragraph">
<p>Example of two <code>af-dist</code> outputs combined into one plot. In this figure, one
of the VCFs (blue line) has many low-probability genotypes (given the 1000Genomes allele
frequencies and assuming HWE), which suggests that many of the REF and ALT
alleles are on the incorrect strand.</p>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_feedback">Feedback</h3>
<div class="paragraph">
<p>We welcome your feedback, please help us improve this page by
either opening an issue on github or editing it directly and sending
a pull request.</p>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="footer">
<div id="footer-text">
</div>
</div>
</body>
</html>