<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Asciidoctor 2.0.15.dev">
<title>Plugins</title>
<link rel="stylesheet" href="./index.css">
</head>
<body class="article">
<div id="header">
</div>
<div id="content">
<div class="sidebarblock navig">
<div class="content">
<div class="ulist">
<div class="title">General</div>
<ul>
<li>
<p><a href="index.html">Main page</a></p>
</li>
<li>
<p><a href="../bcftools.html">Manual page</a></p>
</li>
<li>
<p><a href="install.html">Installation</a></p>
</li>
<li>
<p><a href="publications.html">Publications</a></p>
</li>
</ul>
</div>
<div class="ulist">
<div class="title">Calling</div>
<ul>
<li>
<p><a href="cnv-calling.html">CNV calling</a></p>
</li>
<li>
<p><a href="csq-calling.html">Consequence calling</a></p>
</li>
<li>
<p><a href="consensus-sequence.html">Consensus calling</a></p>
</li>
<li>
<p><a href="roh-calling.html">ROH calling</a></p>
</li>
<li>
<p><a href="variant-calling.html">Variant calling and filtering</a></p>
</li>
</ul>
</div>
<div class="ulist">
<div class="title">Tips and Tricks</div>
<ul>
<li>
<p><a href="convert.html">Converting formats</a></p>
</li>
<li>
<p><a href="annotate.html">Adding annotation</a></p>
</li>
<li>
<p><a href="query.html">Extracting information</a></p>
</li>
<li>
<p><a href="filtering.html">Filtering expressions</a></p>
</li>
<li>
<p><a href="scaling.html">Performance and Scaling</a></p>
</li>
<li>
<p><a href="plugins.html">Plugins</a></p>
</li>
<li>
<p><a href="FAQ.html">FAQ</a></p>
</li>
</ul>
</div>
</div>
</div>
<div id="main">
<div class="sect1">
<h2 id="_plugins">Plugins</h2>
<div class="sectionbody">
<div class="admonitionblock important">
<table>
<tr>
<td class="icon">
<div class="title">Important</div>
</td>
<td class="content">
In order to use the BCFtools plugins, this environment variable must
be set and point to the correct location:<br>
<code>export BCFTOOLS_PLUGINS=/path/to/bcftools/plugins</code><br>
If a plugin still does not work, <code>bcftools +name -vv</code> explains why it would not load.<br>
Note that plugins can be also run by giving the full path, for example <code>bcftools +/path/to/plugin-name.so</code>.
</td>
</tr>
</table>
</div>
<div class="sect2">
<h3 id="_for_developers">For developers</h3>
<div class="paragraph">
<p>Please read <a href="plugin.api.html">here</a>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_list_of_plugins">List of plugins</h3>
<div class="dlist">
<dl>
<dt class="hdlist1">ad-bias</dt>
<dd>
<p>Finds positions with ALT allele frequency significantly different
between two samples. Performs a Fisher test on the FORMAT/AD annotation.</p>
</dd>
<dt class="hdlist1">add-variantkey</dt>
<dd>
<p>add VariantKey INFO fields VKX and RSX</p>
</dd>
<dt class="hdlist1"><a href="plugin.af-dist.html">af-dist</a></dt>
<dd>
<p>Collects AF deviation stats and GT probability distribution given AF and assuming HWE</p>
</dd>
<dt class="hdlist1">allele-length</dt>
<dd>
<p>count the frequency of the length of REF, ALT and REF+ALT</p>
</dd>
<dt class="hdlist1">check-ploidy</dt>
<dd>
<p>Checks if ploidy of samples is consistent for all sites.</p>
</dd>
<dt class="hdlist1">check-sparsity</dt>
<dd>
<p>Prints samples without genotypes in a region (-r/-R) or chromosome (the default)</p>
</dd>
<dt class="hdlist1">color-chrs</dt>
<dd>
<p>Colors shared chromosomal segments shared. Requires phased GTs.</p>
</dd>
<dt class="hdlist1">contrast</dt>
<dd>
<p>Runs a basic association test, per-site or in a region, and checks for novel alleles and genotypes in two groups of samples. Adds the following INFO annotations:</p>
<div class="ulist">
<ul>
<li>
<p>PASSOC  .. Fisher&#8217;s exact test probability of genotypic association (REF vs non-REF allele)</p>
</li>
<li>
<p>FASSOC  .. proportion of non-REF allele in controls and cases</p>
</li>
<li>
<p>NASSOC  .. number of control-ref, control-alt, case-ref and case-alt alleles</p>
</li>
<li>
<p>NOVELAL .. lists samples with a novel allele not observed in the control group</p>
</li>
<li>
<p>NOVELGT .. lists samples with a novel genotype not observed in the control group</p>
</li>
</ul>
</div>
</dd>
<dt class="hdlist1">counts</dt>
<dd>
<p>A minimal plugin which counts number of samples, SNPs, INDELs, MNPs and total number of sites.</p>
</dd>
<dt class="hdlist1">dosage</dt>
<dd>
<p>Prints genotype dosage determined from FORMAT/PL, GL, or GT.</p>
</dd>
<dt class="hdlist1">fill-from-fasta</dt>
<dd>
<p>Fills INFO or REF field based on values in a fasta file.</p>
</dd>
<dt class="hdlist1"><a href="plugin.fill-tags.html">fill-tags</a></dt>
<dd>
<p>Sets INFO tags such as AF, AC, AC_Hemi, etc, for all samples or per population.</p>
</dd>
<dt class="hdlist1">fixploidy</dt>
<dd>
<p>Fix ploidy</p>
</dd>
<dt class="hdlist1"><a href="plugin.fixref.html">fixref</a></dt>
<dd>
<p>Diagnose and fix VCFs with genotypes on incorrect strand.</p>
</dd>
<dt class="hdlist1">frameshifts</dt>
<dd>
<p>Annotates frameshift indels.</p>
</dd>
<dt class="hdlist1">GTisec</dt>
<dd>
<p>Counts genotype intersections across all possible sample subsets in a vcf file.</p>
</dd>
<dt class="hdlist1">GTsubset</dt>
<dd>
<p>Outputs sites where the requested samples all exclusively share a genotype (GT), i.e.
all selected samples must have the same GT, while non of the others can have it.</p>
</dd>
<dt class="hdlist1">guess-ploidy</dt>
<dd>
<p>Determines sample sex by checking genotype likelihoods (GL,PL) or genotypes (GT) in the non-PAR region of chrX.</p>
</dd>
<dt class="hdlist1">gvcfz</dt>
<dd>
<p>compress gVCF file by resizing non-variant blocks according to specified criteria</p>
</dd>
<dt class="hdlist1">impute-info</dt>
<dd>
<p>Adds imputation information metrics to the INFO field based on selected FORMAT tags. Only the IMPUTE2 INFO metric from FORMAT/GP tags is currently available.</p>
</dd>
<dt class="hdlist1">indel-stats</dt>
<dd>
<p>Calculates per-sample or de novo indels stats. The usage and format is similar to <code>smpl-stats</code> and <code>trio-stats</code>.</p>
</dd>
<dt class="hdlist1">isecGT</dt>
<dd>
<p>Compares two files and set non-identical genotypes in the first file to missing.</p>
</dd>
<dt class="hdlist1"><a href="plugin.mendelian.html">mendelian</a></dt>
<dd>
<p>Count Mendelian consistent / inconsistent genotypes</p>
</dd>
<dt class="hdlist1">missing2ref</dt>
<dd>
<p>sets missing genotypes ("./.") to ref allele ("0/0" or "0|0"). DEPRECATED, use setGT instead.</p>
</dd>
<dt class="hdlist1">parental-origin</dt>
<dd>
<p>determine parental origin of a CNV region</p>
</dd>
<dt class="hdlist1">prune</dt>
<dd>
<p>Prune sites by missingness, allele frequency or linkage disequilibrium. Alternatively, annotate sites with r2, Lewontin&#8217;s D' (PMID:19433632), Ragsdale&#8217;s D (PMID:31697386).</p>
</dd>
<dt class="hdlist1"><a href="plugin.setGT.html">setGT</a></dt>
<dd>
<p>Sets genotypes according to the specified criteria and filtering expressions. For example, missing genotypes can be set to ref, but much more than that.</p>
</dd>
<dt class="hdlist1">smpl-stats</dt>
<dd>
<p>calculates basic per-sample stats. The usage and format is similar to <code>indel-stats</code> and <code>trio-stats</code>.</p>
</dd>
<dt class="hdlist1">split</dt>
<dd>
<p>split VCF by sample, creating single-sample VCFs</p>
</dd>
<dt class="hdlist1"><a href="plugin.split-vep.html">split-vep</a></dt>
<dd>
<p>extract fields from structured annotations such as INFO/CSQ created by bcftools/csq or VEP. These can be added as a new INFO field to the VCF or in a custom text format</p>
</dd>
<dt class="hdlist1"><a href="plugin.tag2tag.html">tag2tag</a></dt>
<dd>
<p>Converts between similar tags, such as GL,PL,GP or QR,QA,QS or localized alleles, eg LPL,LAD.</p>
</dd>
<dt class="hdlist1"><a href="plugin.trio-dnm2.html">trio-dnm2</a></dt>
<dd>
<p>screen variants for possible de-novo mutations in trios</p>
</dd>
<dt class="hdlist1">trio-stats</dt>
<dd>
<p>Calculate transmission rate in trio children. The usage and format is similar to <code>indel-stats</code> and <code>smpl-stats</code>.</p>
</dd>
<dt class="hdlist1">trio-switch-rate</dt>
<dd>
<p>Calculates phase switch rate in trio children.</p>
</dd>
<dt class="hdlist1">variantkey-hex</dt>
<dd>
<p>generate unsorted VariantKey-RSid index files in hexadecimal format</p>
</dd>
<dt class="hdlist1">vrfs</dt>
<dd>
<p>Assess site noisiness (variant read frequency score) from a large number of reference samples</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="_feedback">Feedback</h3>
<div class="paragraph">
<p>We welcome your feedback, please help us improve this page by 
either opening an <a href="https://github.com/samtools/bcftools/issues">issue on github</a> or <a href="https://github.com/samtools/bcftools/tree/gh-pages/howtos">editing it directly</a> and sending
a pull request.</p>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="footer">
<div id="footer-text">
</div>
</div>
</body>
</html>