<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Asciidoctor 2.0.15.dev">
<title>Detecting runs of homozygosity (RoH)</title>
<link rel="stylesheet" href="./index.css">
</head>
<body class="article">
<div id="header">
</div>
<div id="content">
<div class="sidebarblock navig">
<div class="content">
<div class="ulist">
<div class="title">General</div>
<ul>
<li>
<p><a href="index.html">Main page</a></p>
</li>
<li>
<p><a href="../bcftools.html">Manual page</a></p>
</li>
<li>
<p><a href="install.html">Installation</a></p>
</li>
<li>
<p><a href="publications.html">Publications</a></p>
</li>
</ul>
</div>
<div class="ulist">
<div class="title">Calling</div>
<ul>
<li>
<p><a href="cnv-calling.html">CNV calling</a></p>
</li>
<li>
<p><a href="csq-calling.html">Consequence calling</a></p>
</li>
<li>
<p><a href="consensus-sequence.html">Consensus calling</a></p>
</li>
<li>
<p><a href="roh-calling.html">ROH calling</a></p>
</li>
<li>
<p><a href="variant-calling.html">Variant calling and filtering</a></p>
</li>
</ul>
</div>
<div class="ulist">
<div class="title">Tips and Tricks</div>
<ul>
<li>
<p><a href="convert.html">Converting formats</a></p>
</li>
<li>
<p><a href="annotate.html">Adding annotation</a></p>
</li>
<li>
<p><a href="query.html">Extracting information</a></p>
</li>
<li>
<p><a href="filtering.html">Filtering expressions</a></p>
</li>
<li>
<p><a href="plugins.html">Plugins</a></p>
</li>
<li>
<p><a href="FAQ.html">FAQ</a></p>
</li>
</ul>
</div>
</div>
</div>
<div id="main">
<div class="sect1">
<h2 id="_detecting_runs_of_homozygosity_roh">Detecting runs of homozygosity (RoH)</h2>
<div class="sectionbody">
<div class="paragraph">
<p>The BCFtools/RoH command detects regions of autozygosity in sequencing data, including
exome data, using a hidden Markov model.</p>
</div>
<div class="sect2">
<h3 id="_preparing_input_data">Preparing input data</h3>
<div class="paragraph">
<p>The <code>roh</code> command takes on input VCF with FORMAT columns containing either
genotype likelihoods (PL) or genotypes (GT). By default, genotype likelihoods
are expected unless the <code>-G</code> option is provided.</p>
</div>
<div class="paragraph">
<p>The second required information is the estimate of the alternate allele
frequencies in the population for each site. There are two ways to obtain this
information. One is to supply a VCF with many samples from the same population
(preferably &gt;100, though often it works down to just 20 individuals or
so). This allows the frequencies to be estimated from the samples provided.
Otherwise, the allele frequency information for your specific population can be
provided from other datasets: from a custom VCF annotation (<code>--AF-tag</code>),
from a tab-delimited file (<code>--AF-file</code>) or a fixed default can be used
if none of the previous is available (<code>--AF-dflt</code>).</p>
</div>
<div class="paragraph">
<p>Finally, a genetic map can be provided (<code>--genetic-map</code>) so that the model
can account for recombination hotspots. If not available, a constant
recombination rate per-base can be given (<code>--rec-rate</code>).
Note that also both options can be given. In that case, the <code>--rec-rate</code> parameter
will be interpreted differently by the program - as a fold increase of transition probabilities. This
allows the model to become more sensitive yet still account for recombination hotspots.
Of course, also the range of the values given to <code>--rec-rate</code> must be different in both cases:
it will be typically in the range (1e-9,1e-3) without the <code>--genetic-map</code> option
and in the range (10,1000) with <code>--genetic-map</code>. In both cases bigger values result in more sensitive calls.</p>
</div>
<div class="paragraph">
<p>The program considers only biallelic sites, sites with non-zero AF,
and sites with non-missing PL/GT.</p>
</div>
</div>
<div class="sect2">
<h3 id="_example_1">Example 1</h3>
<div class="paragraph">
<p>In this example, the FORMAT/PL annotation is not present, therefore we must use FORMAT/GT,
see the -G option.  Also, the VCF does not contain allele frequency information
and there is just one sample so it cannot be estimated on the fly. Hence we use
a default AF value 0.4.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>bcftools roh -G30 --AF-dflt 0.4 file.vcf</pre>
</div>
</div>
<div class="paragraph">
<p>Please check this <a href="roh-calling/usage-example.tgz">usage example</a> for details
and some test data to experiment with.</p>
</div>
<div class="exampleblock">
<div class="content">
<a href="roh-calling/roh-example.png"><img src="roh-calling/roh-example.png" width="800px"></a>
<div class="paragraph">
<p>In this figure, the blue line shows heterozygosity plotted as dosage and the RoH estimate by bcftools/roh.</p>
</div>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_example_2">Example 2</h3>
<div class="paragraph">
<p>In the second example we use the allele frequencies from the
<a href="http://www.internationalgenome.org">1000 Genomes Project</a>
and the genetic map distributed with
<a href="https://mathgen.stats.ox.ac.uk/impute/impute_v2.html#reference">Impute2</a>
(<a href="https://mathgen.stats.ox.ac.uk/impute/1000GP_Phase3.tgz">1000GP_Phase3.tgz</a>).
Because these files are big and require some (although easy) pre-processing, a packaged version
can be downloaded from here: <a href="https://ngs.sanger.ac.uk/production/samtools/1000GP-AFs.tgz">allele frequencies</a> and <a href="https://ngs.sanger.ac.uk/production/samtools/genetic-map.tgz">genetic map</a>.</p>
</div>
<div class="listingblock">
<div class="content">
<pre># Annotate VCF with the AFs and run ROH
bcftools annotate -c CHROM,POS,REF,ALT,AF1KG -h AFs.tab.gz.hdr -a AFs.tab.gz in.bcf | \
   bcftools roh --AF-tag AF1KG -M 100 -m genetic-map/genetic_map_chr{CHROM}_combined_b37.txt -o roh.txt</pre>
</div>
</div>
<div class="paragraph">
<p>In case a different version is needed, the following commands show how the pre-packaged
<a href="https://ngs.sanger.ac.uk/production/samtools/genetic-map.tgz">genetic map</a> and
<a href="https://ngs.sanger.ac.uk/production/samtools/1000GP-AFs.v2023-11-02.tgz">allele frequencies</a> were created:</p>
</div>
<div class="listingblock">
<div class="content">
<pre># Get the genetic map
curl -O https://mathgen.stats.ox.ac.uk/impute/1000GP_Phase3.tgz
tar --wildcards -xzf 1000GP_Phase3.tgz 1000GP_Phase3/genetic_map_chr\*
mv 1000GP_Phase3 genetic-map
rm -f 1000GP_Phase3.tgz

# Get alleles frequencies from the 1000 Genomes Project
curl -Lo AFs.vcf.gz https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.wgs.phase3_shapeit2_mvncall_integrated_v5c.20130502.sites.vcf.gz
bcftools query -f'%CHROM\t%POS\t%REF,%ALT\t%AF\n' AFs.vcf.gz | bgzip -c &gt; AFs.tab.gz
tabix -s1 -b2 -e2 AFs.tab.gz
echo '##INFO=&lt;ID=AF1KG,Number=1,Type=Float,Description="Allele Frequency"&gt;' &gt; AFs.tab.gz.hdr
rm -f AFs.vcf.gz</pre>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_example_3">Example 3</h3>
<div class="paragraph">
<p>Sometimes multiple samples need to be analyzed and compared. There is the convenience
script <a href="https://github.com/samtools/bcftools/blob/develop/misc/run-roh.pl">run-roh.pl</a>
which takes multiple VCF/BCF files and locates regions private to one sample or shared across multiple samples.
On input it expects a directory with .vcf, .vcf.gz or .bcf files, a file with allele frequencies
and a genetic map (optional).  The calls can be then plotted or visualized interactively using
the <a href="https://github.com/samtools/bcftools/blob/develop/misc/plot-roh.py">plot-roh.py</a> script</p>
</div>
<div class="listingblock">
<div class="content">
<pre>run-roh.pl -i input-dir -o output-dir
plot-roh.py -i output-dir</pre>
</div>
</div>
<div class="paragraph">
<p>An example of the output is shown below. In the first window, all chromosomes from three samples
are shown. Each dot corresponds to one marker with either the alternate homozygous genotype (top) or the
heterozygous genotype (bottom). The grey and red rectangles highlight regions predicted by <code>bcftools roh</code>
as homozygous - they should contain few heterozygous genotypes. The grey rectangles highlight homozygous
regions private to one sample, while the red rectangles show homozygous regions shared across multiple samples.
In the second window is chromosome 17, zoomed-in using the button with the magnifying glass icon.</p>
</div>
<div class="exampleblock">
<div class="content">
<a href="roh-calling/plot-roh-1.jpg"><img src="roh-calling/plot-roh-1.jpg" width="800px"></a>
<a href="roh-calling/plot-roh-2.jpg"><img src="roh-calling/plot-roh-2.jpg" width="800px"></a>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_troubleshooting">Troubleshooting</h3>
<div class="paragraph">
<p>If the results look unexpected, check how many sites have been ignored by the
program.  For example in this run many sites were filtered:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>Number of lines: total/processed: 599218/37730</pre>
</div>
</div>
<div class="paragraph">
<p>If the number of the processed sites is too low, check what was the reason for exluding
them. This command should give the number of sites that were processed:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>bcftools view -i'type="SNP" &amp;&amp; AF&gt;0' -m2 -M2 -s Indiv1 -Ou file.vcf | bcftools +counts -i'GT!="./." &amp;&amp; GT!="."</pre>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_references">References</h3>
<div class="paragraph">
<p>Please cite this paper if you find our software useful: <a href="http://www.ncbi.nlm.nih.gov/pubmed/26826718" class="bare">http://www.ncbi.nlm.nih.gov/pubmed/26826718</a></p>
</div>
</div>
<div class="sect2">
<h3 id="_feedback">Feedback</h3>
<div class="paragraph">
<p>We welcome your feedback, please help us improve this page by 
either opening an <a href="https://github.com/samtools/bcftools/issues">issue on github</a> or <a href="https://github.com/samtools/bcftools/tree/gh-pages/howtos">editing it directly</a> and sending
a pull request.</p>
</div>
</div>
</div>
</div>
</div>
</div>
<div id="footer">
<div id="footer-text">
</div>
</div>
</body>
</html>