<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Asciidoctor 2.0.16.dev">
<title>bcftools(1)</title>
<link rel="stylesheet" href="./docbook-xsl.css">
</head>
<body class="manpage">
<div id="header">
<h1>bcftools(1) Manual Page</h1>
<h2 id="_name">NAME</h2>
<div class="sectionbody">
<p>bcftools - utilities for variant calling and manipulating VCFs and BCFs.</p>
</div>
</div>
<div id="content">
<div class="sect1">
<h2 id="_synopsis">SYNOPSIS</h2>
<div class="sectionbody">
<div class="paragraph">
<p><strong>bcftools</strong> [--version|--version-only] [--help] [<em>COMMAND</em>] [<em>OPTIONS</em>]</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_description">DESCRIPTION</h2>
<div class="sectionbody">
<div class="paragraph">
<p>BCFtools  is  a set of utilities that manipulate variant calls in the Variant
Call Format (VCF) and its binary counterpart BCF. All commands work
transparently with both VCFs and BCFs, both uncompressed and BGZF-compressed.</p>
</div>
<div class="paragraph">
<p>Most commands accept VCF, bgzipped VCF and BCF with filetype detected
automatically even when streaming from a pipe. Indexed VCF and BCF
will work in all situations. Un-indexed VCF and BCF and streams will
work in most, but not all situations. In general, whenever multiple VCFs are
read simultaneously, they must be indexed and therefore also compressed.
(Note that files with non-standard index names can be accessed as e.g.
"<code>bcftools view -r X:2928329 file.vcf.gz##idx##non-standard-index-name</code>".)</p>
</div>
<div class="paragraph">
<p>BCFtools is designed to work on a stream. It regards an input file "-" as the
standard input (stdin) and outputs to the standard output (stdout). Several
commands can thus be  combined  with  Unix pipes.</p>
</div>
<div class="sect2">
<h3 id="_version">VERSION</h3>
<div class="paragraph">
<p>This manual page was last updated <strong>2024-09-12</strong> and refers to bcftools git version <strong>1.21</strong>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_bcf1">BCF1</h3>
<div class="paragraph">
<p>The obsolete BCF1 format output by versions of samtools &lt;= 0.1.19 is <strong>not</strong>
compatible with this version of bcftools. To read BCF1 files one can use
the view command from old versions of bcftools packaged with samtools
versions &lt;= 0.1.19 to convert to VCF, which can then be read by
this version of bcftools.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>    samtools-0.1.19/bcftools/bcftools view file.bcf1 | bcftools view</pre>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_variant_calling">VARIANT CALLING</h3>
<div class="paragraph">
<p>See <em>bcftools call</em> for variant calling from the output of the
<em>samtools mpileup</em> command. In versions of samtools &lt;= 0.1.19 calling was
done with <em>bcftools view</em>. Users are now required to choose between the old
samtools calling model (<em>-c/--consensus-caller</em>) and the new multiallelic
calling model (<em>-m/--multiallelic-caller</em>). The multiallelic calling model
is recommended for most tasks.</p>
</div>
</div>
<div class="sect2">
<h3 id="_filtering_expressions">FILTERING EXPRESSIONS</h3>
<div class="paragraph">
<p>See <strong><a href="#expressions">EXPRESSIONS</a></strong></p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_list_of_commands">LIST OF COMMANDS</h2>
<div class="sectionbody">
<div class="paragraph">
<p>For a full list of available commands, run <strong>bcftools</strong> without arguments. For a full
list of available options, run <strong>bcftools</strong> <em>COMMAND</em> without arguments.</p>
</div>
<div class="ulist">
<ul>
<li>
<p><strong><a href="#annotate">annotate</a></strong>  ..  edit VCF files, add or remove annotations</p>
</li>
<li>
<p><strong><a href="#call">call</a></strong>      ..  SNP/indel calling (former "view")</p>
</li>
<li>
<p><strong><a href="#cnv">cnv</a></strong>       ..  Copy Number Variation caller</p>
</li>
<li>
<p><strong><a href="#concat">concat</a></strong>    ..  concatenate VCF/BCF files from the same set of samples</p>
</li>
<li>
<p><strong><a href="#consensus">consensus</a></strong> ..  create consensus sequence by applying VCF variants</p>
</li>
<li>
<p><strong><a href="#convert">convert</a></strong>   ..  convert VCF/BCF to other formats and back</p>
</li>
<li>
<p><strong><a href="#csq">csq</a></strong>       ..  haplotype aware consequence caller</p>
</li>
<li>
<p><strong><a href="#filter">filter</a></strong>    ..  filter VCF/BCF files using fixed thresholds</p>
</li>
<li>
<p><strong><a href="#gtcheck">gtcheck</a></strong>   ..  check sample concordance, detect sample swaps and contamination</p>
</li>
<li>
<p><strong><a href="#head">head</a></strong>      ..  view VCF/BCF file headers</p>
</li>
<li>
<p><strong><a href="#index">index</a></strong>     ..  index VCF/BCF</p>
</li>
<li>
<p><strong><a href="#isec">isec</a></strong>      ..  intersections of VCF/BCF files</p>
</li>
<li>
<p><strong><a href="#merge">merge</a></strong>     ..  merge VCF/BCF files files from non-overlapping sample sets</p>
</li>
<li>
<p><strong><a href="#mpileup">mpileup</a></strong>   ..  multi-way pileup producing genotype likelihoods</p>
</li>
<li>
<p><strong><a href="#norm">norm</a></strong>      ..  normalize indels</p>
</li>
<li>
<p><strong><a href="#plugin">plugin</a></strong>    ..  run user-defined plugin</p>
</li>
<li>
<p><strong><a href="#polysomy">polysomy</a></strong>  ..  detect contaminations and whole-chromosome aberrations</p>
</li>
<li>
<p><strong><a href="#query">query</a></strong>     ..  transform VCF/BCF into user-defined formats</p>
</li>
<li>
<p><strong><a href="#reheader">reheader</a></strong>  ..  modify VCF/BCF header, change sample names</p>
</li>
<li>
<p><strong><a href="#roh">roh</a></strong>       ..  identify runs of homo/auto-zygosity</p>
</li>
<li>
<p><strong><a href="#sort">sort</a></strong>      ..  sort VCF/BCF files</p>
</li>
<li>
<p><strong><a href="#stats">stats</a></strong>     ..  produce VCF/BCF stats (former vcfcheck)</p>
</li>
<li>
<p><strong><a href="#view">view</a></strong>      ..  subset, filter and convert VCF and BCF files</p>
</li>
</ul>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_list_of_scripts">LIST OF SCRIPTS</h2>
<div class="sectionbody">
<div class="paragraph">
<p>Some helper scripts are bundled with the bcftools code.</p>
</div>
<div class="ulist">
<ul>
<li>
<p><strong><a href="#gff2gff">gff2gff</a></strong>  .. converts a GFF file to the format required by <strong><a href="#csq">csq</a></strong></p>
</li>
<li>
<p><strong><a href="#plot-vcfstats">plot-vcfstats</a></strong>  .. plots the output of <strong><a href="#stats">stats</a></strong></p>
</li>
</ul>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_commands_and_options">COMMANDS AND OPTIONS</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="common_options">Common Options</h3>
<div class="paragraph">
<p>The following options are common to many bcftools commands. See usage for
specific commands to see if they apply.</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><em>FILE</em></dt>
<dd>
<p>Files can be both VCF or BCF, uncompressed or BGZF-compressed. The file "-"
is interpreted as standard input. Some tools may require tabix- or
CSI-indexed files.</p>
</dd>
<dt class="hdlist1"><strong>-c, --collapse</strong> <em>snps</em>|<em>indels</em>|<em>both</em>|<em>all</em>|<em>some</em>|<em>none</em>|<em>id</em></dt>
<dd>
<p>Controls  how to treat records with duplicate positions and defines compatible
records across multiple input files. Here by "compatible" we mean records which
should be considered as identical by the tools. For example, when performing
line intersections, the desire may be to consider as identical all sites with
matching positions (<strong>bcftools isec -c</strong> <em>all</em>), or only sites with  matching variant
type (<strong>bcftools isec -c</strong> <em>snps</em>&#160; <strong>-c</strong> <em>indels</em>), or only sites with all alleles
identical (<strong>bcftools isec -c</strong> <em>none</em>).</p>
<div class="dlist">
<dl>
<dt class="hdlist1"><em>none</em></dt>
<dd>
<p>only records with identical REF and ALT alleles are compatible</p>
</dd>
<dt class="hdlist1"><em>some</em></dt>
<dd>
<p>only records where some subset of ALT alleles match are compatible</p>
</dd>
<dt class="hdlist1"><em>all</em></dt>
<dd>
<p>all records are compatible, regardless of whether the ALT alleles
match or not. In the case of records with the same position, only
the first will be considered and appear on output.</p>
</dd>
<dt class="hdlist1"><em>snps</em></dt>
<dd>
<p>any SNP records are compatible, regardless of whether the ALT
alleles match or not. For duplicate positions, only the first SNP
record will be considered and appear on output.</p>
</dd>
<dt class="hdlist1"><em>indels</em></dt>
<dd>
<p>all  indel records are compatible, regardless of whether the REF
and ALT alleles match or not. For duplicate positions, only the
first indel record will be considered and appear on output.</p>
</dd>
<dt class="hdlist1"><em>both</em></dt>
<dd>
<p>abbreviation of "<strong>-c</strong> <em>indels</em>&#160; <strong>-c</strong> <em>snps</em>"</p>
</dd>
<dt class="hdlist1"><em>id</em></dt>
<dd>
<p>only records with identical ID column are compatible.
Supported by <strong><a href="#merge">bcftools merge</a></strong> only.</p>
</dd>
</dl>
</div>
</dd>
<dt class="hdlist1"><strong>-f, --apply-filters</strong> <em>LIST</em></dt>
<dd>
<p>Skip sites where FILTER column does not contain any of the strings listed
in <em>LIST</em>. For example, to include only sites which have no filters set,
use <strong>-f</strong> <em>.,PASS</em>.</p>
</dd>
<dt class="hdlist1"><strong>--no-version</strong></dt>
<dd>
<p>Do not append version and command line information to the output VCF header.</p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>When output consists of a single stream, write it to <em>FILE</em> rather than
to standard output, where it is written by default.
The file type is determined automatically from the file name suffix and in
case a conflicting <strong>-O</strong> option is given, the file name suffix takes precedence.</p>
</dd>
<dt class="hdlist1"><strong>-O, --output-type</strong> <em>b</em>|<em>u</em>|<em>z</em>|<em>v</em>[0-9]</dt>
<dd>
<p>Output compressed BCF (<em>b</em>), uncompressed BCF (<em>u</em>), compressed VCF (<em>z</em>), uncompressed VCF (<em>v</em>).
Use the -Ou option when piping between bcftools subcommands to speed up
performance by removing unnecessary compression/decompression and
VCF&#8592;&#8594;BCF conversion.
&#160;
The compression level of the compressed formats (<em>b</em> and <em>z</em>) can be set by
by appending a number between 0-9.</p>
</dd>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:beg-end</em>|<em>chr:beg-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>Comma-separated list of regions, see also <strong>-R, --regions-file</strong>. Overlapping
records are matched even when the starting coordinate is outside of the
region, unlike the <strong>-t/-T</strong> options where only the POS coordinate is checked.
Note that <strong>-r</strong> cannot be used in combination with <strong>-R</strong>.</p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>FILE</em></dt>
<dd>
<p>Regions can be specified either on command line or in a VCF, BED, or
tab-delimited file (the default). The columns of the tab-delimited file
can contain either positions (two-column format: CHROM, POS) or intervals
(three-column format: CHROM, BEG, END), but not both. Positions are 1-based
and inclusive. The columns of the tab-delimited BED file are also
CHROM, POS and END (trailing columns are ignored), but coordinates
are 0-based, half-open.  To indicate that a file be treated as BED rather
than the 1-based tab-delimited file, the file must have the ".bed" or
".bed.gz" suffix (case-insensitive). Uncompressed files are stored in
memory, while bgzip-compressed and tabix-indexed region files are streamed.
Note that sequence names must match exactly, "chr20" is not the same as
"20".  Also note that chromosome ordering in <em>FILE</em> will be respected,
the VCF will be processed in the order in which chromosomes first appear
in <em>FILE</em>. However, within chromosomes, the VCF will always be
processed in ascending genomic coordinate order no matter what order they
appear in <em>FILE</em>. Note that overlapping regions in <em>FILE</em> can result in
duplicated out of order positions in the output.
This option requires indexed VCF/BCF files. Note that <strong>-R</strong> cannot be used
in combination with <strong>-r</strong>.</p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>pos</em>|<em>record</em>|<em>variant</em>|<em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>This option controls how overlapping records are determined:
set to <strong>pos</strong> or <strong>0</strong> if the VCF record has to have POS inside a region
(this corresponds to the default behavior of <strong>-t/-T</strong>);
set to <strong>record</strong> or <strong>1</strong> if also overlapping records with POS outside a region
should be included (this is the default behavior of <strong>-r/-R</strong>, and includes indels
with POS at the end of a region, which are technically outside the region); or set
to <strong>variant</strong> or <strong>2</strong> to include only true overlapping variation (compare
the full VCF representation "<code>TA&gt;T-</code>" vs the true sequence variation "<code>A&gt;-</code>").</p>
</dd>
<dt class="hdlist1"><strong>-s, --samples</strong> [^]<em>LIST</em></dt>
<dd>
<p>Comma-separated list of samples to include or exclude if prefixed
with "^." (Note that when multiple samples are to be excluded,
the "^" prefix is still present only once, e.g.  "^SAMPLE1,SAMPLE2".)
The sample order is updated to reflect that given on the command line.
Note that in general tags such as INFO/AC, INFO/AN, etc are not updated
to correspond to the subset samples. <strong><a href="#view">bcftools view</a></strong> is the
exception where some tags will be updated (unless the <strong>-I, --no-update</strong>
option is used; see <strong><a href="#view">bcftools view</a></strong> documentation). To use updated
tags for the subset in another command one can pipe from <strong>view</strong> into
that command. For example:</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    bcftools view -Ou -s sample1,sample2 file.vcf | bcftools query -f %INFO/AC\t%INFO/AN\n</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-S, --samples-file</strong> [^]<em>FILE</em><a id="samples_file"></a></dt>
<dd>
<p>File of sample names to include or exclude if prefixed with "^".
One sample per line. See also the note above for the <strong>-s, --samples</strong>
option.
The sample order is updated to reflect that given in the input file.
The command <strong><a href="#call">bcftools call</a></strong> accepts an optional second
column indicating ploidy (0, 1 or 2) or sex (as defined by
<strong><a href="#ploidy">--ploidy</a></strong>, for example "F" or "M"), for example:</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    sample1    1
    sample2    2
    sample3    2</pre>
</div>
</div>
<div class="paragraph">
<p>or</p>
</div>
<div class="listingblock">
<div class="content">
<pre>    sample1    M
    sample2    F
    sample3    F</pre>
</div>
</div>
<div class="paragraph">
<p>If the second column is not present, the sex "F" is assumed.
With <strong><a href="#call">bcftools call</a> -C</strong> <em>trio</em>, PED file is expected.
The program ignores the first column and the last indicates sex (1=male, 2=female), for example:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>    ignored_column  daughterA fatherA  motherA  2
    ignored_column  sonB      fatherB  motherB  1</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-t, --targets</strong> [^]<em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>Similar as <strong>-r, --regions</strong>, but the next position is accessed by streaming the
whole VCF/BCF rather than using the tbi/csi index. Both <strong>-r</strong> and <strong>-t</strong> options
can be applied simultaneously: <strong>-r</strong>  uses  the index  to  jump  to  a  region
and <strong>-t</strong> discards positions which are not in the targets. Unlike <strong>-r</strong>, targets
can be prefixed with "^" to request logical complement. For example, "^X,Y,MT"
indicates that sequences X, Y and MT should be skipped.
Yet another difference between the <strong>-t/-T</strong> and <strong>-r/-R</strong> is that <strong>-r/-R</strong> checks for
proper overlaps and considers both POS and the end position of an indel, while <strong>-t/-T</strong>
considers the POS coordinate only (by default; see also <strong>--regions-overlap</strong> and <strong>--targets-overlap</strong>).
Note that <strong>-t</strong> cannot be used in combination with <strong>-T</strong>.</p>
</dd>
<dt class="hdlist1"><strong>-T, --targets-file</strong> [^]<em>FILE</em></dt>
<dd>
<p>Same <strong>-t, --targets</strong>, but reads regions from a file. Note that <strong>-T</strong>
cannot be used in combination with <strong>-t</strong>.</p>
<div class="paragraph">
<p>With the <strong>call -C</strong> <em>alleles</em> command, third column of the targets file must
be comma-separated list of alleles, starting with the reference allele.
Note that the file must be compressed and indexed.
Such a file can be easily created from a VCF using:</p>
</div>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    bcftools query -f'%CHROM\t%POS\t%REF,%ALT\n' file.vcf | bgzip -c &gt; als.tsv.gz &amp;&amp; tabix -s1 -b2 -e2 als.tsv.gz</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--targets-overlap</strong> <em>pos</em>|<em>record</em>|<em>variant</em>|<em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>Same as <strong>--regions-overlap</strong> but for <strong>-t/-T</strong>.</p>
</dd>
<dt class="hdlist1"><strong>--threads</strong> <em>INT</em></dt>
<dd>
<p>Use multithreading with <em>INT</em> worker threads. The option is currently used only for the compression of the
output stream, only when <em>--output-type</em> is <em>b</em> or <em>z</em>. Default: 0.</p>
</dd>
<dt class="hdlist1"><strong>-W</strong>[<em>FMT</em>]<strong>, -W</strong>[=<em>FMT</em>]<strong>, --write-index</strong>[=<em>FMT</em>]</dt>
<dd>
<p>Automatically index the output files. <em>FMT</em> is optional and can be
one of "tbi" or "csi" depending on output file format. Defaults to
CSI unless specified otherwise. Can be used only for compressed
BCF and VCF output.</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="annotate">bcftools annotate <em>[OPTIONS]</em> <em>FILE</em></h3>
<div class="paragraph">
<p>Add or remove annotations.</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-a, --annotations</strong> <em>file</em></dt>
<dd>
<p>Bgzip-compressed and tabix-indexed file with annotations. The file
can be VCF, BED, or a tab-delimited file with mandatory columns CHROM, POS
(or, alternatively, FROM and TO), optional columns REF and ALT, and arbitrary
number of annotation columns. BED files are expected to have
the ".bed" or ".bed.gz" suffix (case-insensitive), otherwise a tab-delimited file is assumed.
Note that in case of tab-delimited file, the coordinates POS, FROM and TO are
one-based and inclusive.  When REF and ALT are present, only matching VCF
records will be annotated. If the END coordinate is present in the annotation file
and given on command line as "<code>-c ~INFO/END</code>", then VCF records will be matched also by the INFO/END coordinate.
If ID is present in the annotation file and given as "<code>-c ~ID</code>", then VCF records will be matched
also by the ID column.
&#160;<br>
&#160;<br>
When multiple ALT alleles are present in the annotation file (given as
comma-separated list of alleles), at least one must match one of the
alleles in the corresponding VCF record. Similarly, at least one
alternate allele from a multi-allelic VCF record must be present in the
annotation file.
&#160;<br>
&#160;<br>
Missing values can be added by providing "." in place of actual value
and using the missing value modifier with <strong>-c</strong>, such as ".TAG".
&#160;<br>
&#160;<br>
Note that flag types, such as "INFO/FLAG", can be annotated by including
a field with the value "1" to set the flag, "0" to remove it, or "." to
keep existing flags.
See also <strong>-c, --columns</strong> and <strong>-h, --header-lines</strong>.</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    # Sample annotation file with columns CHROM, POS, STRING_TAG, NUMERIC_TAG
    1  752566  SomeString      5
    1  798959  SomeOtherString 6</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-c, --columns</strong> <em>list</em></dt>
<dd>
<p>Comma-separated list of columns or tags to carry over from the annotation file
(see also <strong>-a, --annotations</strong>). If the annotation file is not a VCF/BCF,
<em>list</em> describes the columns of the annotation file and must include CHROM,
POS (or, alternatively, FROM,TO or BEG,END), and optionally REF and ALT. Unused
columns which should be ignored can be indicated by "-".
&#160;<br>
&#160;<br>
If the annotation file is a VCF/BCF, only the edited columns/tags must be present and their
order does not matter. The columns ID, QUAL, FILTER, INFO and FORMAT
can be edited, where INFO tags can be written both as "INFO/TAG" or simply "TAG",
and FORMAT tags can be written as "FORMAT/TAG" or "FMT/TAG".
The imported VCF annotations can be renamed as "DST_TAG:=SRC_TAG" or "FMT/DST_TAG:=FMT/SRC_TAG".
&#160;<br>
&#160;<br>
To carry over all INFO annotations, use "INFO". To add all INFO annotations except
"TAG", use "^INFO/TAG". By default, existing values are replaced.
&#160;<br>
&#160;<br>
By default, existing tags are overwritten unless the source value is a missing value (i.e. ".").
If also missing values should be carried over (and overwrite existing tags), use ".TAG" instead of "TAG".
To add annotations without overwriting existing values (that is, to add tags that are absent or
to add values to existing tags with missing values), use "+TAG" instead of "TAG". These can be combined,
for example ".+TAG" can be used to add TAG even if the source value is missing but only if TAG does not
exist in the target file; existing tags will not be overwritten.
To append to existing values (rather than replacing or leaving untouched), use "=TAG"
(instead of "TAG" or "+TAG").
To replace only existing values without modifying missing annotations, use "-TAG".
As a special case of this, if position needs to be replaced, mark the column with the new coordinate as "-POS".
(Note that in previous releases this used to be "~POS", now deprecated.)
&#160;<br>
&#160;<br>
To match the record also by ID or INFO/END, in addition to REF and ALT, use "~ID" or "~INFO/END".
Note that this works only for ID and POS, for other fields see the description of <strong>-i</strong> below.
&#160;<br>
&#160;<br>
If the annotation file is not a VCF/BCF, all new annotations must be
defined via <strong>-h, --header-lines</strong>.
&#160;<br>
&#160;<br>
See also the <strong>-l, --merge-logic</strong> option.
&#160;<br>
&#160;<br>
<strong>Summary of <code>-c, --columns</code>:</strong></p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    CHROM,POS,TAG       .. match by chromosome and position, transfer annotation from TAG
    CHROM,POS,-,TAG     .. same as above, but ignore the third column of the annotation file
    CHROM,BEG,END,TAG   .. match by region (BEG,END are synonymous to FROM,TO)
    CHROM,POS,REF,ALT   .. match by CHROM, POS, REF and ALT

    DST_TAG:=SRC_TAG    .. transfer the SRC_TAG using the new name DST_TAG
    INFO                .. transfer all INFO annotations
    ^INFO/TAG           .. transfer all INFO annotations except "TAG"

    TAG       .. add or overwrite existing target value if source is not "." and skip otherwise
    +TAG      .. add or overwrite existing target value only it is "."
    .TAG      .. add or overwrite existing target value even if source is "."
    .+TAG     .. add new but never overwrite existing tag, regardless of its value; can transfer "." if target does not exist
    -TAG      .. overwrite existing value, never add new if target does not exist
    =TAG      .. do not overwrite but append value to existing tags

    ~FIELD    .. use this column to match lines with -i/-e expression (see the description of -i below)
    ~ID       .. in addition to CHROM,POS,REF,ALT match by also ID
    ~INFO/END .. in addition to CHROM,POS,REF,ALT match by also INFO/END</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-C, --columns-file</strong> <em>file</em></dt>
<dd>
<p>Read the list of columns from a file (normally given via the <strong>-c, --columns</strong> option).
"-" to skip a column of the annotation file.
One column name per row, an additional space- or tab-separated field can
be present to indicate the merge logic (normally given via the <strong>-l, --merge-logic</strong> option).
This is useful when many annotations are added at once.</p>
</dd>
<dt class="hdlist1"><strong>-e, --exclude</strong> <em>EXPRESSION</em></dt>
<dd>
<p>exclude sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong> and the extension described in <strong>-i, --include</strong> below.</p>
</dd>
<dt class="hdlist1"><strong>--force</strong></dt>
<dd>
<p>continue even when parsing errors, such as undefined tags, are encountered. Note
this can be an unsafe operation and can result in corrupted BCF files. If this
option is used, make sure to sanity check the result thoroughly.</p>
</dd>
<dt class="hdlist1"><strong>-h, --header-lines</strong> <em>file</em></dt>
<dd>
<p>Lines to append to the VCF header, see also <strong>-c, --columns</strong> and <strong>-a, --annotations</strong>. For example:</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    ##INFO=&lt;ID=NUMERIC_TAG,Number=1,Type=Integer,Description="Example header line"&gt;
    ##INFO=&lt;ID=STRING_TAG,Number=1,Type=String,Description="Yet another header line"&gt;</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-I, --set-id</strong> [+]<em>FORMAT</em></dt>
<dd>
<p>assign ID on the fly. The format is the same as in the <strong><a href="#query">query</a></strong>
command (see below).  By default all existing IDs are replaced. If the
format string is preceded by "+", only missing IDs will be set. For example,
one can use</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    bcftools annotate --set-id +'%CHROM\_%POS\_%REF\_%FIRST_ALT' file.vcf</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --include</strong> <em>EXPRESSION</em></dt>
<dd>
<p>include only sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.
&#160;<br>
&#160;<br>
Additionally, the command <strong>bcftools annotate</strong> supports expressions updated from the annotation
file dynamically for each record:</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    # The field 'STR' from the -a file is required to match INFO/TAG in VCF. In the first example
    # the alleles REF,ALT must match, in the second example they are ignored. The option -k is required
    # to output also records that are not annotated. The third example shows the same concept with
    # a numerical expression.
    bcftools annotate -a annots.tsv.gz -c CHROM,POS,REF,ALT,SCORE,~STR -i'TAG={STR}' -k input.vcf
    bcftools annotate -a annots.tsv.gz -c CHROM,POS,-,-,SCORE,~STR     -i'TAG={STR}' -k input.vcf
    bcftools annotate -a annots.tsv.gz -c CHROM,POS,-,-,SCORE,~INT     -i'TAG&gt;{INT}' -k input.vcf</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-k, --keep-sites</strong></dt>
<dd>
<p>keep sites which do not pass <strong>-i</strong> and <strong>-e</strong> expressions instead of discarding them</p>
</dd>
<dt class="hdlist1"><strong>-l, --merge-logic</strong> <em>tag:first</em>|<em>append</em>|<em>append-missing</em>|<em>unique</em>|<em>sum</em>|<em>avg</em>|<em>min</em>|<em>max</em>[,&#8230;&#8203;]</dt>
<dd>
<p>When multiple regions overlap a single record, this option defines how to treat multiple
annotation values when setting <em>tag</em> in the destination file: use the first encountered value ignoring
the rest (<em>first</em>); append allowing duplicates (<em>append</em>); append even if the appended value is missing,
i.e. is a dot (<em>append-missing</em>); append discarding duplicate values (<em>unique</em>);
sum the values (<em>sum</em>, numeric fields only); average the values (<em>avg</em>); use the minimum value (<em>min</em>) or
the maximum (<em>max</em>).
+
Note that this option is intended for use with BED or TAB-delimited annotation files only. Moreover,
it is effective only when either <em>REF</em> and <em>ALT</em> or <em>BEG</em> and <em>END</em> <strong>--columns</strong> are present .
+
Multiple rules can be given either as a comma-separated list or giving the option multiple times.
This is an experimental feature.</p>
</dd>
<dt class="hdlist1"><strong>-m, --mark-sites</strong> <em class="+-">TAG</em></dt>
<dd>
<p>annotate sites which are present ("+") or absent ("-") in the <strong>-a</strong> file with a new INFO/TAG flag</p>
</dd>
<dt class="hdlist1"><strong>--min-overlap</strong> <em>ANN</em>:'VCF'</dt>
<dd>
<p>minimum overlap required as a fraction of the variant in the annotation <strong>-a</strong> file (<em>ANN</em>), in the
target VCF file (<em>:VCF</em>), or both for reciprocal overlap (<em>ANN:VCF</em>).
By default overlaps of arbitrary length are sufficient.
The option can be used only with the tab-delimited annotation <strong>-a</strong> file and with <em>BEG</em> and <em>END</em>
columns present.</p>
</dd>
<dt class="hdlist1"><strong>--no-version</strong></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-O, --output-type</strong> <em>b</em>|<em>u</em>|<em>z</em>|<em>v</em>[0-9]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--pair-logic</strong> <em>snps</em>|<em>indels</em>|<em>both</em>|<em>all</em>|<em>some</em>|<em>exact</em></dt>
<dd>
<p>Controls how to match records from the annotation file to the target VCF.
Effective only when <strong>-a</strong> is a VCF or BCF. The option replaces the former
uninuitive <strong>--collapse</strong>.
See <strong><a href="#common_options">Common Options</a></strong> for more.</p>
</dd>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--rename-annots</strong> <em>file</em></dt>
<dd>
<p>rename annotations according to the map in <em>file</em>, with
"old_name new_name\n" pairs separated by whitespaces, each on a separate
line. The old name must be prefixed with the annotation type:
INFO, FORMAT, or FILTER.</p>
</dd>
<dt class="hdlist1"><strong>--rename-chrs</strong> <em>file</em></dt>
<dd>
<p>rename chromosomes according to the map in <em>file</em>, with
"old_name new_name\n" pairs separated by whitespaces, each on a separate
line.</p>
</dd>
<dt class="hdlist1"><strong>-s, --samples</strong> [^]<em>LIST</em></dt>
<dd>
<p>subset of samples to annotate, see also <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-S, --samples-file</strong> <em>FILE</em></dt>
<dd>
<p>subset of samples to annotate. If the samples are named differently in the
target VCF and the <strong>-a, --annotations</strong> VCF, the name mapping can be
given as "src_name dst_name\n", separated by whitespaces, each pair on a
separate line.</p>
</dd>
<dt class="hdlist1"><strong>--single-overlaps</strong></dt>
<dd>
<p>use this option to keep memory requirements low with very large annotation
files. Note, however, that this comes at a cost, only single overlapping intervals
are considered in this mode. This was the default mode until the commit
af6f0c9 (Feb 24 2019).</p>
</dd>
<dt class="hdlist1"><strong>--threads</strong> <em>INT</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-x, --remove</strong> <em>list</em></dt>
<dd>
<p>List of annotations to remove. Use "FILTER" to remove all filters or
"FILTER/SomeFilter" to remove a specific filter. Similarly, "INFO" can
be used to remove all INFO tags and "FORMAT" to remove all FORMAT tags
except GT. To remove all INFO tags except "FOO" and "BAR", use
"^INFO/FOO,INFO/BAR" (and similarly for FORMAT and FILTER).
"INFO" can be abbreviated to "INF" and "FORMAT" to "FMT".</p>
</dd>
<dt class="hdlist1"><strong>-W</strong>[<em>FMT</em>]<strong>, -W</strong>[=<em>FMT</em>]<strong>, --write-index</strong>[=<em>FMT</em>]</dt>
<dd>
<p>Automatically index the output file.  <em>FMT</em> is optional and can be
one of "tbi" or "csi" depending on output file format.</p>
</dd>
</dl>
</div>
<div class="paragraph">
<p><strong>Examples:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre>    # Remove three fields
    bcftools annotate -x ID,INFO/DP,FORMAT/DP file.vcf.gz

    # Remove all INFO fields and all FORMAT fields except for GT and PL
    bcftools annotate -x INFO,^FORMAT/GT,FORMAT/PL file.vcf

    # Add ID, QUAL and INFO/TAG, not replacing TAG if already present
    bcftools annotate -a src.bcf -c ID,QUAL,+TAG dst.bcf

    # Carry over all INFO and FORMAT annotations except FORMAT/GT
    bcftools annotate -a src.bcf -c INFO,^FORMAT/GT dst.bcf

    # Annotate from a tab-delimited file with six columns (the fifth is ignored),
    # first indexing with tabix. The coordinates are 1-based.
    tabix -s1 -b2 -e2 annots.tab.gz
    bcftools annotate -a annots.tab.gz -h annots.hdr -c CHROM,POS,REF,ALT,-,TAG file.vcf

    # Annotate from a tab-delimited file with regions (1-based coordinates, inclusive)
    tabix -s1 -b2 -e3 annots.tab.gz
    bcftools annotate -a annots.tab.gz -h annots.hdr -c CHROM,FROM,TO,TAG input.vcf

    # Annotate from a bed file (0-based coordinates, half-closed, half-open intervals)
    bcftools annotate -a annots.bed.gz -h annots.hdr -c CHROM,FROM,TO,TAG input.vcf

    # Transfer the INFO/END tag, matching by POS,REF,ALT and ID. This example assumes
    # that INFO/END is already present in the VCF header.
    bcftools annotate -a annots.tab.gz  -c CHROM,POS,~ID,REF,ALT,INFO/END input.vcf

    # For (many) more examples see http://samtools.github.io/bcftools/howtos/annotate.html</pre>
</div>
</div>
</div>
<div class="sect2">
<h3 id="call">bcftools call <em>[OPTIONS]</em> <em>FILE</em></h3>
<div class="paragraph">
<p>This command replaces the former <strong>bcftools view</strong> caller. Some of the original
functionality has been temporarily lost in the process of transition under
<a href="http://github.com/samtools/htslib">htslib</a>, but will be added back on popular
demand. The original calling model can be invoked with the <strong>-c</strong> option.</p>
</div>
<div class="sect3">
<h4 id="_file_format_options">File format options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--no-version</strong></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-O, --output-type</strong> <em>b</em>|<em>u</em>|<em>z</em>|<em>v</em>[0-9]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--ploidy</strong> <em>ASSEMBLY</em>[<em>?</em>]<a id="ploidy"></a></dt>
<dd>
<p>predefined ploidy, use <em>list</em> (or any other unused word) to print a list
of all predefined assemblies. Append a question mark to print the actual
definition. See also <strong>--ploidy-file</strong>.</p>
</dd>
<dt class="hdlist1"><strong>--ploidy-file</strong> <em>FILE</em></dt>
<dd>
<p>ploidy definition given as a space/tab-delimited list of
CHROM, FROM, TO, SEX, PLOIDY. The SEX codes are arbitrary and
correspond to the ones used by <strong><a href="#samples_file">--samples-file</a></strong>.
The default ploidy can be given using the starred records (see
below), unlisted regions have ploidy 2. The default ploidy definition is</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    X 1 60000 M 1
    X 2699521 154931043 M 1
    Y 1 59373566 M 1
    Y 1 59373566 F 0
    MT 1 16569 M 1
    MT 1 16569 F 1
    *  * *     M 2
    *  * *     F 2</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-s, --samples</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-S, --samples-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-t, --targets</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-T, --targets-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--targets-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--threads</strong> <em>INT</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-W</strong>[<em>FMT</em>]<strong>, -W</strong>[=<em>FMT</em>]<strong>, --write-index</strong>[=<em>FMT</em>]</dt>
<dd>
<p>Automatically index the output file.  <em>FMT</em> is optional and can be
one of "tbi" or "csi" depending on output file format.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_inputoutput_options">Input/output options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-A, --keep-alts</strong></dt>
<dd>
<p>output all alternate alleles present in the alignments even if they do not
appear in any of the genotypes</p>
</dd>
<dt class="hdlist1"><strong>-</strong>*<strong>, --keep-unseen-allele</strong></dt>
<dd>
<p>keep the unobserved allele &lt;*&gt; or &lt;NON_REF&gt;, useful mainly for gVCF output</p>
</dd>
<dt class="hdlist1"><strong>-f, --format-fields</strong> <em>list</em></dt>
<dd>
<p>comma-separated list of FORMAT fields to output for each sample. Currently
GQ and GP fields are supported. For convenience, the fields can be given
as lower case letters. Prefixed with "^" indicates a request for tag
removal of auxiliary tags useful only for calling.</p>
</dd>
<dt class="hdlist1"><strong>-F, --prior-freqs</strong> <em>AN</em>,<em>AC</em></dt>
<dd>
<p>take advantage of prior knowledge of population allele frequencies. The
workflow looks like this:</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    # Extract AN,AC values from an existing VCF, such 1000Genomes
    bcftools query -f'%CHROM\t%POS\t%REF\t%ALT\t%AN\t%AC\n' 1000Genomes.bcf | bgzip -c &gt; AFs.tab.gz

    # If the tags AN,AC are not already present, use the +fill-tags plugin
    bcftools +fill-tags 1000Genomes.bcf | bcftools query -f'%CHROM\t%POS\t%REF\t%ALT\t%AN\t%AC\n' | bgzip -c &gt; AFs.tab.gz
    tabix -s1 -b2 -e2 AFs.tab.gz

    # Create a VCF header description, here we name the tags REF_AN,REF_AC
    cat AFs.hdr
    ##INFO=&lt;ID=REF_AN,Number=1,Type=Integer,Description="Total number of alleles in reference genotypes"&gt;
    ##INFO=&lt;ID=REF_AC,Number=A,Type=Integer,Description="Allele count in reference genotypes for each ALT allele"&gt;

    # Now before calling, stream the raw mpileup output through `bcftools annotate` to add the frequencies
    bcftools mpileup [...] -Ou | bcftools annotate -a AFs.tab.gz -h AFs.hdr -c CHROM,POS,REF,ALT,REF_AN,REF_AC -Ou | bcftools call -mv -F REF_AN,REF_AC [...]</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-G, --group-samples</strong> <em class="TAG:">FILE</em>|<em>-</em></dt>
<dd>
<p>by default, all samples are assumed to come from a single population. This option groups samples
into populations and apply the HWE assumption within but not across the populations. <em>FILE</em> is a tab-delimited
text file with sample names in the first column and group names in the second column. If <em>-</em> is
given instead, no HWE assumption is made at all and single-sample calling is performed. (Note that
in low coverage data this inflates the rate of false positives.) The <strong>-G</strong> option requires the presence of
per-sample FORMAT/QS or FORMAT/AD tag generated with <strong>bcftools mpileup -a QS</strong> (or <strong>-a AD</strong>).</p>
</dd>
<dt class="hdlist1"><strong>-g, --gvcf</strong> <em>INT</em>[,&#8230;&#8203;]</dt>
<dd>
<p>output gVCF blocks of homozygous REF calls, with depth (DP) ranges
specified by the list of integers. For example, passing <em>5,15</em> will
group sites into two types of gVCF blocks, the first with minimum
per-sample DP from the interval [5,15) and the latter with minimum
depth 15 or more. In this example, sites with minimum per-sample
depth less than 5 will be printed as separate records, outside of
gVCF blocks.</p>
</dd>
<dt class="hdlist1"><strong>-i, --insert-missed</strong> <em>INT</em></dt>
<dd>
<p>output also sites missed by mpileup but present in <strong>-T, --targets-file</strong>.</p>
</dd>
<dt class="hdlist1"><strong>-M, --keep-masked-ref</strong></dt>
<dd>
<p>output sites where REF allele is N</p>
</dd>
<dt class="hdlist1"><strong>-V, --skip-variants</strong> <em>snps</em>|<em>indels</em></dt>
<dd>
<p>skip indel/SNP sites</p>
</dd>
<dt class="hdlist1"><strong>-v, --variants-only</strong></dt>
<dd>
<p>output variant sites only</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_consensusvariant_calling_options">Consensus/variant calling options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-c, --consensus-caller</strong></dt>
<dd>
<p>the original <strong>samtools</strong>/<strong>bcftools</strong> calling method (conflicts with <strong>-m</strong>)</p>
</dd>
<dt class="hdlist1"><strong>-C, --constrain</strong> <em>alleles</em>|<em>trio</em></dt>
<dd>
<div class="dlist">
<dl>
<dt class="hdlist1"><em>alleles</em></dt>
<dd>
<p>call genotypes given alleles. See also <strong>-T, --targets-file</strong>.</p>
</dd>
<dt class="hdlist1"><em>trio</em></dt>
<dd>
<p>call genotypes given the father-mother-child constraint. See also
<strong>-s, --samples</strong> and <strong>-n, --novel-rate</strong>.</p>
</dd>
</dl>
</div>
</dd>
<dt class="hdlist1"><strong>-m, --multiallelic-caller</strong></dt>
<dd>
<p>alternative model for multiallelic and rare-variant calling designed to
overcome known limitations in <strong>-c</strong> calling model (conflicts with <strong>-c</strong>)</p>
</dd>
<dt class="hdlist1"><strong>-n, --novel-rate</strong> <em>float</em>[,&#8230;&#8203;]</dt>
<dd>
<p>likelihood of novel mutation for constrained <strong>-C</strong> <em>trio</em> calling. The trio
genotype calling maximizes likelihood of a particular combination of
genotypes for father, mother and the child
P(F=i,M=j,C=k) = P(unconstrained) * Pn + P(constrained) * (1-Pn).
By providing three values, the mutation rate Pn is set explicitly for SNPs,
deletions and insertions, respectively.  If two values are given, the first
is interpreted as the mutation rate of SNPs and the second is used to
calculate the mutation rate of indels according to their length as
Pn=<em>float</em>*exp(-a-b*len), where a=22.8689, b=0.2994 for insertions and
a=21.9313, b=0.2856 for deletions [pubmed:23975140].  If only one value is
given, the same mutation rate Pn is used for SNPs and indels.</p>
</dd>
<dt class="hdlist1"><strong>-p, --pval-threshold</strong> <em>float</em></dt>
<dd>
<p>with <strong>-c</strong>, accept variant if P(ref|D) &lt; <em>float</em>.</p>
</dd>
<dt class="hdlist1"><strong>-P, --prior</strong> <em>float</em></dt>
<dd>
<p>expected substitution rate, or 0 to disable the prior. Only with <strong>-m</strong>.</p>
</dd>
<dt class="hdlist1"><strong>-t, --targets</strong> <em>file</em>|<em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-X, --chromosome-X</strong></dt>
<dd>
<p>haploid output for male samples (requires PED file with <strong>-s</strong>)</p>
</dd>
<dt class="hdlist1"><strong>-Y, --chromosome-Y</strong></dt>
<dd>
<p>haploid output for males and skips females (requires PED file with <strong>-s</strong>)</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="cnv">bcftools cnv <em>[OPTIONS]</em> <em>FILE</em></h3>
<div class="paragraph">
<p>Copy number variation caller, requires a VCF annotated with the Illumina&#8217;s
B-allele frequency (BAF) and Log R Ratio intensity (LRR) values. The HMM
considers the following copy number states: CN 2 (normal), 1 (single-copy
loss), 0 (complete loss), 3 (single-copy gain).</p>
</div>
<div class="sect3">
<h4 id="_general_options">General Options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-c, --control-sample</strong> <em>string</em></dt>
<dd>
<p>optional control sample name. If given, pairwise calling is performed
and the <strong>-P</strong>  option can be used</p>
</dd>
<dt class="hdlist1"><strong>-f, --AF-file</strong> <em>file</em></dt>
<dd>
<p>read allele frequencies from  a tab-delimited file with the columns CHR,POS,REF,ALT,AF</p>
</dd>
<dt class="hdlist1"><strong>-o, --output-dir</strong> <em>path</em></dt>
<dd>
<p>output directory</p>
</dd>
<dt class="hdlist1"><strong>-p, --plot-threshold</strong> <em>float</em></dt>
<dd>
<p>call <strong>matplotlib</strong> to produce plots for chromosomes with quality at least <em>float</em>,
useful for visual inspection of the calls. With <strong>-p 0</strong>, plots for all chromosomes will be
generated. If not given, a <strong>matplotlib</strong> script will be created but not called.</p>
</dd>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-s, --query-sample</strong> <em>string</em></dt>
<dd>
<p>query sample name</p>
</dd>
<dt class="hdlist1"><strong>-t, --targets</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-T, --targets-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--targets-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_hmm_options">HMM Options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-a, --aberrant</strong> <em>float</em>[,<em>float</em>]</dt>
<dd>
<p>fraction of aberrant cells in query and control. The hallmark of
duplications and contaminations is the BAF value of heterozygous markers
which is dependent on the fraction of aberrant cells. Sensitivity to
smaller fractions of cells can be increased by setting <strong>-a</strong> to a lower value. Note
however, that this comes at the cost of increased false discovery rate.</p>
</dd>
<dt class="hdlist1"><strong>-b, --BAF-weight</strong> <em>float</em></dt>
<dd>
<p>relative contribution from BAF</p>
</dd>
<dt class="hdlist1"><strong>-d, --BAF-dev</strong> <em>float</em>[,<em>float</em>]</dt>
<dd>
<p>expected BAF deviation in query and control, i.e. the noise observed
in the data.</p>
</dd>
<dt class="hdlist1"><strong>-e, --err-prob</strong> <em>float</em></dt>
<dd>
<p>uniform error probability</p>
</dd>
<dt class="hdlist1"><strong>-l, --LRR-weight</strong> <em>float</em></dt>
<dd>
<p>relative contribution from LRR. With noisy data, this option can have big effect
on the number of calls produced. In truly random noise (such as in simulated data),
the value should be set high (1.0), but in the presence of systematic noise
when LRR are not informative, lower values result in cleaner calls (0.2).</p>
</dd>
<dt class="hdlist1"><strong>-L, --LRR-smooth-win</strong> <em>int</em></dt>
<dd>
<p>reduce LRR noise by applying moving average given this window size</p>
</dd>
<dt class="hdlist1"><strong>-O, --optimize</strong> <em>float</em></dt>
<dd>
<p>iteratively estimate the fraction of aberrant cells, down to the given fraction.
Lowering this value from the default 1.0 to say, 0.3, can help discover more
events but also increases noise</p>
</dd>
<dt class="hdlist1"><strong>-P, --same-prob</strong> <em>float</em></dt>
<dd>
<p>the prior probability of the query and the control sample being the same.
Setting to 0 calls both independently, setting to 1 forces the same copy
number state in both.</p>
</dd>
<dt class="hdlist1"><strong>-x, --xy-prob</strong> <em>float</em></dt>
<dd>
<p>the HMM probability of transition to another copy number state. Increasing this
values leads to smaller and more frequent calls.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="concat">bcftools concat <em>[OPTIONS]</em> <em>FILE1</em> <em>FILE2</em> [&#8230;&#8203;]</h3>
<div class="paragraph">
<p>Concatenate or combine VCF/BCF files. All source files must have the same sample
columns appearing in the same order. Can be used, for example, to
concatenate chromosome VCFs into one VCF, or combine a SNP VCF and an indel
VCF into one. The input files must be sorted by chr and position. The files
must be given in the correct order to produce sorted VCF on output unless
the <strong>-a, --allow-overlaps</strong> option is specified. With the --naive option, the files
are concatenated without being recompressed, which is very fast..</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-a, --allow-overlaps</strong></dt>
<dd>
<p>First coordinate of the next file can precede last record of the current file.</p>
</dd>
<dt class="hdlist1"><strong>-c, --compact-PS</strong></dt>
<dd>
<p>Do not output PS tag at each site, only at the start of a new phase set block.</p>
</dd>
<dt class="hdlist1"><strong>-d, --rm-dups</strong> <em>snps</em>|<em>indels</em>|<em>both</em>|<em>all</em>|<em>exact</em></dt>
<dd>
<p>Output duplicate records of specified type present in multiple files only once.
Note that records duplicate within one file are not removed with this option,
for that use <strong><a href="#norm">bcftools norm -d</a></strong> instead.<br>
In other words, the default behavior of the program is similar to unix "<code>cat</code>" in
that when two files contain a record with the same position, that position will appear
twice on output. With <strong>-d</strong>, every line that finds a matching record in another
file will be printed only once.<br>
Requires <strong>-a, --allow-overlaps</strong>.</p>
</dd>
<dt class="hdlist1"><strong>-D, --remove-duplicates</strong></dt>
<dd>
<p>Alias for <strong>-d exact</strong></p>
</dd>
<dt class="hdlist1"><strong>-f, --file-list</strong> <em>FILE</em></dt>
<dd>
<p>Read file names from <em>FILE</em>, one file name per line.</p>
</dd>
<dt class="hdlist1"><strong>-l, --ligate</strong></dt>
<dd>
<p>Ligate phased VCFs by matching phase at overlapping haplotypes.
Note that the option is intended for VCFs with perfect overlap, sites
in overlapping regions present in one but missing in the other are dropped.</p>
</dd>
<dt class="hdlist1"><strong>--ligate-force</strong></dt>
<dd>
<p>Keep all sites and ligate even non-overlapping chunks and chunks with imperfect overlap</p>
</dd>
<dt class="hdlist1"><strong>--ligate-warn</strong></dt>
<dd>
<p>Drop sites in imperfect overlaps</p>
</dd>
<dt class="hdlist1"><strong>--no-version</strong></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-n, --naive</strong></dt>
<dd>
<p>Concatenate VCF or BCF files without recompression. This is very fast but requires
that all files are of the same type (all VCF or all BCF) and have the same headers.
This is because all tags and chromosome names in the BCF body rely on the order
of the contig and tag definitions in the header. A header check compatibility
is performed and the program throws an error if it is not safe to use the option.</p>
</dd>
<dt class="hdlist1"><strong>--naive-force</strong></dt>
<dd>
<p>Same as --naive, but header compatibility is not checked. Dangerous, use with caution.</p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-O, --output-type</strong> <em>b</em>|<em>u</em>|<em>z</em>|<em>v</em>[0-9]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-q, --min-PQ</strong> <em>INT</em></dt>
<dd>
<p>Break phase set if phasing quality is lower than <em>INT</em></p>
</dd>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong>. Requires <strong>-a, --allow-overlaps</strong>.</p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong>. Requires <strong>-a, --allow-overlaps</strong>.</p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--threads</strong> <em>INT</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-W</strong>[<em>FMT</em>]<strong>, -W</strong>[=<em>FMT</em>]<strong>, --write-index</strong>[=<em>FMT</em>]</dt>
<dd>
<p>Automatically index the output file.  <em>FMT</em> is optional and can be
one of "tbi" or "csi" depending on output file format.</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="consensus">bcftools consensus <em>[OPTIONS]</em> <em>FILE</em></h3>
<div class="paragraph">
<p>Create consensus sequence by applying VCF variants to a reference fasta file.
By default, the program will apply all ALT variants to the reference fasta to
obtain the consensus sequence. Using the <strong>--sample</strong> (and, optionally,
<strong>--haplotype</strong>) option will apply genotype (haplotype) calls from FORMAT/GT.
Note that the program does not act as a primitive variant caller and ignores allelic
depth information, such as INFO/AD or FORMAT/AD. For that, consider using the
<strong>setGT</strong> plugin.</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-a, --absent</strong> <em>CHAR</em></dt>
<dd>
<p>replace positions absent from VCF with CHAR</p>
</dd>
<dt class="hdlist1"><strong>-c, --chain</strong> <em>FILE</em></dt>
<dd>
<p>write a chain file for liftover</p>
</dd>
<dt class="hdlist1"><strong>-e, --exclude</strong> <em>EXPRESSION</em></dt>
<dd>
<p>exclude sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-f, --fasta-ref</strong> <em>FILE</em></dt>
<dd>
<p>reference sequence in fasta format</p>
</dd>
<dt class="hdlist1"><strong>-H, --haplotype</strong> N|<em>R</em>|<em>A</em>|<em>I</em>|<em>LR</em>|<em>LA</em>|<em>SR</em>|<em>SA</em>|<em>NpIu</em></dt>
<dd>
<p>choose which allele from the FORMAT/GT field to use (the codes are case-insensitive):</p>
<div class="dlist">
<dl>
<dt class="hdlist1"><em>N</em></dt>
<dd>
<p>N={1,2,3,&#8230;&#8203;}, the allele index within the genotype, regardless of phasing</p>
</dd>
<dt class="hdlist1"><em>R</em></dt>
<dd>
<p>the REF allele (in heterozygous genotypes)</p>
</dd>
<dt class="hdlist1"><em>A</em></dt>
<dd>
<p>the ALT allele (in heterozygous genotypes)</p>
</dd>
<dt class="hdlist1"><em>I</em></dt>
<dd>
<p>IUPAC code for all genotypes</p>
</dd>
<dt class="hdlist1"><em>LR, LA</em></dt>
<dd>
<p>the longer allele. If both have the same length, use the REF allele (LR), or the ALT allele  (LA)</p>
</dd>
<dt class="hdlist1"><em>SR, SA</em></dt>
<dd>
<p>the shorter allele. If both have the same length, use the REF allele (SR), or the ALT allele  (SA)</p>
</dd>
<dt class="hdlist1"><em>NpIu</em></dt>
<dd>
<p>N={1,2,3,&#8230;&#8203;}, the allele index within genotype for phased genotypes and IUPAC code for unphased genotypes.
For example, <em>1pIu</em> or <em>2pIu</em></p>
</dd>
</dl>
</div>
</dd>
</dl>
</div>
<div class="paragraph">
<p>Note that the <strong>-H, --haplotype</strong> option requires the <strong>-s, --samples</strong> option, unless exactly one sample is present in the VCF</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --include</strong> <em>EXPRESSION</em></dt>
<dd>
<p>include only sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-I, --iupac-codes</strong></dt>
<dd>
<p>output variants in the form of IUPAC ambiguity codes determined from FORMAT/GT fields. By default all
samples are used and can be subset with <strong>-s, --samples</strong> and <strong>-S, --samples-file</strong>. Use <strong>-s -</strong> to ignore
samples and use only the REF and ALT columns.  NOTE: prior to version 1.17 the IUPAC codes were determined solely
from REF,ALT columns and sample genotypes were not considered.</p>
</dd>
<dt class="hdlist1"><strong>--mark-del</strong> <em>CHAR</em></dt>
<dd>
<p>instead of removing sequence, insert character CHAR for deletions</p>
</dd>
<dt class="hdlist1"><strong>--mark-ins</strong> <em>uc</em>|<em>lc</em>|<em>CHAR</em></dt>
<dd>
<p>highlight inserted sequence in uppercase (uc), lowercase (lc), or a provided character CHAR, leaving the rest of the sequence as is</p>
</dd>
<dt class="hdlist1"><strong>--mark-snv</strong> <em>uc</em>|<em>lc</em></dt>
<dd>
<p>highlight substitutions in uppercase (uc), lowercase (lc), or a provided character CHAR, leaving the rest of the sequence as is</p>
</dd>
<dt class="hdlist1"><strong>-m, --mask</strong> <em>FILE</em></dt>
<dd>
<p>BED file or TAB file with regions to be replaced with N (the default) or as specified by
the next <strong>--mask-with</strong> option. See discussion
of <strong>--regions-file</strong> in <strong><a href="#common_options">Common Options</a></strong> for file
format details.</p>
</dd>
<dt class="hdlist1"><strong>--mask-with</strong> <em>CHAR</em>|<em>lc</em>|<em>uc</em></dt>
<dd>
<p>replace sequence from <strong>--mask</strong> with CHAR, skipping overlapping variants, or change to lowercase (lc) or uppercase (uc)</p>
</dd>
<dt class="hdlist1"><strong>-M, --missing</strong> <em>CHAR</em></dt>
<dd>
<p>instead of skipping the missing genotypes, output the character CHAR (e.g. "?")</p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>write output to a file</p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>how to treat VCF variants overlapping the target region in the fasta file:
see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-s, --samples</strong> <em>LIST</em></dt>
<dd>
<p>apply variants of the listed samples. See also the option <strong>-I, --iupac-codes</strong></p>
</dd>
<dt class="hdlist1"><strong>-S, --samples-file</strong> <em>FILE</em></dt>
<dd>
<p>apply variants of the samples listed in the file. See also the option <strong>-I, --iupac-codes</strong></p>
</dd>
</dl>
</div>
<div class="paragraph">
<p><strong>Examples:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre>    # Apply variants present in sample "NA001", output IUPAC codes for hets
    bcftools consensus -i -s NA001 -f in.fa in.vcf.gz &gt; out.fa

    # Create consensus for one region. The fasta header lines are then expected
    # in the form "&gt;chr:from-to". Ignore samples and consider only the REF and ALT columns
    samtools faidx ref.fa 8:11870-11890 | bcftools consensus -s - in.vcf.gz -o out.fa

    # For more examples see http://samtools.github.io/bcftools/howtos/consensus-sequence.html</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>Notes:</strong></dt>
<dd>
<p>Masking options are applied in the following order</p>
<div class="olist arabic">
<ol class="arabic">
<li>
<p>mask regions with <strong>--mask-with</strong> character if <strong>--mask</strong> is given. All overlapping VCF variants are ignored</p>
</li>
<li>
<p>replace sequence not mentioned in the VCF with the requested character if <strong>--absent</strong> is given</p>
</li>
<li>
<p>finally apply <strong>--mark-del</strong>, <strong>--mark-ins</strong>, <strong>--mark-snv</strong> masks</p>
</li>
</ol>
</div>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="convert">bcftools convert <em>[OPTIONS]</em> <em>FILE</em></h3>
<div class="sect3">
<h4 id="_vcf_input_options">VCF input options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-e, --exclude</strong> <em>EXPRESSION</em></dt>
<dd>
<p>exclude sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-i, --include</strong> <em>EXPRESSION</em></dt>
<dd>
<p>include only sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-s, --samples</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-S, --samples-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-t, --targets</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-T, --targets-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--targets-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-W</strong>[<em>FMT</em>]<strong>, -W</strong>[=<em>FMT</em>]<strong>, --write-index</strong>[=<em>FMT</em>]</dt>
<dd>
<p>Automatically index the output file.  <em>FMT</em> is optional and can be
one of "tbi" or "csi" depending on output file format.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_vcf_output_options">VCF output options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--no-version</strong></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-O, --output-type</strong> <em>b</em>|<em>u</em>|<em>z</em>|<em>v</em>[0-9]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--threads</strong> <em>INT</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_gensample_conversion">GEN/SAMPLE conversion:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-G, --gensample2vcf</strong> <em>prefix</em> or <em>gen-file</em>,<em>sample-file</em></dt>
<dd>
<p>convert IMPUTE2 output to VCF. One of the ID columns ("SNP ID" or "rsID" in
<a href="https://www.cog-genomics.org/plink/2.0/formats#gen" class="bare">https://www.cog-genomics.org/plink/2.0/formats#gen</a>) must be of the form
"CHROM:POS_REF_ALT" to detect possible strand swaps.
&#160;<br>
When the <strong>--vcf-ids</strong> option is given, the other column (autodetected) is used
to fill the ID column of the VCF.
&#160;<br>
See also <strong>-g</strong> and <strong>--3N6</strong> options.</p>
</dd>
<dt class="hdlist1"><strong>-g, --gensample</strong> <em>prefix</em> or <em>gen-file</em>,<em>sample-file</em></dt>
<dd>
<p>convert from VCF to gen/sample format used by IMPUTE2 and SHAPEIT.
The columns of .gen file format are ID1,ID2,POS,A,B followed by three
genotype probabilities P(AA), P(AB), P(BB) for each sample.  In order to
prevent strand swaps, the program uses IDs of the form "CHROM:POS_REF_ALT".
When the <strong>--vcf-ids</strong> option is given, the second column is set to match the ID
column of the VCF.
&#160;<br>
See also <strong>-G</strong> and <strong>--3N6</strong> options.
&#160;<br>
The file .gen and .sample file format are:</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>  .gen (with --3N6 --vcf-ids)
  ---------------------------
  chr1 1:111485207_G_A rsID1 111485207 G A 0 1 0 0 1 0
  chr1 1:111494194_C_T rsID2 111494194 C T 0 1 0 0 0 1

  .gen (with --vcf-ids)
  ---------------------------
  1:111485207_G_A rsID1 111485207 G A 0 1 0 0 1 0
  1:111494194_C_T rsID2 111494194 C T 0 1 0 0 0 1

  .gen (the default)
  ------------------------------
  1:111485207_G_A 1:111485207_G_A 111485207 G A 0 1 0 0 1 0
  1:111494194_C_T 1:111494194_C_T 111494194 C T 0 1 0 0 0 1

  .sample
  -------
  ID_1 ID_2 missing
  0 0 0
  sample1 sample1 0
  sample2 sample2 0</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--3N6</strong></dt>
<dd>
<p>Expect/Create files in the 3*N+6 column format. This is the new .gen file format with the first
column containing the chromosome name, see <a href="https://www.cog-genomics.org/plink/2.0/formats#gen" class="bare">https://www.cog-genomics.org/plink/2.0/formats#gen</a></p>
</dd>
<dt class="hdlist1"><strong>--tag</strong> <em>STRING</em></dt>
<dd>
<p>tag to take values for .gen file: GT,PL,GL,GP</p>
</dd>
<dt class="hdlist1"><strong>--sex</strong> <em>FILE</em></dt>
<dd>
<p>output sex column in the sample file. The FILE format is</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    MaleSample    M
    FemaleSample  F</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--vcf-ids</strong></dt>
<dd>
<p>output VCF IDs in the second column instead of CHROM:POS_REF_ALT</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_gvcf_conversion">gVCF conversion:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--gvcf2vcf</strong></dt>
<dd>
<p>convert gVCF to VCF, expanding REF blocks into sites. Note that
the <strong>-i</strong> and <strong>-e</strong> options work differently with this switch. In this situation
the filtering expressions define which sites should be expanded and
which sites should be left unmodified, but all sites are printed on
output. In order to drop sites, stream first through <strong><a href="#view">bcftools view</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-f, --fasta-ref</strong> <em>file</em></dt>
<dd>
<p>reference sequence in fasta format. Must be indexed with samtools faidx</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_hapsample_conversion">HAP/SAMPLE conversion:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--hapsample2vcf</strong> <em>prefix</em> or <em>hap-file</em>,<em>sample-file</em></dt>
<dd>
<p>convert from hap/sample format to VCF. The columns of .hap file are
similar to .gen file above, but there are only two haplotype columns per
sample.  Note that the first or the second column of the .hap file is expected to be in
the form "CHR:POS_REF_ALT[_END]", with the _END being optional for
defining the INFO/END tag when ALT is a symbolic allele. For example:</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>  .hap (with --vcf-ids)
  ---------------------
  1:111485207_G_A rsID1 111485207 G A 0 1 0 0
  1:111495231_A_&lt;DEL&gt;_111495784 rsID3 111495231 A &lt;DEL&gt; 0 0 1 0

  .hap (the default)
  ------------------
  1 1:111485207_G_A 111485207 G A 0 1 0 0
  1 1:111495231_A_&lt;DEL&gt;_111495784 111495231 A &lt;DEL&gt; 0 0 1 0</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--hapsample</strong> <em>prefix</em> or <em>hap-file</em>,<em>sample-file</em></dt>
<dd>
<p>convert from VCF to hap/sample format used by IMPUTE2 and SHAPEIT.
The columns of .hap file begin with ID,RSID,POS,REF,ALT. In order to
prevent strand swaps, the program uses IDs of the form
"CHROM:POS_REF_ALT".</p>
</dd>
<dt class="hdlist1"><strong>--haploid2diploid</strong></dt>
<dd>
<p>with <strong>-h</strong> option converts haploid genotypes to homozygous diploid
genotypes. For example, the program will print <em>0 0</em> instead of the
default <em>0 -</em>. This is useful for programs which do not handle haploid
genotypes correctly.</p>
</dd>
<dt class="hdlist1"><strong>--sex</strong> <em>FILE</em></dt>
<dd>
<p>output sex column in the sample file. The FILE format is</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    MaleSample    M
    FemaleSample  F</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--vcf-ids</strong></dt>
<dd>
<p>the second column of the .hap file holds the VCF ids, the first
column is of the form "CHR:POS_REF_ALT[_END]". Without the option,
the format follows <a href="https://www.cog-genomics.org/plink/2.0/formats#haps" class="bare">https://www.cog-genomics.org/plink/2.0/formats#haps</a>
with ids (the second column) of the form "CHR:POS_REF_ALT[_END]"</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_haplegendsample_conversion">HAP/LEGEND/SAMPLE conversion:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-H, --haplegendsample2vcf</strong> <em>prefix</em> or <em>hap-file</em>,<em>legend-file</em>,<em>sample-file</em></dt>
<dd>
<p>convert from hap/legend/sample format used by IMPUTE2 to VCF.
See also <strong>-h, --hapslegendsample</strong> below.</p>
</dd>
<dt class="hdlist1"><strong>-h, --haplegendsample</strong> <em>prefix</em> or <em>hap-file</em>,<em>legend-file</em>,<em>sample-file</em></dt>
<dd>
<p>convert from VCF to hap/legend/sample format used by IMPUTE2 and SHAPEIT.
The columns of .legend file ID,POS,REF,ALT. In order to prevent strand
swaps, the program uses IDs of the form "CHROM:POS_REF_ALT".
The .sample file is quite basic at the moment with columns for population, group and
sex expected to be edited by the user. For example:</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>  .hap
  -----
  0 1 0 0 1 0
  0 1 0 0 0 1

  .legend
  -------
  id position a0 a1
  1:111485207_G_A 111485207 G A
  1:111494194_C_T 111494194 C T

  .sample
  -------
  sample population group sex
  sample1 sample1 sample1 2
  sample2 sample2 sample2 2</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--haploid2diploid</strong></dt>
<dd>
<p>with <strong>-h</strong> option converts haploid genotypes to homozygous diploid
genotypes. For example, the program will print <em>0 0</em> instead of the
default <em>0 -</em>. This is useful for programs which do not handle haploid
genotypes correctly.</p>
</dd>
<dt class="hdlist1"><strong>--sex</strong> <em>FILE</em></dt>
<dd>
<p>output sex column in the sample file. The FILE format is</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    MaleSample    M
    FemaleSample  F</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--vcf-ids</strong></dt>
<dd>
<p>output VCF IDs instead of "CHROM:POS_REF_ALT". Note that this option can
be used with <strong>--haplegendsample</strong> but not with <strong>--haplegendsample2vcf</strong>.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_tsv_conversion">TSV conversion:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--tsv2vcf</strong> <em>file</em></dt>
<dd>
<p>convert from TSV (tab-separated values) format (such as generated by
23andMe) to VCF. The input file fields can be tab- or space- delimited</p>
</dd>
<dt class="hdlist1"><strong>-c, --columns</strong> <em>list</em></dt>
<dd>
<p>comma-separated list of fields in the input file. In the current
version, the fields CHROM, POS, ID, and AA or REF, ALT are expected and
can appear in arbitrary order. Columns which should be ignored in the input
file can be indicated by "-".
The AA field lists alleles on the forward reference strand,
for example "CC" or "CT" for diploid genotypes or "C"
for haploid genotypes (sex chromosomes). Insertions and deletions
are supported only with REF and ALT but not with AA. Missing data can be indicated with "--" or ".".</p>
</dd>
<dt class="hdlist1"><strong>-f, --fasta-ref</strong> <em>file</em></dt>
<dd>
<p>reference sequence in fasta format. Must be indexed with samtools faidx</p>
</dd>
<dt class="hdlist1"><strong>-s, --samples</strong> <em>LIST</em></dt>
<dd>
<p>list of sample names. See <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-S, --samples-file</strong> <em>FILE</em></dt>
<dd>
<p>file of sample names. See <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
</dl>
</div>
<div class="paragraph">
<p><strong>Example:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre># Convert 23andme results into VCF
bcftools convert -c ID,CHROM,POS,AA -s SampleName -f 23andme-ref.fa --tsv2vcf 23andme.txt -o out.vcf.gz

# Convert tab-delimited file into a sites-only VCF (no genotypes), in this example first column to be ignored
bcftools convert -c -,CHROM,POS,REF,ALT -f ref.fa --tsv2vcf calls.txt -o out.bcf</pre>
</div>
</div>
</div>
</div>
<div class="sect2">
<h3 id="csq">bcftools csq <em>[OPTIONS]</em> <em>FILE</em></h3>
<div class="paragraph">
<p>Haplotype aware consequence predictor which correctly handles combined
variants such as MNPs split over multiple VCF records, SNPs separated by
an intron (but adjacent in the spliced transcript) or nearby frame-shifting
indels which in combination in fact are not frame-shifting.</p>
</div>
<div class="paragraph">
<p>The output VCF is annotated with INFO/BCSQ and FORMAT/BCSQ tag (configurable
with the <strong>-c</strong> option). The latter is a bitmask of indexes to INFO/BCSQ, with
interleaved haplotypes. See the usage examples below for using the %TBCSQ
converter in <strong>query</strong> for extracting a more human readable form from this
bitmask. The construction of the bitmask limits the number of consequences
that can be referenced per sample in the FORMAT/BCSQ tags. By default this is 15, but
if more are required, see the <strong>--ncsq</strong> option.</p>
</div>
<div class="paragraph">
<p>Note that the program annotates only records with a functional consequence and
intergenic regions will pass through unchanged.</p>
</div>
<div class="paragraph">
<p>The program requires on input a VCF/BCF file, the reference genome in fasta
format (<strong>--fasta-ref</strong>) and genomic features in the GFF3 format downloadable
from the Ensembl website (<strong>--gff-annot</strong>), and outputs an annotated VCF/BCF
file. Currently, only Ensembl GFF3 files are supported.</p>
</div>
<div class="paragraph">
<p>By default, the input VCF should be phased. If phase is unknown, or only
partially known, the <strong>--phase</strong> option can be used to indicate how to handle
unphased data. Alternatively, haplotype aware calling can be turned off
with the <strong>--local-csq</strong> option.</p>
</div>
<div class="paragraph">
<p>If conflicting (overlapping) variants within one haplotype are detected,
a warning will be emitted and predictions will be based on only the first
variant in the analysis.</p>
</div>
<div class="paragraph">
<p>Symbolic alleles are not supported. They will remain unannotated in the
output VCF and are ignored for the prediction analysis.</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-c, --custom-tag</strong> <em>STRING</em></dt>
<dd>
<p>use this custom tag to store consequences rather than the default BCSQ tag</p>
</dd>
<dt class="hdlist1"><strong>-B, --trim-protein-seq</strong> <em>INT</em></dt>
<dd>
<p>abbreviate protein-changing predictions to maximum of INT aminoacids.
For example, instead of writing the whole modified protein sequence with potentially hundreds of
aminoacids, with <strong>-B 1</strong> only an abbreviated version such as <em>25E..329&gt;25G..94</em> will be
written.</p>
</dd>
<dt class="hdlist1"><strong>--dump-gff</strong> <em>FILE</em></dt>
<dd>
<p>dump the parsed GFF into a gzipped FILE. Intended for debugging purposes,
shows how is the input GFF viewed by the program.</p>
</dd>
<dt class="hdlist1"><strong>-e, --exclude</strong> <em>EXPRESSION</em></dt>
<dd>
<p>exclude sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-f, --fasta-ref</strong> <em>FILE</em></dt>
<dd>
<p>reference sequence in fasta format (required)</p>
</dd>
<dt class="hdlist1"><strong>--force</strong></dt>
<dd>
<p>run even if some sanity checks fail. Currently the option enables skipping
transcripts in malformatted GFFs with incorrect phase</p>
</dd>
<dt class="hdlist1"><strong>-g, --gff-annot</strong> <em>FILE</em></dt>
<dd>
<p>GFF3 annotation file (required), such as <a href="http://ftp.ensembl.org/pub/current_gff3/homo_sapiens/" class="bare">http://ftp.ensembl.org/pub/current_gff3/homo_sapiens/</a>.
The script <strong><a href="#gff2gff">gff2gff</a></strong> can help with conversion from non-standard GFF formats.
An example of a minimal working GFF file:</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    # The program looks for "CDS", "exon", "three_prime_UTR" and "five_prime_UTR" lines,
    # looks up their parent transcript (determined from the "Parent=transcript:" attribute),
    # the gene (determined from the transcript's "Parent=gene:" attribute), and the biotype
    # (the most interesting is "protein_coding").
    #
    # Empty and commented lines are skipped, the following GFF columns are required
    #   1.  chromosome
    #   2.  IGNORED
    #   3.  type (CDS, exon, three_prime_UTR, five_prime_UTR, gene, transcript, etc.)
    #   4.  start of the feature (1-based)
    #   5.  end of the feature (1-based)
    #   6.  IGNORED
    #   7.  strand (+ or -)
    #   8.  phase (0, 1, 2 or .)
    #   9.  semicolon-separated attributes (see below)
    #
    # Attributes required for
    #   gene lines:
    #   - ID=gene:&lt;gene_id&gt;
    #   - biotype=&lt;biotype&gt;
    #   - Name=&lt;gene_name&gt;      [optional]
    #
    #   transcript lines:
    #   - ID=transcript:&lt;transcript_id&gt;
    #   - Parent=gene:&lt;gene_id&gt;
    #   - biotype=&lt;biotype&gt;
    #
    #   other lines (CDS, exon, five_prime_UTR, three_prime_UTR):
    #   - Parent=transcript:&lt;transcript_id&gt;
    #
    # Supported biotypes:
    #   - see the function gff_parse_biotype() in bcftools/csq.c

    1   ignored_field  gene            21  2148  .   -   .   ID=gene:GeneId;biotype=protein_coding;Name=GeneName
    1   ignored_field  transcript      21  2148  .   -   .   ID=transcript:TranscriptId;Parent=gene:GeneId;biotype=protein_coding
    1   ignored_field  three_prime_UTR 21  2054  .   -   .   Parent=transcript:TranscriptId
    1   ignored_field  exon            21  2148  .   -   .   Parent=transcript:TranscriptId
    1   ignored_field  CDS             21  2148  .   -   1   Parent=transcript:TranscriptId
    1   ignored_field  five_prime_UTR  210 2148  .   -   .   Parent=transcript:TranscriptId</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --include</strong> <em>EXPRESSION</em></dt>
<dd>
<p>include only sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-l, --local-csq</strong></dt>
<dd>
<p>switch off haplotype-aware calling, run localized predictions considering
only one VCF record at a time</p>
</dd>
<dt class="hdlist1"><strong>-n, --ncsq</strong> <em>INT</em></dt>
<dd>
<p>maximum number of per-haplotype consequences to consider for each site. The INFO/BCSQ column includes
all consequences, but only the first <em>INT</em> will be referenced by the FORMAT/BCSQ fields.
The default value is 15 which corresponds to one 32-bit integer per diploid
sample, after accounting for values reserved by the BCF specification.
Note that increasing the value leads to increased size of the output BCF.</p>
</dd>
<dt class="hdlist1"><strong>--no-version</strong></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-O, --output-type</strong> <em>t</em>|<em>b</em>|<em>u</em>|<em>z</em>|<em>v</em>[0-9]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong>. In addition, a custom tab-delimited
plain text output can be printed (<em>t</em>).</p>
</dd>
<dt class="hdlist1"><strong>-p, --phase</strong> <em>a</em>|<em>m</em>|<em>r</em>|<em>R</em>|<em>s</em></dt>
<dd>
<p>how to handle unphased heterozygous genotypes:</p>
<div class="dlist">
<dl>
<dt class="hdlist1"><em>a</em></dt>
<dd>
<p>take GTs as is, create haplotypes regardless of phase (0/1 &#8594; 0|1)</p>
</dd>
<dt class="hdlist1"><em>m</em></dt>
<dd>
<p>merge all GTs into a single haplotype (0/1 &#8594; 1, 1/2 &#8594; 1)</p>
</dd>
<dt class="hdlist1"><em>r</em></dt>
<dd>
<p>require phased GTs, throw an error on unphased heterozygous GTs</p>
</dd>
<dt class="hdlist1"><em>R</em></dt>
<dd>
<p>create non-reference haplotypes if possible (0/1 &#8594; 1|1, 1/2 &#8594; 1|2)</p>
</dd>
<dt class="hdlist1"><em>s</em></dt>
<dd>
<p>skip unphased heterozygous GTs</p>
</dd>
</dl>
</div>
</dd>
<dt class="hdlist1"><strong>-q, --quiet</strong></dt>
<dd>
<p>suppress warning messages</p>
</dd>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-s, --samples</strong> <em>LIST</em></dt>
<dd>
<p>samples to include or "-" to apply all variants and ignore samples</p>
</dd>
<dt class="hdlist1"><strong>-S, --samples-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-t, --targets</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-T, --targets-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--targets-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--unify-chr-names</strong> <em>0</em>|<em>1</em></dt>
<dd>
<p>Automatically detect and unify chromosome naming conventions in the GFF, fasta
and VCF, such as "chrX" vs "X". The chromosome names in the output VCF will match
that of the input VCF. The default is to attempt the automatic translation.</p>
</dd>
<dt class="hdlist1"><strong>-v, --verbose</strong> <em>INT</em></dt>
<dd>
<p>verbosity level (0-2)</p>
</dd>
<dt class="hdlist1"><strong>-W</strong>[<em>FMT</em>]<strong>, -W</strong>[=<em>FMT</em>]<strong>, --write-index</strong>[=<em>FMT</em>]</dt>
<dd>
<p>Automatically index the output file.  <em>FMT</em> is optional and can be
one of "tbi" or "csi" depending on output file format.</p>
</dd>
</dl>
</div>
<div class="paragraph">
<p><strong>Examples:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre>    # Basic usage
    bcftools csq -f hs37d5.fa -g Homo_sapiens.GRCh37.82.gff3.gz in.vcf -Ob -o out.bcf

    # Extract the translated haplotype consequences. The following TBCSQ variations
    # are recognised:
    #   %TBCSQ    .. print consequences in all haplotypes in separate columns
    #   %TBCSQ{0} .. print the first haplotype only
    #   %TBCSQ{1} .. print the second haplotype only
    #   %TBCSQ{*} .. print a list of unique consequences present in either haplotype
    bcftools query -f'[%CHROM\t%POS\t%SAMPLE\t%TBCSQ\n]' out.bcf</pre>
</div>
</div>
<div class="paragraph">
<p><strong>Examples of BCSQ annotation:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre>    # Two separate VCF records at positions 2:122106101 and 2:122106102
    # change the same codon. This UV-induced C&gt;T dinucleotide mutation
    # has been annotated fully at the position 2:122106101 with
    #   - consequence type
    #   - gene name
    #   - ensembl transcript ID
    #   - coding strand (+ fwd, - rev)
    #   - amino acid position (in the coding strand orientation)
    #   - list of corresponding VCF variants
    # The annotation at the second position gives the position of the full
    # annotation
    BCSQ=missense|CLASP1|ENST00000545861|-|1174P&gt;1174L|122106101G&gt;A+122106102G&gt;A
    BCSQ=@122106101

    # A frame-restoring combination of two frameshift insertions C&gt;CG and T&gt;TGG
    BCSQ=@46115084
    BCSQ=inframe_insertion|COPZ2|ENST00000006101|-|18AGRGP&gt;18AQAGGP|46115072C&gt;CG+46115084T&gt;TGG

    # Stop gained variant
    BCSQ=stop_gained|C2orf83|ENST00000264387|-|141W&gt;141*|228476140C&gt;T

    # The consequence type of a variant downstream from a stop are prefixed with *
    BCSQ=*missense|PER3|ENST00000361923|+|1028M&gt;1028T|7890117T&gt;C</pre>
</div>
</div>
<div class="paragraph">
<p><strong>Supported consequence types</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre>3_prime_utr
5_prime_utr
coding_sequence
feature_elongation
frameshift
inframe_altering
inframe_deletion
inframe_insertion
intergenic
intron
missense
non_coding
splice_acceptor
splice_donor
splice_region
start_lost
start_retained
stop_gained
stop_lost
stop_retained
synonymous</pre>
</div>
</div>
<div class="paragraph">
<p>See also <a href="https://ensembl.org/info/genome/variation/prediction/predicted_data.html" class="bare">https://ensembl.org/info/genome/variation/prediction/predicted_data.html</a></p>
</div>
</div>
<div class="sect2">
<h3 id="filter">bcftools filter <em>[OPTIONS]</em> <em>FILE</em></h3>
<div class="paragraph">
<p>Apply fixed-threshold filters.</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-e, --exclude</strong> <em>EXPRESSION</em></dt>
<dd>
<p>exclude sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-g, --SnpGap</strong> <em>INT</em>[:'indel',<em>mnp</em>,<em>bnd</em>,<em>other</em>,<em>overlap</em>]</dt>
<dd>
<p>filter SNPs within <em>INT</em> base pairs of an indel or other other variant type. The following example
demonstrates the logic of <strong>--SnpGap</strong> <em>3</em> applied on a deletion and
an insertion:</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>The SNPs at positions 1 and 7 are filtered, positions 0 and 8 are not:
         0123456789
    ref  .G.GT..G..
    del  .A.G-..A..
Here the positions 1 and 6 are filtered, 0 and 7 are not:
         0123-456789
    ref  .G.G-..G..
    ins  .A.GT..A..</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-G, --IndelGap</strong> <em>INT</em></dt>
<dd>
<p>filter clusters of indels separated by <em>INT</em> or fewer base pairs allowing
only one to pass. The following example demonstrates the logic of
<strong>--IndelGap</strong> <em>2</em> applied on a deletion and an insertion:</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>The second indel is filtered:
         012345678901
    ref  .GT.GT..GT..
    del  .G-.G-..G-..
And similarly here, the second is filtered:
         01 23 456 78
    ref  .A-.A-..A-..
    ins  .AT.AT..AT..</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-i, --include</strong> <em>EXPRESSION</em></dt>
<dd>
<p>include only sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>--mask</strong> [^]<em>REGION</em></dt>
<dd>
<p>Soft filter regions, prepepend "^" to negate. Requires <strong>-s, --soft-filter</strong>.</p>
</dd>
<dt class="hdlist1"><strong>-M, --mask-file</strong> [^]<em>FILE</em></dt>
<dd>
<p>Soft filter regions listed in a file, "^" to negate.  Requires <strong>-s, --soft-filter</strong>.</p>
</dd>
<dt class="hdlist1"><strong>--mask-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>Same as <strong>--regions-overlap</strong> but for <strong>--mask/--mask-file</strong>.
See <strong><a href="#common_options">Common Options</a></strong>. [1]</p>
</dd>
<dt class="hdlist1"><strong>-m, --mode</strong> [<em>+x</em>]</dt>
<dd>
<p>define behaviour at sites with existing FILTER annotations. The default
mode replaces existing filters of failed sites with a new FILTER string
while leaving sites which pass untouched when non-empty and setting to
"PASS" when the FILTER string is absent. The "+" mode appends new FILTER
strings of failed sites instead of replacing them. The "x" mode resets
filters of sites which pass to "PASS". Modes "+" and "x" can both be set.</p>
</dd>
<dt class="hdlist1"><strong>--no-version</strong></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-O, --output-type</strong> <em>b</em>|<em>u</em>|<em>z</em>|<em>v</em>[0-9]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-s, --soft-filter</strong> <em>STRING</em>|<em>+</em></dt>
<dd>
<p>annotate FILTER column with <em>STRING</em> or, with <em>+</em>, a unique filter name generated
by the program ("Filter%d"). Applies to records that do not meet filter expression.</p>
</dd>
<dt class="hdlist1"><strong>-S, --set-GTs</strong> <em>.</em>|<em>0</em></dt>
<dd>
<p>set genotypes of failed samples to missing value (<em>.</em>) or reference allele (<em>0</em>)</p>
</dd>
<dt class="hdlist1"><strong>-t, --targets</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-T, --targets-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--targets-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--threads</strong> <em>INT</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-W</strong>[<em>FMT</em>]<strong>, -W</strong>[=<em>FMT</em>]<strong>, --write-index</strong>[=<em>FMT</em>]</dt>
<dd>
<p>Automatically index the output file.  <em>FMT</em> is optional and can be
one of "tbi" or "csi" depending on output file format.</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="gtcheck">bcftools gtcheck [<em>OPTIONS</em>] [<strong>-g</strong> <em>genotypes.vcf.gz</em>] <em>query.vcf.gz</em></h3>
<div class="paragraph">
<p>Checks sample identity. The program can operate in two modes. If the <strong>-g</strong>
option is given, the identity of samples from <em>query.vcf.gz</em>
is checked against the samples in the <strong>-g</strong> file.
Without the <strong>-g</strong> option, multi-sample cross-check of samples in <em>query.vcf.gz</em> is performed.</p>
</div>
<div class="paragraph">
<p>Note that the interpretation of the discordance score depends on the options provided (specifically <strong>-e</strong> and
<strong>-u</strong>) and on the available annotations (FORMAT/PL vs FORMAT/GT).
The discordance score can be interpreted as the number of mismatching genotypes only if GT-vs-GT matching is performed.</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--distinctive-sites</strong> <em>NUM[,MEM[,DIR]]</em></dt>
<dd>
<p>Find sites that can distinguish between at least NUM sample pairs. If the number is smaller or equal to 1,
it is interpreted as the fraction of pairs.  The optional MEM string sets the maximum memory used for
in-memory sorting and DIR is the temporary directory for external sorting. This option requires also
<strong>--pairs</strong> to be given.</p>
</dd>
<dt class="hdlist1"><strong>--dry-run</strong></dt>
<dd>
<p>Stop after first record to estimate required time.</p>
</dd>
<dt class="hdlist1"><strong>-e, --exclude</strong> [<em>qry</em>|<em>gt</em>]:'EXPRESSION'</dt>
<dd>
<p>Exclude sites from query file (<em>qry:</em>) or genotype file (<em>gt:</em>) for which <em>EXPRESSION</em> is true.
For valid expressions see <strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-E, --error-probability</strong> <em>INT</em></dt>
<dd>
<p>Interpret genotypes and genotype likelihoods probabilistically. The value of <em>INT</em>
represents genotype quality when GT tag is used (e.g. Q=30 represents one error in 1,000 genotypes and
Q=40 one error in 10,000 genotypes) and is ignored when PL tag is used (in that case an arbitrary
non-zero integer can be provided).
&#160;<br>
&#160;<br>
If <strong>-E</strong> is set to 0, the discordance score can be interpreted as the number of mismatching genotypes,
but only in the GT-vs-GT matching mode. See the <strong>-u, --use</strong> option below for additional notes and caveats.
&#160;<br>
&#160;<br>
If performance is an issue, set <strong>-E 0</strong> for faster run times but less accurate results.
&#160;<br>
&#160;<br>
Note that in previous versions of bcftools (&#8656;1.18), this option used to be a smaller case <strong>-e</strong>. It
changed to make room for the filtering option <strong>-e, --exclude</strong> to stay consistent across other
commands.</p>
</dd>
<dt class="hdlist1"><strong>-g, --genotypes</strong> <em>FILE</em></dt>
<dd>
<p>VCF/BCF file with reference genotypes to compare against</p>
</dd>
<dt class="hdlist1"><strong>-H, --homs-only</strong></dt>
<dd>
<p>Homozygous genotypes only, useful with low coverage data (requires <strong>-g, --genotypes</strong>)</p>
</dd>
<dt class="hdlist1"><strong>-i, --include</strong> [<em>qry</em>|<em>gt</em>]:'EXPRESSION'</dt>
<dd>
<p>Include sites from query file (<em>qry:</em>) or genotype file (<em>gt:</em>) for which <em>EXPRESSION</em> is true.
For valid expressions see <strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>--n-matches</strong> <em>INT</em></dt>
<dd>
<p>Print only top INT matches for each sample, 0 for unlimited. Use negative value
to sort by HWE probability rather than the number of discordant sites. Note
that average score is used to determine the top matches, not absolute values.</p>
</dd>
<dt class="hdlist1"><strong>--no-HWE-prob</strong></dt>
<dd>
<p>Disable calculation of HWE probability to reduce memory requirements with
comparisons between very large number of sample pairs.</p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>Write to <em>FILE</em> rather than to standard output, where it is written by default.</p>
</dd>
<dt class="hdlist1"><strong>-O, --output-type</strong> <em>t</em>|<em>z</em></dt>
<dd>
<p>Write a plain (<em>t</em>) or compressed (<em>z</em>) text tab-delimited output.</p>
</dd>
<dt class="hdlist1"><strong>-p, --pairs</strong> <em>LIST</em></dt>
<dd>
<p>A comma-separated list of sample pairs to compare. When the <strong>-g</strong> option is given, the first
sample must be from the query file, the second from the <strong>-g</strong> file, third from the query file
etc (qry,gt[,qry,gt..]). Without the <strong>-g</strong> option, the pairs are created the same way but both
samples are from the query file (qry,qry[,qry,qry..])</p>
</dd>
<dt class="hdlist1"><strong>-P, --pairs-file</strong> <em>FILE</em></dt>
<dd>
<p>A file with tab-delimited sample pairs to compare. The first sample in the pair must come
from the query file, the second from the genotypes file when <strong>-g</strong> is given</p>
</dd>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>Restrict to comma-separated list of regions, see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1">*-R, --regions-file' <em>FILE</em></dt>
<dd>
<p>Restrict to regions listed in a file, see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
</dl>
</div>
<div class="paragraph">
<p><strong>-s, --samples</strong> [<em>qry</em>|<em>gt</em>]:'LIST':
    List of query samples or <strong>-g</strong> samples. If neither <strong>-s</strong> nor <strong>-S</strong> are given, all possible sample
    pair combinations are compared</p>
</div>
<div class="paragraph">
<p><strong>-S, --samples-file</strong> [<em>qry</em>|<em>gt</em>]:'FILE'
    File with the query or <strong>-g</strong> samples to compare.  If neither <strong>-s</strong> nor <strong>-S</strong> are given, all possible sample
    pair combinations are compared</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-t, --targets</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-T, --targets-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--targets-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-u, --use</strong> <em>TAG1</em>[,<em>TAG2</em>]</dt>
<dd>
<p>specifies which tag to use in the query file (<em>TAG1</em>) and the <strong>-g</strong> (<em>TAG2</em>) file.
By default, the PL tag is used in the query file and, when available, the GT tags in the
<strong>-g</strong> file.
&#160;<br>
&#160;<br>
Note that when the requested tag is not available, the program will attempt to use
the other tag. The output includes the number of sites that were matched by the four
possible modes (for example GT-vs-GT or GT-vs-PL).</p>
</dd>
</dl>
</div>
<div class="paragraph">
<p><strong>Examples:</strong></p>
</div>
<div class="listingblock">
<div class="content">
<pre>   # Check discordance of all samples from B against all samples in A
   bcftools gtcheck -g A.bcf B.bcf

   # Limit comparisons to the given list of samples
   bcftools gtcheck -s gt:a1,a2,a3 -s qry:b1,b2 -g A.bcf B.bcf

   # Compare only two pairs a1,b1 and a1,b2
   bcftools gtcheck -p a1,b1,a1,b2 -g A.bcf B.bcf</pre>
</div>
</div>
</div>
<div class="sect2">
<h3 id="head">bcftools head [<em>OPTIONS</em>] [<em>FILE</em>]</h3>
<div class="paragraph">
<p>By default, prints all headers from the specified input file to standard output
in VCF format. The input file may be in VCF or BCF format; if no <em>FILE</em> is
specified, standard input will be read. With appropriate options, only some
of the headers and/or additionally some of the variant records will be printed.</p>
</div>
<div class="paragraph">
<p>The <strong>bcftools head</strong> command outputs VCF headers almost exactly as they appear
in the input file: it may add a <code>##FILTER=&lt;ID=PASS&gt;</code> header if not already
present, but it never adds version or command line information itself.</p>
</div>
<div class="sect3">
<h4 id="_options">Options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --header</strong> <em>INT</em></dt>
<dd>
<p>Display only the first <em>INT</em> header lines.
By default, all header lines are displayed.</p>
</dd>
<dt class="hdlist1"><strong>-n, --records</strong> <em>INT</em></dt>
<dd>
<p>Also display the first <em>INT</em> variant records.
By default, no variant records are displayed.</p>
</dd>
<dt class="hdlist1"><strong>-s, --samples</strong> <em>INT</em></dt>
<dd>
<p>Display the first <em>INT</em> variant records including the last #CHROM header line with samples.
Running with <strong>-s 0</strong> alone outputs the #CHROM header line only. Note that
the list of samples, with each sample per line, can be obtained with <code>bcftools query</code> using
the option <strong>-l, --list-samples</strong>.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="index">bcftools index [<em>OPTIONS</em>]  <em>in.bcf</em>|<em>in.vcf.gz</em></h3>
<div class="paragraph">
<p>Creates index for bgzip compressed VCF/BCF files for random access. CSI
(coordinate-sorted index) is created by default. The CSI format
supports indexing of chromosomes up to length 2^31. TBI (tabix index)
index files, which support chromosome lengths up to 2^29, can be
created by using the <em>-t/--tbi</em> option or using the <em>tabix</em> program
packaged with htslib. When loading an index file, bcftools will try
the CSI first and then the TBI.</p>
</div>
<div class="sect3">
<h4 id="_indexing_options">Indexing options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-c, --csi</strong></dt>
<dd>
<p>generate CSI-format index for VCF/BCF files [default]</p>
</dd>
<dt class="hdlist1"><strong>-f, --force</strong></dt>
<dd>
<p>overwrite index if it already exists</p>
</dd>
<dt class="hdlist1"><strong>-m, --min-shift <em>INT</em></strong></dt>
<dd>
<p>set minimal interval size for CSI indices to 2^INT; default: 14</p>
</dd>
<dt class="hdlist1"><strong>-o, --output <em>FILE</em></strong></dt>
<dd>
<p>output file name. If not set, then the index will be created
using the input file name plus a <em>.csi</em> or <em>.tbi</em> extension</p>
</dd>
<dt class="hdlist1"><strong>-t, --tbi</strong></dt>
<dd>
<p>generate TBI-format index for VCF files</p>
</dd>
<dt class="hdlist1"><strong>--threads</strong> <em>INT</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_stats_options">Stats options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-a, --all</strong></dt>
<dd>
<p>Used in conjunction with <strong>-s, --stats</strong>, print per contig stats
for all contigs, even those with zero records and those for which
no stats are recorded in the index file (shown as <em>.</em>).</p>
</dd>
<dt class="hdlist1"><strong>-n, --nrecords</strong></dt>
<dd>
<p>print the number of records based on the CSI or TBI index files</p>
</dd>
<dt class="hdlist1"><strong>-s, --stats</strong></dt>
<dd>
<p>Print per contig stats based on the CSI or TBI index files.
Output format is three tab-delimited columns listing the contig
name, contig length (<em>.</em> if unknown) and number of records for
the contig. Contigs with zero records are not printed by default.</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="isec">bcftools isec [<em>OPTIONS</em>]  <em>A.vcf.gz</em> <em>B.vcf.gz</em> [&#8230;&#8203;]</h3>
<div class="paragraph">
<p>Creates intersections, unions and complements of VCF files. Depending
on the options, the program can output records from one (or more) files
which have (or do not have) corresponding records with the same position
in the other files.</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-c, --collapse</strong> <em>snps</em>|<em>indels</em>|<em>both</em>|<em>all</em>|<em>some</em>|<em>none</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-C, --complement</strong></dt>
<dd>
<p>output positions present only in the first file but missing in the others</p>
</dd>
<dt class="hdlist1"><strong>-e, --exclude</strong> <em>-</em>|<em>EXPRESSION</em></dt>
<dd>
<p>exclude sites for which <em>EXPRESSION</em> is true. If <strong>-e</strong> (or <strong>-i</strong>)
appears only once, the same filtering expression will be applied to all
input files.  Otherwise, <strong>-e</strong> or <strong>-i</strong> must be given for each input file.
To indicate that no filtering should be performed on a file, use "-" in
place of <em>EXPRESSION</em>, as shown in the example below.
For valid expressions see <strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-f, --apply-filters</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-i, --include</strong> <em>EXPRESSION</em></dt>
<dd>
<p>include only sites for which <em>EXPRESSION</em> is true. See discussion
of <strong>-e, --exclude</strong> above.</p>
</dd>
<dt class="hdlist1"><strong>-f, --file-list</strong> <em>FILE</em></dt>
<dd>
<p>Read file names from <em>FILE</em>, one file name per line.</p>
</dd>
<dt class="hdlist1"><strong>-n, --nfiles</strong> [+-=]<em>INT</em>|~<em>BITMAP</em></dt>
<dd>
<p>output positions present in this many (=), this many or more (+), this
many or fewer (-), or the exact same (~) files</p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong>.  When several files are being
output, their names are controlled via <strong>-p</strong> instead.</p>
</dd>
<dt class="hdlist1"><strong>-O, --output-type</strong> <em>b</em>|<em>u</em>|<em>z</em>|<em>v</em>[0-9]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-p, --prefix</strong> <em>DIR</em></dt>
<dd>
<p>if given, subset each of the input files accordingly. See also <strong>-w</strong>.</p>
</dd>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-t, --targets</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-T, --targets-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--targets-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-w, --write</strong> <em>LIST</em></dt>
<dd>
<p>comma-separated list of input files to output given as 1-based indices. With <strong>-p</strong> and no
<strong>-w</strong>, all files are written.</p>
</dd>
<dt class="hdlist1"><strong>-W</strong>[<em>FMT</em>]<strong>, -W</strong>[=<em>FMT</em>]<strong>, --write-index</strong>[=<em>FMT</em>]</dt>
<dd>
<p>Automatically index the output file.  <em>FMT</em> is optional and defaults
to tbi for vcf.gz and csi for bcf.  This is done automatically
with the <strong>-p</strong> option if the output format is compressed.</p>
</dd>
</dl>
</div>
<div class="sect3">
<h4 id="_examples">Examples:</h4>
<div class="paragraph">
<p>Create intersection and complements of two sets saving the output in dir/*</p>
</div>
<div class="listingblock">
<div class="content">
<pre>    bcftools isec -p dir A.vcf.gz B.vcf.gz</pre>
</div>
</div>
<div class="paragraph">
<p>Filter sites in A (require INFO/MAF&gt;=0.01) and B (require INFO/dbSNP) but not in C,
and create an intersection, including only sites which appear in at least two of
the files after filters have been applied</p>
</div>
<div class="listingblock">
<div class="content">
<pre>    bcftools isec -e'MAF&lt;0.01' -i'dbSNP=1' -e- A.vcf.gz B.vcf.gz C.vcf.gz -n +2 -p dir</pre>
</div>
</div>
<div class="paragraph">
<p>Extract and write records from A shared by both A and B using exact allele match</p>
</div>
<div class="listingblock">
<div class="content">
<pre>    bcftools isec -p dir -n=2 -w1 A.vcf.gz B.vcf.gz</pre>
</div>
</div>
<div class="paragraph">
<p>Extract records private to A or B comparing by position only</p>
</div>
<div class="listingblock">
<div class="content">
<pre>    bcftools isec -p dir -n-1 -c all A.vcf.gz B.vcf.gz</pre>
</div>
</div>
<div class="paragraph">
<p>Print a list of records which are present in A and B but not in C and D</p>
</div>
<div class="listingblock">
<div class="content">
<pre>    bcftools isec -n~1100 -c all A.vcf.gz B.vcf.gz C.vcf.gz D.vcf.gz</pre>
</div>
</div>
</div>
</div>
<div class="sect2">
<h3 id="merge">bcftools merge [<em>OPTIONS</em>] <em>A.vcf.gz</em> <em>B.vcf.gz</em> [&#8230;&#8203;]</h3>
<div class="paragraph">
<p>Merge multiple VCF/BCF files from non-overlapping sample sets to create one
multi-sample file.  For example, when merging file <em>A.vcf.gz</em> containing
samples <em>S1</em>, <em>S2</em> and <em>S3</em> and file <em>B.vcf.gz</em> containing samples <em>S3</em> and
<em>S4</em>, the output file will contain five samples named <em>S1</em>, <em>S2</em>, <em>S3</em>, <em>2:S3</em>
and <em>S4</em>.</p>
</div>
<div class="paragraph">
<p>Note that it is responsibility of the user to ensure that the sample names are
unique across all files. If they are not, the program will exit with an error
unless the option <strong>--force-samples</strong> is given.  The sample names can be
also given explicitly using the <strong>--print-header</strong> and <strong>--use-header</strong> options.</p>
</div>
<div class="paragraph">
<p>Note that only records from different files can be merged, never from the same file.
For "vertical" merge take a look at <strong><a href="#concat">bcftools concat</a></strong> or <strong><a href="#norm">bcftools norm</a> -m</strong> instead.</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--force-no-index</strong></dt>
<dd>
<p>synonymous to <strong>--no-index</strong></p>
</dd>
<dt class="hdlist1"><strong>--force-samples</strong></dt>
<dd>
<p>if the merged files contain duplicate samples names, proceed anyway.
Duplicate sample names will be resolved by prepending the index of the file
as it appeared on the command line to the conflicting sample name (see
<em>2:S3</em> in the above example).</p>
</dd>
<dt class="hdlist1"><strong>--force-single</strong></dt>
<dd>
<p>run even if only one file is given on input</p>
</dd>
<dt class="hdlist1"><strong>--print-header</strong></dt>
<dd>
<p>print only merged header and exit</p>
</dd>
<dt class="hdlist1"><strong>--use-header</strong> <em>FILE</em></dt>
<dd>
<p>use the VCF header in the provided text <em>FILE</em></p>
</dd>
<dt class="hdlist1"><strong>-0  --missing-to-ref</strong></dt>
<dd>
<p>assume genotypes at missing sites are 0/0</p>
</dd>
<dt class="hdlist1"><strong>-f, --apply-filters</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-F, --filter-logic</strong> <em>x</em>|<em>+</em></dt>
<dd>
<p>Set the output record to PASS if any of the inputs is PASS (<em>x</em>),
or apply all filters (<em>+</em>), which is the default.</p>
</dd>
<dt class="hdlist1"><strong>-g, --gvcf</strong> <em>-</em>|<em>FILE</em></dt>
<dd>
<p>merge gVCF blocks, INFO/END tag is expected. If the reference fasta
file <em>FILE</em> is not given and the dash (<em>-</em>) is given, unknown reference
bases generated at gVCF block splits will be substituted with N&#8217;s.
The <strong>--gvcf</strong> option uses the following default INFO rules:
<strong>-i QS:sum,MinDP:min,I16:sum,IDV:max,IMF:max</strong> and the following missing
rules: <strong>-M PL:max,AD:0</strong>.</p>
</dd>
<dt class="hdlist1"><strong>-i, --info-rules</strong> <em>-</em>|<em>TAG:METHOD</em>[,&#8230;&#8203;]</dt>
<dd>
<p>Rules for merging INFO fields (scalars or vectors) or <em>-</em> to disable the
default rules. <em>METHOD</em> is one of <em>sum</em>, <em>avg</em>, <em>min</em>, <em>max</em>, <em>join</em>.
Default is <em>DP:sum,DP4:sum</em> if these fields exist in the input files.
Fields with no specified rule will take the value from the first input file.
The merged QUAL value is currently set to the maximum. This behaviour is
not user controllable at the moment.</p>
</dd>
<dt class="hdlist1"><strong>-l, --file-list</strong> <em>FILE</em></dt>
<dd>
<p>Read file names from <em>FILE</em>, one file name per line.</p>
</dd>
<dt class="hdlist1"><strong>-L, --local-alleles</strong> <em>INT</em></dt>
<dd>
<p>Sites with many alternate alleles can require extremely large storage space which
can exceed the 2GB size limit representable by BCF. This is caused
by Number=G tags (such as FORMAT/PL) which store a value for each combination of reference
and alternate alleles. The <strong>-L, --local-alleles</strong> option allows replacement of such tags
with a localized tag (FORMAT/LPL) which only includes a subset of alternate alleles relevant
for that sample. A new FORMAT/LAA tag is added which lists 1-based indices of the
alternate alleles relevant (local) for the current sample. The number <em>INT</em> gives the
maximum number of alternate alleles that can be included in the PL tag. The default value
is 0 which disables the feature and outputs values for all alternate alleles.</p>
</dd>
<dt class="hdlist1"><strong>-m, --merge</strong> <em>snps</em>|<em>indels</em>|<em>both</em>|<em>snp-ins-del</em>|<em>all</em>|<em>none</em>|<em>id</em>[,<em>*</em>]</dt>
<dd>
<p>The option controls what types of multiallelic records can be created. If single asterisk
<em>*</em> is appended, the unobserved allele <em>&lt;*&gt;</em> or <em>&lt;NON_REF&gt;</em> will be removed at variant sites;
if two asterisks <em>**</em> are appended, the unobserved allele will be removed all sites.</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>-m none        ..  no new multiallelics, output multiple records instead
-m snps        ..  allow multiallelic SNP records
-m indels      ..  allow multiallelic indel records
-m both        ..  both SNP and indel records can be multiallelic
-m both,*      ..  same as above but remove &lt;*&gt; (or &lt;NON_REF&gt;) from variant sites
-m both,**     ..  same as above but remove &lt;*&gt; (or &lt;NON_REF&gt;) at all sites
-m all         ..  SNP records can be merged with indel records
-m snp-ins-del ..  allow multiallelic SNVs, insertions, deletions, but don't mix them
-m id          ..  merge by ID</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-M, --missing-rules</strong> <em>-</em>|<em>TAG:METHOD</em>[,&#8230;&#8203;]</dt>
<dd>
<p>Rules for merging vector tags at multiallelic sites. When input files have different alternate
alleles, vector fields pertaining to unobserved alleles are set to missing (<em>.</em>) by default.
The <em>METHOD</em> is one of <em>.</em> (the default, use missing values), <em>NUMBER</em> (use a constant value, e.g. 0),
<em>max</em> (the maximum value observed for other alleles in the sample). When <strong>--gvcf</strong> option is set,
the rule <strong>-M PL:max,AD:0</strong> is implied. This can be overridden with providing <strong>-M -</strong> or <strong>-M PL:.,AD:.</strong>.
Note that if the unobserved allele is explicitly present as <em>&lt;*&gt;</em> or <em>&lt;NON_REF&gt;</em>, then its corresponding
value will be used regardless of <strong>-M</strong> settings.</p>
</dd>
<dt class="hdlist1"><strong>--no-index</strong></dt>
<dd>
<p>the option allows files to be merged without indexing them first. In order for this
option to work, the user must ensure that the input files have chromosomes in
the same order and consistent with the order of sequences in the VCF header.</p>
</dd>
<dt class="hdlist1"><strong>--no-version</strong></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-O, --output-type</strong> <em>b</em>|<em>u</em>|<em>z</em>|<em>v</em>[0-9]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--threads</strong> <em>INT</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-W</strong>[<em>FMT</em>]<strong>, -W</strong>[=<em>FMT</em>]<strong>, --write-index</strong>[=<em>FMT</em>]</dt>
<dd>
<p>Automatically index the output file.  <em>FMT</em> is optional and can be
one of "tbi" or "csi" depending on output file format.</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="mpileup">bcftools mpileup [<em>OPTIONS</em>] <strong>-f</strong> <em>ref.fa</em> <em>in.bam</em> [<em>in2.bam</em> [&#8230;&#8203;]]</h3>
<div class="paragraph">
<p>Generate VCF or BCF containing genotype likelihoods for one or multiple
alignment (BAM or CRAM) files. This is based on the original
<strong>samtools mpileup</strong> command (with the <strong>-v</strong> or <strong>-g</strong> options) producing
genotype likelihoods in VCF or BCF format, but not the textual pileup
output. The <strong>mpileup</strong> command was transferred to bcftools in order to
avoid errors resulting from use of incompatible versions of samtools
and bcftools when using in the mpileup+bcftools call pipeline.</p>
</div>
<div class="paragraph">
<p>Individuals are identified from the SM tags in the @RG header lines.  Multiple
individuals can be pooled in one alignment file, also one individual can be
separated into multiple files.  If sample identifiers are absent, each input
file is regarded as one sample.</p>
</div>
<div class="paragraph">
<p>Note that there are two orthogonal ways to specify locations in the
input file; via <strong>-r</strong> <em>region</em> and <strong>-t</strong> <em>positions</em>.  The
former uses (and requires) an index to do random access while the
latter streams through the file contents filtering out the specified
regions, requiring no index.  The two may be used in conjunction.  For
example a BED file containing locations of genes in chromosome 20
could be specified using <strong>-r 20 -t chr20.bed</strong>, meaning that the
index is used to find chromosome 20 and then it is filtered for the
regions listed in the BED file. Also note that the <strong>-r</strong> option can be much
slower than <strong>-t</strong> with many regions and can require more memory when
multiple regions and many alignment files are processed.</p>
</div>
<div class="sect3">
<h4 id="_input_options">Input options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-6, --illumina1.3+</strong></dt>
<dd>
<p>Assume the quality is in the Illumina 1.3+ encoding.</p>
</dd>
<dt class="hdlist1"><strong>-A, --count-orphans</strong></dt>
<dd>
<p>Include anomalous read pairs in variant calling, i.e. reads with
flag PAIRED but not PROPER_PAIR set. By default such reads are discarded.</p>
</dd>
<dt class="hdlist1"><strong>-b, --bam-list</strong> <em>FILE</em></dt>
<dd>
<p>List of input alignment files, one file per line [null]</p>
</dd>
<dt class="hdlist1"><strong>-B, --no-BAQ</strong></dt>
<dd>
<p>Disable probabilistic realignment for the computation of base alignment
quality (BAQ). BAQ is the Phred-scaled probability of a read base being
misaligned. Applying this option greatly helps to reduce false SNPs caused
by misalignments.</p>
</dd>
<dt class="hdlist1"><strong>-C, --adjust-MQ</strong> <em>INT</em></dt>
<dd>
<p>Coefficient for downgrading mapping quality for reads containing
excessive mismatches.  Mismatches are counted as a proportion of the
number of aligned bases ("M", "X" or "=" CIGAR operations), along with
their quality, to derive an upper-bound of the mapping quality.
Original mapping qualities lower than this are left intact, while
higher ones are capped at the new adjusted score.</p>
<div class="paragraph">
<p>The exact formula is complex and likely tuned to specific instruments
and specific alignment tools, so this option is disabled by default
(indicated as having a zero value).  Variables in the formulae and
their meaning are defined below.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>Variable    Meaning / formula
M           The number of matching CIGAR bases (operation "M", "X" or "=").
X           The number of substitutions with quality &gt;= 13.
SubQ        The summed quality of substitution bases included in X, capped
            at a maximum of quality 33 per mismatching base.
ClipQ       The summed quality of soft-clipped or hard-clipped bases. This
            has no minimum or maximum quality threshold per base.  For
            hard-clipped bases the per-base quality is taken as 13.

T           SubQ - 10 * log10(M^X / X!) + ClipQ/5
Cap         MAX(0, INT * sqrt((INT - T) / INT))</pre>
</div>
</div>
<div class="paragraph">
<p>Some notes on the impact of this.</p>
</div>
<div class="ulist">
<ul>
<li>
<p>As the number of mismatches increases, the mapping quality cap
reduces, eventually resulting in discarded alignments.</p>
</li>
<li>
<p>High quality mismatches reduces the cap faster than low quality
mismatches.</p>
</li>
<li>
<p>The starting INT value also acts as a hard cap on mapping quality,
even when zero mismatches are observed.</p>
</li>
<li>
<p>Indels have no impact on the mapping quality.</p>
<div class="paragraph">
<p>The intent of this option is to work around aligners that compute a
mapping quality using a local alignment without having any regard to
the degree of clipping required or consideration of potential
contamination or large scale insertions with respect to the reference.
A record may align uniquely and have no close second match, but having
a high number of mismatches may still imply that the reference is not
the correct site.</p>
</div>
<div class="paragraph">
<p>However we do not recommend use of this parameter unless you fully
understand the impact of it and have determined that it is appropriate
for your sequencing technology.</p>
</div>
</li>
</ul>
</div>
</dd>
<dt class="hdlist1"><strong>-D, --full-BAQ</strong></dt>
<dd>
<p>Run the BAQ algorithm on all reads, not just those in problematic
regions.  This matches the behaviour for Bcftools 1.12 and earlier.</p>
<div class="paragraph">
<p>By default mpileup uses heuristics to decide when to apply the BAQ
algorithm.  Most sequences will not be BAQ adjusted, giving a CPU
time closer to --no-BAQ, but it will still be applied in regions
with suspected problematic alignments.  This has been tested to
work well on single sample data with even allele frequency, but
the reliability is unknown for multi-sample calling and for low
allele frequency variants so full BAQ is still recommended in
those scenarios.</p>
</div>
</dd>
<dt class="hdlist1"><strong>-d, --max-depth</strong> <em>INT</em></dt>
<dd>
<p>At a position, read maximally <em>INT</em> reads per input file. Note that
the original <strong>samtools mpileup</strong> command had a minimum value of <em>8000/n</em>
where <em>n</em> was the number of input files given to mpileup. This means that
in <strong>samtools mpileup</strong> the default was highly likely to be increased and the
<strong>-d</strong> parameter would have an effect only once above the cross-sample minimum of 8000.
This behavior was problematic when working with a combination of
single- and multi-sample bams, therefore in <strong>bcftools mpileup</strong> the user
is given the full control (and responsibility), and an informative message
is printed instead [250]</p>
</dd>
<dt class="hdlist1"><strong>-E, --redo-BAQ</strong></dt>
<dd>
<p>Recalculate BAQ on the fly, ignore existing BQ tags</p>
</dd>
<dt class="hdlist1"><strong>-f, --fasta-ref</strong> <em>FILE</em></dt>
<dd>
<p>The <strong>faidx</strong>-indexed reference file in the FASTA format. The file can be
optionally compressed by <strong>bgzip</strong>. Reference is required by default
unless the <strong>--no-reference</strong> option is set [null]</p>
</dd>
<dt class="hdlist1"><strong>--no-reference</strong></dt>
<dd>
<p>Do not require the <strong>--fasta-ref</strong> option.</p>
</dd>
<dt class="hdlist1"><strong>-G, --read-groups</strong> <em class="^">FILE</em></dt>
<dd>
<p>list of read groups to include or exclude if prefixed with "^".
One read group per line.  This file can also be used to assign new sample
names to read groups by giving the new sample name as a second
white-space-separated field, like this: "read_group_id new_sample_name".
If the read group name is not unique, also the bam file name can
be included: "read_group_id file_name sample_name".  If all
reads from the alignment file should be treated as a single sample, the
asterisk symbol can be used: "* file_name sample_name". Alignments without
a read group ID can be matched with "?". <strong>NOTE:</strong> The meaning of <strong>bcftools mpileup -G</strong>
is the opposite of <strong>samtools mpileup -G</strong>.</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    RG_ID_1
    RG_ID_2  SAMPLE_A
    RG_ID_3  SAMPLE_A
    RG_ID_4  SAMPLE_B
    RG_ID_5  FILE_1.bam  SAMPLE_A
    RG_ID_6  FILE_2.bam  SAMPLE_A
    *        FILE_3.bam  SAMPLE_C
    ?        FILE_3.bam  SAMPLE_D</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--indels-2.0</strong></dt>
<dd>
<p>A new EXPERIMENTAL indel calling model which aims to address some known deficiencies of
the current indel calling algorithm. Specifically, it uses diploid reference consensus
sequence. Note that in the current version it has the potential to increase sensitivity
but at the cost of decreased specificity.
Only works with short-read sequencing technologies.</p>
</dd>
<dt class="hdlist1"><strong>--indels-cns</strong></dt>
<dd>
<p>Another EXPERIMENTAL indel calling method, predating indels-2.0 in
PR form, but merged more recently.  It also uses a diploid
reference consensus, but with added parameters and heuristics to
optimise for a variety of sequencing platforms.  This is usually
faster and more accurate than the default caller and --indels-2.0,
but has not been tested on non-diploid samples and samples without
approximately even allele frequency.</p>
</dd>
<dt class="hdlist1"><strong>--no-indels-cns</strong></dt>
<dd>
<p>May be used to turn off --indels-cns mode when using one of the
newer profiles that has this enabled by default.</p>
</dd>
<dt class="hdlist1"><strong>-q, -min-MQ</strong> <em>INT</em></dt>
<dd>
<p>Minimum mapping quality for an alignment to be used [0]</p>
</dd>
<dt class="hdlist1"><strong>-Q, --min-BQ</strong> <em>INT</em></dt>
<dd>
<p>Minimum base quality for a base to be considered [13]</p>
</dd>
<dt class="hdlist1"><strong>--max-BQ</strong> <em>INT</em></dt>
<dd>
<p>Caps the base quality to a maximum value [60].  This can be
particularly useful on technologies that produce overly optimistic
high qualities, leading to too many false positives or incorrect
genotype assignments.</p>
</dd>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>CHR</em>|<em>CHR:POS</em>|<em>CHR:FROM-TO</em>|<em>CHR:FROM-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>Only generate mpileup output in given regions. Requires the alignment files
to be indexed.  If used in conjunction with -l then considers the intersection;
see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>FILE</em></dt>
<dd>
<p>As for <strong>-r, --regions</strong>, but regions read from FILE;
see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--ignore-RG</strong></dt>
<dd>
<p>Ignore RG tags. Treat all reads in one alignment file as one sample.</p>
</dd>
<dt class="hdlist1"><strong>--ls, --skip-all-set</strong></dt>
<dd>
<p>Skip reads with all of the FLAG bits set [null]</p>
</dd>
<dt class="hdlist1"><strong>--ns, --skip-any-set</strong></dt>
<dd>
<p>Skip reads with any of the FLAG bits set. This option replaces and
is synonymous to the deprecated <strong>--ff, --excl-flags</strong> [UNMAP,SECONDARY,QCFAIL,DUP]</p>
</dd>
<dt class="hdlist1"><strong>--lu, --skip-all-unset</strong></dt>
<dd>
<p>Skip reads with all of the FLAG bits unset. This option replaces and
is synonymous to the deprecated <strong>--rf, --incl-flags</strong> [null]</p>
</dd>
<dt class="hdlist1"><strong>--nu, --skip-any-unset</strong></dt>
<dd>
<p>Skip reads with any of the FLAG bits unset [null]</p>
</dd>
<dt class="hdlist1"><strong>-s, --samples</strong> <em class="^">LIST</em></dt>
<dd>
<p>list of sample names. See <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-S, --samples-file</strong> <em class="^">FILE</em></dt>
<dd>
<p>file of sample names to include or exclude if prefixed with "^".
One sample per line. This file can also be used to rename samples by giving
the new sample name as a second white-space-separated column, like this:
"old_name new_name".  If a sample name contains spaces, the spaces can be
escaped using the backslash character, for example "Not\ a\ good\ sample\
name".</p>
</dd>
<dt class="hdlist1"><strong>-t, --targets</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-T, --targets-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--targets-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-x, --ignore-overlaps</strong></dt>
<dd>
<p>Disable read-pair overlap detection.</p>
</dd>
<dt class="hdlist1"><strong>--seed</strong> <em>INT</em></dt>
<dd>
<p>Set the random number seed used when sub-sampling deep regions [0].</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_output_options">Output options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-a, --annotate</strong> <em>LIST</em></dt>
<dd>
<p>Comma-separated list of FORMAT and INFO tags to output. (case-insensitive,
the "FORMAT/" prefix is optional, and use "?" to list available annotations
on the command line) [null]:</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>FORMAT/AD   .. Allelic depth (Number=R,Type=Integer)
FORMAT/ADF  .. Allelic depths on the forward strand (Number=R,Type=Integer)
FORMAT/ADR  .. Allelic depths on the reverse strand (Number=R,Type=Integer)
FORMAT/DP   .. Number of high-quality bases (Number=1,Type=Integer)
FORMAT/SP   .. Phred-scaled strand bias P-value (Number=1,Type=Integer)
FORMAT/SCR  .. Number of soft-clipped reads (Number=1,Type=Integer)

INFO/AD     .. Total allelic depth (Number=R,Type=Integer)
INFO/ADF    .. Total allelic depths on the forward strand (Number=R,Type=Integer)
INFO/ADR    .. Total allelic depths on the reverse strand (Number=R,Type=Integer)
INFO/SCR    .. Number of soft-clipped reads (Number=1,Type=Integer)

FORMAT/DV   .. Deprecated in favor of FORMAT/AD; Number of high-quality non-reference bases, (Number=1,Type=Integer)
FORMAT/DP4  .. Deprecated in favor of FORMAT/ADF and FORMAT/ADR; Number of high-quality ref-forward, ref-reverse,
               alt-forward and alt-reverse bases (Number=4,Type=Integer)
FORMAT/DPR  .. Deprecated in favor of FORMAT/AD; Number of high-quality bases for each observed allele (Number=R,Type=Integer)
INFO/DPR    .. Deprecated in favor of INFO/AD; Number of high-quality bases for each observed allele (Number=R,Type=Integer)</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-g, --gvcf</strong> <em>INT</em>[,&#8230;&#8203;]</dt>
<dd>
<p>output gVCF blocks of homozygous REF calls, with depth (DP) ranges
specified by the list of integers. For example, passing <em>5,15</em> will
group sites into two types of gVCF blocks, the first with minimum
per-sample DP from the interval [5,15) and the latter with minimum
depth 15 or more. In this example, sites with minimum per-sample
depth less than 5 will be printed as separate records, outside of
gVCF blocks.</p>
</dd>
<dt class="hdlist1"><strong>--no-version</strong></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>Write output to <em>FILE</em>, rather than the default of standard output.
(The same short option is used for both <strong>--open-prob</strong> and <strong>--output</strong>.  If <strong>-o</strong>'s
argument contains any non-digit characters other than a leading + or -
sign,  it  is  interpreted  as <strong>--output</strong>.  Usually the filename extension
will take care of this, but to write to an entirely numeric filename use <strong>-o
./123</strong> or <strong>--output 123</strong>.)</p>
</dd>
<dt class="hdlist1"><strong>-O, --output-type</strong> <em>b</em>|<em>u</em>|<em>z</em>|<em>v</em>[0-9]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--threads</strong> <em>INT</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-U, --mwu-u</strong></dt>
<dd>
<p>The the previous Mann-Whitney U test score from version 1.12 and
earlier.  This is a probability score, but importantly it folds
probabilities above or below the desired score into the same P.
The new Mann-Whitney U test score is a "Z score", expressing the
score as the number of standard deviations away from the mean (with
zero being matching the mean).  It keeps both positive and
negative values.  This can be important for some tests where
errors are asymmetric.</p>
<div class="literalblock">
<div class="content">
<pre>This option changes the INFO field names produced back to the ones
used by the earlier Bcftools releases.  For excample BQBZ becomes
BQB.</pre>
</div>
</div>
</dd>
<dt class="hdlist1"><strong>-W</strong>[<em>FMT</em>]<strong>, -W</strong>[=<em>FMT</em>]<strong>, --write-index</strong>[=<em>FMT</em>]</dt>
<dd>
<p>Automatically index the output file.  <em>FMT</em> is optional and can be
one of "tbi" or "csi" depending on output file format.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_options_for_snpindel_genotype_likelihood_computation">Options for SNP/INDEL genotype likelihood computation</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-X, --config</strong> <em>STR</em></dt>
<dd>
<p>Specify a platform specific configuration profile.  Specifying the
profile as "list" will list the available profile names and the
parameters they change.  There are profiles named after a release,
which should be used if you wish to ensure forward compatibility
of results.  The non-versioned names (eg "illumina") will always
point to the most recent set of parameters for that instrument type.
The current values are:</p>
<div class="literalblock">
<div class="content">
<pre>1.12            -Q13 -h100 -m1</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>bgi
bgi-1.20        --indels-cns -B --indel-size 80 -F0.1 --indel-bias 0.9
                --seqq-offset 120</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>illumina-1.18   [ default values ]</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>illumina
illumina-1.20   --indels-cns --seqq-offset 125</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>ont             -B -Q5 --max-BQ 30 -I</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>ont-sup
ont-sup-1.20    --indels-cns -B -Q1 --max-BQ 35 --delta-BQ 99 -F0.2
                -o15 -e1 -h110 --del-bias 0.4 --indel-bias 0.7
                --poly-mqual --seqq-offset 130 --indel-size 80</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>pacbio-ccs-1.18 -D -Q5 --max-BQ 50 -F0.1 -o25 -e1 -M99999</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>pacbio-ccs
pacbio-ccs-1.20  --indels-cns -B -Q5 --max-BQ 50 -F0.1 -o25 -e1 -h300
                 --delta-BQ 10 --del-bias 0.4 --poly-mqual
                 --indel-bias 0.9 --seqq-offset 118 --indel-size 80
                 --score-vs-ref 0.7</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>ultima
ultima-1.20      --indels-cns -B -Q1 --max-BQ 30 --delta-BQ 10 -F0.15
                 -o20 -e10 -h250 --del-bias 0.3 --indel-bias 0.7
                 --poly-mqual --seqq-offset 140 --score-vs-ref 0.3
                 --indel-size 80</pre>
</div>
</div>
</dd>
<dt class="hdlist1"><strong>--ar, --ambig-reads</strong> <em>drop</em>|<em>incAD</em>|<em>incAD0</em></dt>
<dd>
<p>What to do with ambiguous indel reads that do not span an entire
short tandem repeat region: discard ambiguous reads from calling
and do not increment high-quality AD depth counters (<em>drop</em>),
exclude from calling but increment AD counters proportionally (<em>incAD</em>),
exclude from calling and increment the first value of the AD counter
(<em>incAD0</em>) [<em>drop</em>]</p>
</dd>
<dt class="hdlist1"><strong>-e, --ext-prob</strong> <em>INT</em></dt>
<dd>
<p>Phred-scaled gap extension sequencing error probability. Reducing <em>INT</em>
leads to longer indels [20]</p>
</dd>
<dt class="hdlist1"><strong>-F, --gap-frac</strong> <em>FLOAT</em></dt>
<dd>
<p>Minimum fraction of gapped reads [0.002]</p>
</dd>
<dt class="hdlist1"><strong>-h, --tandem-qual</strong> <em>INT</em></dt>
<dd>
<p>Coefficient for modeling homopolymer errors. Given an <em>l</em>-long homopolymer
run, the sequencing error of an indel of size s is modeled as <em>INT</em>*s/l [500]
Increasing this informs the caller that indels in long
homopolymers are more likely genuine and less likely to be
sequencing artifacts.  Hence increasing tandem-qual will have
higher recall and lower precision.  Bcftools 1.12 and earlier had
a default of 100, which was tuned around more error prone instruments.
Note changing this may have a minor impact on SNP calling too.
For maximum SNP calling accuracy, it may be preferable to adjust
this lower again, although this will adversely affect indels.</p>
</dd>
<dt class="hdlist1"><strong>--indel-bias</strong> <em>FLOAT</em></dt>
<dd>
<p>Skews the indel scores up or down, trading recall (low
false-negative) vs precision (low false-positive) [1.0].  In Bcftools
1.12 and earlier this parameter didn&#8217;t exist, but had an implied
value of 1.0.  If you are planning to do heavy filtering of
variants, selecting the best quality ones only (favouring
precision over recall), it is advisable to set this lower (such as
0.75) while higher depth samples or where you favour recall rates
over precision may work better with a higher value such as 2.0.</p>
</dd>
<dt class="hdlist1"><strong>--del-bias</strong> <em>FLOAT</em></dt>
<dd>
<p>Skews the likelihood of deletions over insertions.  Defaults to an
even distribution value of 1.0.  Lower values imply a higher rate
of false positive deletions (meaning candidate deletions are less
likely to be real).</p>
</dd>
<dt class="hdlist1"><strong>--indel-size</strong> <em>INT</em></dt>
<dd>
<p>Indel window size to use when assessing the quality of candidate indels.
Note that although the window size approximately corresponds to the maximum
indel size considered, it is not an exact threshold [110]</p>
</dd>
<dt class="hdlist1"><strong>--seqq-offset</strong> <em>INT</em></dt>
<dd>
<p>Tunes the importance of indel sequence quality per depth.  The
final "seqQ" quality used is "offset - 5*MIN(depth,20)". [120]</p>
</dd>
<dt class="hdlist1"><strong>--poly-mqual</strong></dt>
<dd>
<p>Use the lowest quality value within a homopolymer run, instead of
the quality immediately adjacent to the indel.  This may be
important for unclocked instruments, particularly ones with a flow
chemistry where runs of bases of identical type are incorporated
together.</p>
</dd>
<dt class="hdlist1"><strong>-I, --skip-indels</strong></dt>
<dd>
<p>Do not perform INDEL calling</p>
</dd>
<dt class="hdlist1"><strong>-L, --max-idepth</strong> <em>INT</em></dt>
<dd>
<p>Skip INDEL calling if the average per-sample depth is above <em>INT</em> [250]</p>
</dd>
<dt class="hdlist1"><strong>-m, --min-ireads</strong> <em>INT</em></dt>
<dd>
<p>Minimum number gapped reads for indel candidates <em>INT</em> [1]</p>
</dd>
<dt class="hdlist1"><strong>-M, --max-read-len</strong> <em>INT</em></dt>
<dd>
<p>The maximum read length permitted by the BAQ algorithm [500].
Variants are still called on longer reads, but they will not be
passed through the BAQ method.  This limit exists to prevent
excessively long BAQ times and high memory usage.  Note if partial
BAQ is enabled with <em>-D</em> then raising this parameter will likely
not have a significant a CPU cost.</p>
</dd>
<dt class="hdlist1"><strong>-o, --open-prob</strong> <em>INT</em></dt>
<dd>
<p>Phred-scaled gap open sequencing error probability. Reducing <em>INT</em> leads
to more indel calls. (The same short option is used for both <strong>--open-prob</strong>
and <strong>--output</strong>.  When -o&#8217;s argument contains only an optional + or - sign
followed by the digits 0 to 9, it is interpreted  as <strong>--open-prob</strong>.) [40]</p>
</dd>
<dt class="hdlist1"><strong>-p, --per-sample-mF</strong></dt>
<dd>
<p>Apply <strong>-m</strong> and <strong>-F</strong> thresholds per sample to increase sensitivity of calling.
By default both options are applied to reads pooled from all samples.</p>
</dd>
<dt class="hdlist1"><strong>-P, --platforms</strong> <em>STR</em></dt>
<dd>
<p>Comma-delimited  list  of  platforms (determined by <strong>@RG-PL</strong>) from which
indel candidates are obtained. It is recommended to collect indel
candidates from sequencing technologies that have low indel error rate
such as ILLUMINA [all]</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_examples_2">Examples:</h4>
<div class="paragraph">
<p>Call SNPs and short INDELs, then mark low quality sites and sites with the read
depth exceeding a limit. (The read depth should be adjusted to about twice the
average read depth as higher read depths usually indicate problematic regions
which are often enriched for artefacts.) One may consider to add <strong>-C50</strong> to
mpileup if mapping quality is overestimated  for reads containing  excessive
mismatches.  Applying this option usually helps for BWA-backtrack alignments,
but may not other aligners.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>    bcftools mpileup -Ou -f ref.fa aln.bam | \
    bcftools call -Ou -mv | \
    bcftools filter -s LowQual -e '%QUAL&lt;20 || DP&gt;100' &gt; var.flt.vcf</pre>
</div>
</div>
</div>
</div>
<div class="sect2">
<h3 id="norm">bcftools norm [<em>OPTIONS</em>] <em>file.vcf.gz</em></h3>
<div class="paragraph">
<p>Left-align and normalize indels, check if REF alleles match the reference,
split multiallelic sites into multiple rows; recover multiallelics from
multiple rows. Left-alignment and normalization will only be applied if
the <strong><a href="#fasta_ref">--fasta-ref</a></strong> option is supplied.</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-a, --atomize</strong></dt>
<dd>
<p>Decompose complex variants, e.g. split MNVs into consecutive SNVs.
See also <strong>--atom-overlaps</strong> and <strong>--old-rec-tag</strong>.</p>
</dd>
<dt class="hdlist1"><strong>--atom-overlaps</strong> <em>.</em>|<em>*</em></dt>
<dd>
<p>Alleles missing because of an overlapping variant can be set either
to missing (.) or to the star alele (*), as recommended by
the VCF specification. IMPORTANT: Note that asterisk is expaneded
by shell and must be put in quotes or escaped by a backslash:</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    # Before atomization:
    100  CC  C,GG   1/2

    # After:
    #   bcftools norm -a --atom-overlaps .
    100	 C	 G      ./1
    100	 CC	 C      1/.
    101	 C	 G      ./1

    # After:
    #   bcftools norm -a --atom-overlaps '*'
    #   bcftools norm -a --atom-overlaps \*
    100	 C	 G,*    2/1
    100	 CC	 C,*    1/2
    101	 C	 G,*    2/1</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-c, --check-ref</strong> <em>e</em>|<em>w</em>|<em>x</em>|<em>s</em></dt>
<dd>
<p>what to do when incorrect or missing REF allele is encountered:
exit (<em>e</em>), warn (<em>w</em>), exclude (<em>x</em>), or set/fix (<em>s</em>) bad sites.
The <em>w</em> option can be combined with <em>x</em> and <em>s</em>. Note that <em>s</em>
can swap alleles and will update genotypes (GT) and AC counts,
but will not attempt to fix PL or other fields. Also note, and this
cannot be stressed enough, that <em>s</em> will NOT fix strand issues in
your VCF, do NOT use it for that purpose!!! (Instead see
<a href="http://samtools.github.io/bcftools/howtos/plugin.af-dist.html" class="bare">http://samtools.github.io/bcftools/howtos/plugin.af-dist.html</a> and
&lt;<a href="http://samtools.github.io/bcftools/howtos/plugin.fixref.html&gt;" class="bare">http://samtools.github.io/bcftools/howtos/plugin.fixref.html&gt;</a>.)</p>
</dd>
<dt class="hdlist1"><strong>-d, --rm-dup</strong> <em>snps</em>|<em>indels</em>|<em>both</em>|<em>all</em>|<em>exact</em></dt>
<dd>
<p>If a record is present multiple times, output only the first instance.
See also <strong>--collapse</strong> in <strong><a href="#common_options">Common Options</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-D, --remove-duplicates</strong></dt>
<dd>
<p>If a record is present in multiple files, output only the first instance.
Alias for <strong>-d none</strong>, deprecated.</p>
</dd>
<dt class="hdlist1"><strong>-e, --exclude</strong> <em>EXPRESSION</em></dt>
<dd>
<p>do not normalize input records for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>. Note that duplicate removal ignores this option.</p>
</dd>
<dt class="hdlist1"><strong>-f, --fasta-ref</strong> <em>FILE</em><a id="fasta_ref"></a></dt>
<dd>
<p>reference sequence. Supplying this option will turn on left-alignment
and normalization, however, see also the <strong><a href="#do_not_normalize">--do-not-normalize</a></strong>
option below.</p>
</dd>
<dt class="hdlist1"><strong>--force</strong></dt>
<dd>
<p>try to proceed with <strong>-m-</strong> even if malformed tags with incorrect number of fields
are encountered, discarding such tags. (Experimental, use at your own risk.)</p>
</dd>
<dt class="hdlist1"><strong>-g, --gff-annot</strong> <em>FILE</em></dt>
<dd>
<p>when a GFF file is provided, follow HGVS 3&#8217;rule and right-align variants in transcripts on the forward
strand.  In case of overlapping transcripts, the default mode is to left-align the variant. For a
description of the supported GFF3 file format see <strong><a href="#csq">bcftools csq</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-i, --include</strong> <em>EXPRESSION</em></dt>
<dd>
<p>normalize only input records for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>. Note that duplicate removal ignores this option.</p>
</dd>
<dt class="hdlist1"><strong>--keep-sum</strong> <em>TAG</em>[,&#8230;&#8203;]</dt>
<dd>
<p>keep vector sum constant when splitting multiallelic sites. Only AD tag
is currently supported. See also <a href="https://github.com/samtools/bcftools/issues/360" class="bare">https://github.com/samtools/bcftools/issues/360</a></p>
</dd>
<dt class="hdlist1"><strong>-m, --multiallelics</strong> <strong>-</strong>|<strong>+</strong>[<em>snps</em>|<em>indels</em>|<em>both</em>|<em>any</em>]</dt>
<dd>
<p>split multiallelic sites into biallelic records (<strong>-</strong>) or join
biallelic sites into multiallelic records (<strong>+</strong>). An optional type string
can follow which controls variant types which should be split or merged
together: If only SNP records should be split or merged, specify <em>snps</em>; if
both SNPs and indels should be merged separately into two records, specify
<em>both</em>; if SNPs and indels should be merged into a single record, specify
<em>any</em>.
&#160;<br>
&#160;<br>
Note that multiallelic sites with both SNPs and indels will be split into
biallelic sites with both <strong>-m -snps</strong> and <strong>-m -indels</strong>.</p>
</dd>
<dt class="hdlist1"><strong>--multi-overlaps</strong> <em>0</em>|<em>.</em></dt>
<dd>
<p>use the reference (<em>0</em>) or missing (<em>.</em>) allele for overlapping alleles after
splitting multiallelic sites</p>
</dd>
<dt class="hdlist1"><strong>--no-version</strong></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-N, --do-not-normalize</strong><a id="do_not_normalize"></a></dt>
<dd>
<p>the <em>-c s</em> option can be used to fix or set the REF allele from the
reference <em>-f</em>. The <em>-N</em> option will not turn on indel normalisation
as the <em>-f</em> option normally implies</p>
</dd>
<dt class="hdlist1"><strong>--old-rec-tag</strong> <em>STR</em></dt>
<dd>
<p>Add INFO/STR annotation with the original record. The format of the
annotation is CHROM|POS|REF|ALT|USED_ALT_IDX.</p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-O, --output-type</strong> <em>b</em>|<em>u</em>|<em>z</em>|<em>v</em>[0-9]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-s, --strict-filter</strong></dt>
<dd>
<p>when merging (<em>-m+</em>), merged site is PASS only if all sites being merged PASS</p>
</dd>
<dt class="hdlist1"><strong>-S, --sort</strong> <em>pos</em>|<em>lex</em></dt>
<dd>
<p>when splitting sites or processing duplicates, sort records on output by
POS only (<em>pos</em>, the default) or by POS and lexicographically by REF+ALT (<em>lex</em>)</p>
</dd>
<dt class="hdlist1"><strong>-t, --targets</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-T, --targets-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--targets-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--threads</strong> <em>INT</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-v, --verbose</strong> <em>INT</em></dt>
<dd>
<p>verbosity level of GFF parsing (0-2)</p>
</dd>
<dt class="hdlist1"><strong>-w, --site-win</strong> <em>INT</em></dt>
<dd>
<p>maximum distance between two records to consider when locally
sorting variants which changed position during the realignment</p>
</dd>
<dt class="hdlist1"><strong>-W</strong>[<em>FMT</em>]<strong>, -W</strong>[=<em>FMT</em>]<strong>, --write-index</strong>[=<em>FMT</em>]</dt>
<dd>
<p>Automatically index the output file.  <em>FMT</em> is optional and can be
one of "tbi" or "csi" depending on output file format.</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="plugin">bcftools [plugin <em>NAME</em>|+<em>NAME</em>] <em>[OPTIONS]</em> <em>FILE</em>&#8201;&#8212;&#8201;<em>[PLUGIN OPTIONS]</em></h3>
<div class="paragraph">
<p>A common framework for various utilities. The plugins can be used
the same way as normal commands only their name is prefixed with "+".
Most plugins accept two types of parameters: general options shared by all
plugins followed by a separator, and a list of plugin-specific options.  There
are some exceptions to this rule, some plugins do not accept the common
options and implement their own parameters. Therefore please pay attention to
the usage examples that each plugin comes with.</p>
</div>
<div class="sect3">
<h4 id="_vcf_input_options_2">VCF input options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-e, --exclude</strong> <em>EXPRESSION</em></dt>
<dd>
<p>exclude sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-i, --include</strong> <em>EXPRESSION</em></dt>
<dd>
<p>include only sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-t, --targets</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-T, --targets-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--targets-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_vcf_output_options_2">VCF output options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--no-version</strong></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-O, --output-type</strong> <em>b</em>|<em>u</em>|<em>z</em>|<em>v</em>[0-9]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--threads</strong> <em>INT</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-W</strong>[<em>FMT</em>]<strong>, -W</strong>[=<em>FMT</em>]<strong>, --write-index</strong>[=<em>FMT</em>]</dt>
<dd>
<p>Automatically index the output file.  <em>FMT</em> is optional and can be
one of "tbi" or "csi" depending on output file format.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_plugin_options">Plugin options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-h, --help</strong></dt>
<dd>
<p>list plugin&#8217;s options</p>
</dd>
<dt class="hdlist1"><strong>-l, --list-plugins</strong></dt>
<dd>
<p>List all available plugins.</p>
<div class="paragraph">
<p>By default, appropriate system directories are searched for installed plugins.
    You can override this by setting the BCFTOOLS_PLUGINS environment variable
    to a colon-separated list of directories to search.
    If BCFTOOLS_PLUGINS begins with a colon, ends with a colon, or contains
    adjacent colons, the system directories are also searched at that position
    in the list of directories.</p>
</div>
</dd>
<dt class="hdlist1"><strong>-v, --verbose</strong></dt>
<dd>
<p>print debugging information to debug plugin failure</p>
</dd>
<dt class="hdlist1"><strong>-V, --version</strong></dt>
<dd>
<p>print version string and exit</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_list_of_plugins_coming_with_the_distribution">List of plugins coming with the distribution:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>ad-bias</strong></dt>
<dd>
<p>find positions with wildly varying ALT allele frequency (Fisher test on FMT/AD)</p>
</dd>
<dt class="hdlist1"><strong>add-variantkey</strong></dt>
<dd>
<p>add VariantKey INFO fields VKX and RSX</p>
</dd>
<dt class="hdlist1"><strong>af-dist</strong></dt>
<dd>
<p>collect AF deviation stats and GT probability distribution given AF and assuming HWE</p>
</dd>
<dt class="hdlist1"><strong>allele-length</strong></dt>
<dd>
<p>count the frequency of the length of REF, ALT and REF+ALT</p>
</dd>
<dt class="hdlist1"><strong>check-ploidy</strong></dt>
<dd>
<p>check if ploidy of samples is consistent for all sites</p>
</dd>
<dt class="hdlist1"><strong>check-sparsity</strong></dt>
<dd>
<p>print samples without genotypes in a region or chromosome</p>
</dd>
<dt class="hdlist1"><strong>color-chrs</strong></dt>
<dd>
<p>color shared chromosomal segments, requires trio VCF with phased GTs</p>
</dd>
<dt class="hdlist1"><strong>contrast</strong></dt>
<dd>
<p>runs a basic association test, per-site or in a region, and checks for novel alleles and
genotypes in two groups of samples. Adds the following INFO annotations:</p>
<div class="ulist">
<ul>
<li>
<p>PASSOC  .. Fisher&#8217;s exact test probability of genotypic association (REF vs non-REF allele)</p>
</li>
<li>
<p>FASSOC  .. proportion of non-REF allele in controls and cases</p>
</li>
<li>
<p>NASSOC  .. number of control-ref, control-alt, case-ref and case-alt alleles</p>
</li>
<li>
<p>NOVELAL .. lists samples with a novel allele not observed in the control group</p>
</li>
<li>
<p>NOVELGT .. lists samples with a novel genotype not observed in the control group</p>
</li>
</ul>
</div>
</dd>
<dt class="hdlist1"><strong>counts</strong></dt>
<dd>
<p>a minimal plugin which counts number of SNPs, Indels, and total number of sites.</p>
</dd>
<dt class="hdlist1"><strong>dosage</strong></dt>
<dd>
<p>print genotype dosage. By default the plugin searches for PL, GL and GT, in
that order.</p>
</dd>
<dt class="hdlist1"><strong>fill-from-fasta</strong></dt>
<dd>
<p>fill INFO or REF field based on values in a fasta file</p>
</dd>
<dt class="hdlist1"><strong>fill-tags</strong></dt>
<dd>
<p>set various INFO tags. The list of tags supported in this version:</p>
<div class="ulist">
<ul>
<li>
<p>INFO/AC         Number:A  Type:Integer  ..  Allele count in genotypes</p>
</li>
<li>
<p>INFO/AC_Hom     Number:A  Type:Integer  ..  Allele counts in homozygous genotypes</p>
</li>
<li>
<p>INFO/AC_Het     Number:A  Type:Integer  ..  Allele counts in heterozygous genotypes</p>
</li>
<li>
<p>INFO/AC_Hemi    Number:A  Type:Integer  ..  Allele counts in hemizygous genotypes</p>
</li>
<li>
<p>INFO/AF         Number:A  Type:Float    ..  Allele frequency</p>
</li>
<li>
<p>INFO/AN         Number:1  Type:Integer  ..  Total number of alleles in called genotypes</p>
</li>
<li>
<p>INFO/ExcHet     Number:A  Type:Float    ..  Test excess heterozygosity; 1=good, 0=bad</p>
</li>
<li>
<p>INFO/END        Number:1  Type:Integer  ..  End position of the variant</p>
</li>
<li>
<p>INFO/F_MISSING  Number:1  Type:Float    ..  Fraction of missing genotypes</p>
</li>
<li>
<p>INFO/HWE        Number:A  Type:Float    ..  HWE test (PMID:15789306); 1=good, 0=bad</p>
</li>
<li>
<p>INFO/MAF        Number:A  Type:Float    ..  Minor Allele frequency</p>
</li>
<li>
<p>INFO/NS         Number:1  Type:Integer  ..  Number of samples with data</p>
</li>
<li>
<p>INFO/TYPE       Number:.  Type:String   ..  The record type (REF,SNP,MNP,INDEL,etc)</p>
</li>
<li>
<p>FORMAT/VAF      Number:A  Type:Float    ..  The fraction of reads with the alternate allele, requires FORMAT/AD or ADF+ADR</p>
</li>
<li>
<p>FORMAT/VAF1     Number:1  Type:Float    ..  The same as FORMAT/VAF but for all alternate alleles cumulatively</p>
</li>
<li>
<p>TAG=func(TAG)   Number:1  Type:Integer  ..  Experimental support for user-defined expressions such as "DP=sum(DP)"</p>
</li>
</ul>
</div>
</dd>
<dt class="hdlist1"><strong>fixploidy</strong></dt>
<dd>
<p>sets correct ploidy</p>
</dd>
<dt class="hdlist1"><strong>fixref</strong></dt>
<dd>
<p>determine and fix strand orientation</p>
</dd>
<dt class="hdlist1"><strong>frameshifts</strong></dt>
<dd>
<p>annotate frameshift indels</p>
</dd>
<dt class="hdlist1"><strong>GTisec</strong></dt>
<dd>
<p>count genotype intersections across all possible sample subsets in a vcf file</p>
</dd>
<dt class="hdlist1"><strong>GTsubset</strong></dt>
<dd>
<p>output only sites where the requested samples all exclusively share a genotype</p>
</dd>
<dt class="hdlist1"><strong>guess-ploidy</strong></dt>
<dd>
<p>determine sample sex by checking genotype likelihoods (GL,PL) or genotypes (GT)
in the non-PAR region of chrX.</p>
</dd>
<dt class="hdlist1"><strong>gvcfz</strong></dt>
<dd>
<p>compress gVCF file by resizing non-variant blocks according to specified criteria</p>
</dd>
<dt class="hdlist1"><strong>impute-info</strong></dt>
<dd>
<p>add imputation information metrics to the INFO field based on selected FORMAT tags</p>
</dd>
<dt class="hdlist1"><strong>indel-stats</strong></dt>
<dd>
<p>calculates per-sample or de novo indels stats. The usage and format is similar
to <strong>smpl-stats</strong> and <strong>trio-stats</strong></p>
</dd>
<dt class="hdlist1"><strong>isecGT</strong></dt>
<dd>
<p>compare two files and set non-identical genotypes to missing</p>
</dd>
<dt class="hdlist1"><strong>mendelian</strong></dt>
<dd>
<p>count Mendelian consistent / inconsistent genotypes (deprecated, use +mendelian2 instead)</p>
</dd>
<dt class="hdlist1"><strong>mendelian2</strong></dt>
<dd>
<p>count Mendelian consistent / inconsistent genotypes.</p>
</dd>
<dt class="hdlist1"><strong>missing2ref</strong></dt>
<dd>
<p>sets missing genotypes ("./.") to ref allele ("0/0" or "0|0")</p>
</dd>
<dt class="hdlist1"><strong>parental-origin</strong></dt>
<dd>
<p>determine parental origin of a CNV region</p>
</dd>
<dt class="hdlist1"><strong>prune</strong></dt>
<dd>
<p>prune sites by missingness, allele frequency or linkage disequilibrium.
Alternatively, annotate sites with r2, Lewontin&#8217;s D' (PMID:19433632), Ragsdale&#8217;s D (PMID:31697386).</p>
</dd>
<dt class="hdlist1"><strong>remove-overlaps</strong></dt>
<dd>
<p>remove overlapping variants and duplicate sites</p>
</dd>
<dt class="hdlist1"><strong>scatter</strong></dt>
<dd>
<p>intended as an inverse to <code>bcftools concat</code>, scatter VCF by chunks or regions, creating multiple VCFs.</p>
</dd>
<dt class="hdlist1"><strong>setGT</strong></dt>
<dd>
<p>general tool to set genotypes according to rules requested by the user</p>
</dd>
<dt class="hdlist1"><strong>smpl-stats</strong></dt>
<dd>
<p>calculates basic per-sample stats. The usage and format is similar to
<strong>indel-stats</strong> and <strong>trio-stats</strong>.</p>
</dd>
<dt class="hdlist1"><strong>split</strong></dt>
<dd>
<p>split VCF by sample, creating single- or multi-sample VCFs</p>
</dd>
<dt class="hdlist1"><strong>split-vep</strong></dt>
<dd>
<p>extract fields from structured annotations such as INFO/CSQ created by VEP or INFO/BCSQ created by
bcftools/csq. These can be added as a new INFO field to the VCF or in a custom text format. See
<a href="http://samtools.github.io/bcftools/howtos/plugin.split-vep.html" class="bare">http://samtools.github.io/bcftools/howtos/plugin.split-vep.html</a> for more.</p>
</dd>
<dt class="hdlist1"><strong>tag2tag</strong></dt>
<dd>
<p>Convert between similar tags, such as GL,PL,GP or QR,QA,QS or tags with localized alleles e.g. LPL,LAD.
See <a href="http://samtools.github.io/bcftools/howtos/plugin.tag2tag.html" class="bare">http://samtools.github.io/bcftools/howtos/plugin.tag2tag.html</a> for more.</p>
</dd>
<dt class="hdlist1"><strong>trio-dnm2</strong></dt>
<dd>
<p>screen variants for possible de-novo mutations in trios</p>
</dd>
<dt class="hdlist1"><strong>trio-stats</strong></dt>
<dd>
<p>calculate transmission rate in trio children. The usage and format is similar to
<strong>indel-stats</strong> and <strong>smpl-stats</strong>.</p>
</dd>
<dt class="hdlist1"><strong>trio-switch-rate</strong></dt>
<dd>
<p>calculate phase switch rate in trio samples, children samples must have phased GTs</p>
</dd>
<dt class="hdlist1"><strong>variant-distance</strong></dt>
<dd>
<p>annotate sites with the distance to the nearest variant</p>
</dd>
<dt class="hdlist1"><strong>variantkey-hex</strong></dt>
<dd>
<p>generate unsorted VariantKey-RSid index files in hexadecimal format</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_examples_3">Examples:</h4>
<div class="listingblock">
<div class="content">
<pre># List options common to all plugins
bcftools plugin

# List available plugins
bcftools plugin -l

# Run a plugin
bcftools plugin counts in.vcf

# Run a plugin using the abbreviated "+" notation
bcftools +counts in.vcf

# Run a plugin from an explicit location
bcftools +/path/to/counts.so in.vcf

# The input VCF can be streamed just like in other commands
cat in.vcf | bcftools +counts

# Print usage information of plugin "dosage"
bcftools +dosage -h

# Replace missing genotypes with 0/0
bcftools +missing2ref in.vcf

# Replace missing genotypes with 0|0
bcftools +missing2ref in.vcf -- -p</pre>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_plugins_troubleshooting">Plugins troubleshooting:</h4>
<div class="paragraph">
<p>Things to check if your plugin does not show up in the <strong>bcftools plugin -l</strong> output:</p>
</div>
<div class="ulist">
<ul>
<li>
<p>Run with the <strong>-v</strong> option for verbose output: <strong>bcftools plugin -lv</strong></p>
</li>
<li>
<p>Does the environment variable BCFTOOLS_PLUGINS include the correct path?</p>
</li>
</ul>
</div>
</div>
<div class="sect3">
<h4 id="_plugins_api">Plugins API:</h4>
<div class="listingblock">
<div class="content">
<pre>// Short description used by 'bcftools plugin -l'
const char *about(void);

// Longer description used by 'bcftools +name -h'
const char *usage(void);

// Called once at startup, allows initialization of local variables.
// Return 1 to suppress normal VCF/BCF header output, -1 on critical
// errors, 0 otherwise.
int init(int argc, char **argv, bcf_hdr_t *in_hdr, bcf_hdr_t *out_hdr);

// Called for each VCF record, return NULL to suppress the output
bcf1_t *process(bcf1_t *rec);

// Called after all lines have been processed to clean up
void destroy(void);</pre>
</div>
</div>
</div>
</div>
<div class="sect2">
<h3 id="polysomy">bcftools polysomy [<em>OPTIONS</em>] <em>file.vcf.gz</em></h3>
<div class="paragraph">
<p>Detect number of chromosomal copies in VCFs annotates with the Illumina&#8217;s
B-allele frequency (BAF) values. Note that this command is not compiled
in by default, see the section <strong>Optional Compilation with GSL</strong> in the INSTALL
file for help.</p>
</div>
<div class="sect3">
<h4 id="_general_options_2">General options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-o, --output-dir</strong> <em>path</em></dt>
<dd>
<p>output directory</p>
</dd>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-s, --sample</strong> <em>string</em></dt>
<dd>
<p>sample name</p>
</dd>
<dt class="hdlist1"><strong>-t, --targets</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-T, --targets-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--targets-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-v, --verbose</strong></dt>
<dd>
<p>verbose debugging output which gives hints about the thresholds and decisions made
by the program. Note that the exact output can change between versions.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_algorithm_options">Algorithm options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-b, --peak-size</strong> <em>float</em></dt>
<dd>
<p>the minimum peak size considered as a good match can be from the interval [0,1]
where larger is stricter</p>
</dd>
<dt class="hdlist1"><strong>-c, --cn-penalty</strong> <em>float</em></dt>
<dd>
<p>a penalty for increasing copy number state. How this works: multiple peaks
are always a better fit than a single peak, therefore the program prefers
a single peak (normal copy number) unless the absolute deviation of the
multiple peaks fit is significantly smaller. Here the meaning of
"significant" is given by the <em>float</em> from the interval [0,1] where
larger is stricter.</p>
</dd>
<dt class="hdlist1"><strong>-f, --fit-th</strong> <em>float</em></dt>
<dd>
<p>threshold for goodness of fit (normalized absolute deviation), smaller is stricter</p>
</dd>
<dt class="hdlist1"><strong>-i, --include-aa</strong></dt>
<dd>
<p>include also the AA peak in CN2 and CN3 evaluation. This usually requires increasing <strong>-f</strong>.</p>
</dd>
<dt class="hdlist1"><strong>-m, --min-fraction</strong> <em>float</em></dt>
<dd>
<p>minimum distinguishable fraction of aberrant cells. The experience shows that trustworthy
are estimates of 20% and more.</p>
</dd>
<dt class="hdlist1"><strong>-p, --peak-symmetry</strong> <em>float</em></dt>
<dd>
<p>a heuristics to filter failed fits where the expected peak symmetry is violated.
The <em>float</em> is from the interval [0,1] and larger is stricter</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="query">bcftools query [<em>OPTIONS</em>] <em>file.vcf.gz</em> [<em>file.vcf.gz</em> [&#8230;&#8203;]]</h3>
<div class="paragraph">
<p>Extracts fields from VCF or BCF files and outputs them in user-defined format.</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-e, --exclude</strong> <em>EXPRESSION</em></dt>
<dd>
<p>exclude sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>--force-samples</strong></dt>
<dd>
<p>continue even when some samples requested via <strong>-s/-S</strong> do not exist</p>
</dd>
<dt class="hdlist1"><strong>-f, --format</strong> <em>FORMAT</em></dt>
<dd>
<p>learn by example, see below</p>
</dd>
<dt class="hdlist1"><strong>-F, --print-filtered</strong> <em>STR</em></dt>
<dd>
<p>by default, samples failing <strong>-i/-e</strong> filtering expressions are suppressed from output
when FORMAT fields are queried (for example <em>%CHROM %POS [ %GT]</em>).  With <strong>-F</strong>, such
fields will be still printed but instead of their actual value, <em>STR</em> will be used.</p>
</dd>
<dt class="hdlist1"><strong>-H, --print-header</strong></dt>
<dd>
<p>print header. By default, the header is printed with column indices, e.g. "#[1]CHROM".
These can be suppressed by giving the option twice, "<code>-HH</code>".</p>
</dd>
<dt class="hdlist1"><strong>-i, --include</strong> <em>EXPRESSION</em></dt>
<dd>
<p>include only sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-l, --list-samples</strong></dt>
<dd>
<p>list sample names and exit</p>
</dd>
<dt class="hdlist1"><strong>-N, --disable-automatic-newline</strong></dt>
<dd>
<p>disable automatic addition of a missing newline character at the end of the formatting
expression. By default, the program checks if the expression contains a newline
and appends it if not, to prevent formatting the entire output into a single
line by mistake. Note that versions prior to 1.18 had no automatic check and newline
had to be included explicitly.</p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-s, --samples</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-S, --samples-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-t, --targets</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-T, --targets-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--targets-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-u, --allow-undef-tags</strong></dt>
<dd>
<p>do not throw an error if there are undefined tags in the format string,
print "." instead</p>
</dd>
<dt class="hdlist1"><strong>-v, --vcf-list</strong> <em>FILE</em></dt>
<dd>
<p>process multiple VCFs listed in the file</p>
</dd>
</dl>
</div>
<div class="sect3">
<h4 id="_format">Format:</h4>
<div class="literalblock">
<div class="content">
<pre>%CHROM          The CHROM column (similarly also other columns: POS, ID, REF, ALT, QUAL, FILTER)
%END            End position of the REF allele
%END0           End position of the REF allele in 0-based coordinates
%FIRST_ALT      Alias for %ALT{0}
%FORMAT         Prints all FORMAT fields or a subset of samples with -s or -S
%GT             Genotype (e.g. 0/1)
%INFO           Prints the whole INFO column
%INFO/TAG       Any tag in the INFO column
%IUPACGT        Genotype translated to IUPAC ambiguity codes (e.g. M instead of C/A)
%LINE           Prints the whole line
%MASK           Indicates presence of the site in other files (with multiple files)
%N_PASS(expr)   Number of samples that pass the filtering expression (see *&lt;&lt;expressions,EXPRESSIONS&gt;&gt;*)
%POS0           POS in 0-based coordinates
%PBINOM(TAG)    Calculate phred-scaled binomial probability, the allele index is determined from GT
%SAMPLE         Sample name
%TAG{INT}       Curly brackets to print a subfield (e.g. INFO/TAG{1}, the indexes are 0-based)
%TBCSQ          Translated FORMAT/BCSQ. See the csq command above for explanation and examples.
%TGT            Translated genotype (e.g. C/A)
%TYPE           Variant type (REF, SNP, MNP, INDEL, BND, OTHER)
%VKX            VariantKey, biallelic hexadecimal encoding of CHROM,POS,REF,ALT (https://github.com/tecnickcom/variantkey)
[]              Format fields must be enclosed in brackets to loop over all samples
\n              new line
\t              tab character</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>Everything else is printed verbatim.</pre>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_examples_4">Examples:</h4>
<div class="literalblock">
<div class="content">
<pre># Print chromosome, position, ref allele and the first alternate allele
bcftools query -f '%CHROM  %POS  %REF  %ALT{0}\n' file.vcf.gz</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre># Similar to above, but use tabs instead of spaces, add sample name and genotype
bcftools query -f '%CHROM\t%POS\t%REF\t%ALT[\t%SAMPLE=%GT]\n' file.vcf.gz</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre># Print FORMAT/GT fields followed by FORMAT/GT fields
bcftools query -f 'GQ:[ %GQ] \t GT:[ %GT]\n' file.vcf</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre># Make a BED file: chr, pos (0-based), end pos (1-based), id
bcftools query -f'%CHROM\t%POS0\t%END\t%ID\n' file.bcf</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre># Print only samples with alternate (non-reference) genotypes
bcftools query -f'[%CHROM:%POS %SAMPLE %GT\n]' -i'GT="alt"' file.bcf</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre># Print all samples at sites with at least one alternate genotype
bcftools view -i'GT="alt"' file.bcf -Ou | bcftools query -f'[%CHROM:%POS %SAMPLE %GT\n]'</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre># Print phred-scaled binomial probability from FORMAT/AD tag for all heterozygous genotypes
bcftools query -i'GT="het"' -f'[%CHROM:%POS %SAMPLE %GT %PBINOM(AD)\n]' file.vcf</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre># Print the second value of AC field if bigger than 10. Note the (unfortunate) difference in
# index subscript notation: formatting expressions (-f) uses "{}" while filtering expressions
# (-i) use "[]". This is for historic reasons and backward-compatibility.
bcftools query -f '%AC{1}\n' -i 'AC[1]&gt;10' file.vcf.gz</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre># Print all samples at sites where at least one sample has DP=1 or DP=2. In the second case
# print only samples with DP=1 or DP=2, the difference is in the logical operator used, || vs |.
bcftools query -f '[%SAMPLE %GT %DP\n]' -i 'FMT/DP=1 || FMT/DP=2' file.vcf
bcftools query -f '[%SAMPLE %GT %DP\n]' -i 'FMT/DP=1 |  FMT/DP=2' file.vcf</pre>
</div>
</div>
</div>
</div>
<div class="sect2">
<h3 id="reheader">bcftools reheader [<em>OPTIONS</em>] <em>file.vcf.gz</em></h3>
<div class="paragraph">
<p>Modify header of VCF/BCF files, change sample names.</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-f, --fai</strong> <em>FILE</em></dt>
<dd>
<p>add to the header contig names and their lengths from the provided fasta index file (.fai).
Lengths of existing contig lines will be updated and contig lines not present in
the fai file will be removed</p>
</dd>
<dt class="hdlist1"><strong>-h, --header</strong> <em>FILE</em></dt>
<dd>
<p>new VCF header</p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-s, --samples</strong> <em>FILE</em></dt>
<dd>
<p>new sample names, one name per line, in the same order as they appear
in the VCF file. Alternatively, only samples which need to be renamed
can be listed as "old_name new_name\n" pairs separated by whitespaces,
each on a separate line. If a sample name contains spaces, the
spaces can be escaped using the backslash character, for example
"Not\ a\ good\ sample\ name".</p>
</dd>
<dt class="hdlist1"><strong>-T, --temp-prefix</strong> <em>PATH</em></dt>
<dd>
<p>this option is ignored, but left for compatibility with earlier versions of bcftools.</p>
</dd>
<dt class="hdlist1"><strong>--threads</strong> <em>INT</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="roh">bcftools roh [<em>OPTIONS</em>] <em>file.vcf.gz</em></h3>
<div class="paragraph">
<p>A program for detecting runs of homo/autozygosity. Only bi-allelic sites
are considered.</p>
</div>
<div class="sect3">
<h4 id="_the_hmm_model">The HMM model:</h4>
<div class="listingblock">
<div class="content">
<pre>Notation:
  D  = Data, AZ = autozygosity, HW = Hardy-Weinberg (non-autozygosity),
  f  = non-ref allele frequency

Emission probabilities:
  oAZ = P_i(D|AZ) = (1-f)*P(D|RR) + f*P(D|AA)
  oHW = P_i(D|HW) = (1-f)^2 * P(D|RR) + f^2 * P(D|AA) + 2*f*(1-f)*P(D|RA)

Transition probabilities:
  tAZ = P(AZ|HW)  .. from HW to AZ, the -a parameter
  tHW = P(HW|AZ)  .. from AZ to HW, the -H parameter

  ci  = P_i(C)  .. probability of cross-over at site i, from genetic map
  AZi = P_i(AZ) .. probability of site i being AZ/non-AZ, scaled so that AZi+HWi = 1
  HWi = P_i(HW)

  P_{i+1}(AZ) = oAZ * max[(1 - tAZ * ci) * AZ{i-1} , tAZ * ci * (1-AZ{i-1})]
  P_{i+1}(HW) = oHW * max[(1 - tHW * ci) * (1-AZ{i-1}) , tHW * ci * AZ{i-1}]</pre>
</div>
</div>
</div>
<div class="sect3">
<h4 id="_general_options_3">General Options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--AF-dflt</strong> <em>FLOAT</em></dt>
<dd>
<p>in case allele frequency is not known, use the <em>FLOAT</em>. By default, sites where
allele frequency cannot be determined, or is 0, are skipped.</p>
</dd>
<dt class="hdlist1"><strong>--AF-tag</strong> <em>TAG</em></dt>
<dd>
<p>use the specified INFO tag <em>TAG</em> as an allele frequency estimate
instead of the default AC and AN tags. Sites which do not have <em>TAG</em>
will be skipped.</p>
</dd>
<dt class="hdlist1"><strong>--AF-file</strong> <em>FILE</em></dt>
<dd>
<p>Read allele frequencies from a tab-delimited file containing
the columns: CHROM\tPOS\tREF,ALT\tAF. The file can be compressed with
<strong>bgzip</strong> and indexed with tabix -s1 -b2 -e2.  Sites which are not present in
the <em>FILE</em> or have different reference or alternate allele will be skipped.
Note that such a file can be easily created from a VCF using:</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    bcftools query -f'%CHROM\t%POS\t%REF,%ALT\t%INFO/TAG\n' file.vcf | bgzip -c &gt; freqs.tab.gz</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-b, --buffer-size</strong> <em>INT</em>[,<em>INT</em>]</dt>
<dd>
<p>when the entire many-sample file cannot fit into memory, a sliding
buffer approach can be used. The first value is the number of sites
to keep in memory. If negative, it is interpreted as the maximum
memory to use, in MB. The second, optional, value sets the number
of overlapping sites. The default overlap is set to roughly 1% of
the buffer size.</p>
</dd>
<dt class="hdlist1"><strong>-e, --estimate-AF</strong> <em class="'TAG'">FILE</em></dt>
<dd>
<p>estimate the allele frequency by recalculating INFO/AC and INFO/AN on
the fly, using the specified <em>TAG</em> which can be either FORMAT/GT ("GT")
or FORMAT/PL ("PL"). If <em>TAG</em> is not given, "GT" is assumed.  Either all
samples ("-") or samples listed in <em>FILE</em> will be included. For example,
use "PL,-" to estimate AF from FORMAT/PL of all samples.
If neither <strong>-e</strong> nor the other <strong>--AF-&#8230;&#8203;</strong> options are given, the allele frequency is
estimated from AC and AN counts which are already present in the INFO field.</p>
</dd>
<dt class="hdlist1"><strong>--exclude</strong> <em>EXPRESSION</em></dt>
<dd>
<p>exclude sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-G, --GTs-only</strong> <em>FLOAT</em></dt>
<dd>
<p>use genotypes (FORMAT/GT fields) ignoring genotype likelihoods (FORMAT/PL),
setting PL of unseen genotypes to <em>FLOAT</em>. Safe value to use is 30 to
account for GT errors.</p>
</dd>
<dt class="hdlist1"><strong>--include</strong> <em>EXPRESSION</em></dt>
<dd>
<p>include only sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-I, --skip-indels</strong></dt>
<dd>
<p>skip indels as their genotypes are usually enriched for errors</p>
</dd>
<dt class="hdlist1"><strong>-m, --genetic-map</strong> <em>FILE</em></dt>
<dd>
<p>genetic map in the format required also by IMPUTE2. Only the first and
third column are used (position and Genetic_Map(cM)). The <em>FILE</em> can
be a single file or a file mask, where string "{CHROM}" is replaced with
chromosome name.</p>
</dd>
<dt class="hdlist1"><strong>-M, --rec-rate</strong> <em>FLOAT</em></dt>
<dd>
<p>constant recombination rate per bp. In combination with <strong>--genetic-map</strong>,
the <strong>--rec-rate</strong> parameter is interpreted differently, as <em>FLOAT</em>-fold increase of
transition probabilities, which allows the model to become more sensitive
yet still account for recombination hotspots. Note that also the range
of the values is therefore different in both cases: normally the
parameter will be in the range (1e-3,1e-9) but with <strong>--genetic-map</strong>
it will be in the range (10,1000).</p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>Write output to the <em>FILE</em>, by default the output is printed on stdout</p>
</dd>
<dt class="hdlist1"><strong>-O, --output-type</strong> <em>s</em>|<em>r</em>[<em>z</em>]</dt>
<dd>
<p>Generate per-site output (<em>s</em>) or per-region output (<em>r</em>). By default
both types are printed and the output is uncompressed. Add <em>z</em> for
a compressed output.</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    # Output fields:
    RG = predicted homo/autozygous regions
        - Sample
        - Chromosome
        - Start
        - End
        - Length (bp)
        - Number of markers
        - Quality .. average phred score in the region from the forward-backward algorithm

    ST = per-site output showing:
        - Sample
        - Chromosome
        - Position
        - State .. predicted state from the Viterbi algorithm, 0 for normal (HW, Hardy-Weinberg) or 1 for autozygous (AZ)
        - Quality .. quality score from the forward-backward algorithm</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-s, --samples</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-S, --samples-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-t, --targets</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-T, --targets-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--targets-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_hmm_options_2">HMM Options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-a, --hw-to-az</strong> <em>FLOAT</em></dt>
<dd>
<p>P(AZ|HW) transition probability from AZ (autozygous) to HW (Hardy-Weinberg) state</p>
</dd>
<dt class="hdlist1"><strong>-H, --az-to-hw</strong> <em>FLOAT</em></dt>
<dd>
<p>P(HW|AZ) transition probability from HW to AZ state</p>
</dd>
<dt class="hdlist1"><strong>-V, --viterbi-training</strong> <em>FLOAT</em></dt>
<dd>
<p>estimate HMM parameters using Baum-Welch algorithm, using the convergence threshold
<em>FLOAT</em>, e.g. 1e-10 (experimental)</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="sort">bcftools sort [<em>OPTIONS</em>] file.bcf</h3>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-m, --max-mem</strong> <em>FLOAT</em>[<em>kMG</em>]</dt>
<dd>
<p>Maximum memory to use. Approximate, affects the number of temporary files written
to the disk. Note that if the command fails at this step because of too many open files,
your system limit on the number of open files ("ulimit") may need to be increased.</p>
</dd>
<dt class="hdlist1"><strong>-o, --output</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-O, --output-type</strong> <em>b</em>|<em>u</em>|<em>z</em>|<em>v</em>[0-9]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-T, --temp-dir</strong> <em>DIR</em></dt>
<dd>
<p>Use this directory to store temporary files. If the last six characters of the string DIR are XXXXXX,
then these are replaced with a string that makes the directory name unique.</p>
</dd>
<dt class="hdlist1"><strong>-W</strong>[<em>FMT</em>]<strong>, -W</strong>[=<em>FMT</em>]<strong>, --write-index</strong>[=<em>FMT</em>]</dt>
<dd>
<p>Automatically index the output file.  <em>FMT</em> is optional and can be
one of "tbi" or "csi" depending on output file format.</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="stats">bcftools stats [<em>OPTIONS</em>] <em>A.vcf.gz</em> [<em>B.vcf.gz</em>]</h3>
<div class="paragraph">
<p>Parses VCF or BCF and produces text file stats which is suitable for machine
processing and can be plotted using <strong><a href="#plot-vcfstats">plot-vcfstats</a></strong>.  When two files are given,
the program generates separate stats for intersection and the complements. By
default only sites are compared, <strong>-s</strong>/<strong>-S</strong> must given to include also sample
columns.
When one VCF file is specified on the command line, then stats by non-reference allele
frequency, depth distribution, stats by quality and per-sample counts, singleton stats,
etc. are printed.
When two VCF files are given, then stats such as concordance (Genotype concordance by
non-reference allele frequency, Genotype concordance by sample, Non-Reference Discordance)
and correlation are also printed. Per-site discordance (PSD) is also printed in <strong>--verbose</strong> mode.</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>--af-bins</strong> <em>LIST</em>|<em>FILE</em></dt>
<dd>
<p>comma separated list of allele frequency bins (e.g. 0.1,0.5,1)
or a file listing the allele frequency bins one per line (e.g. 0.1\n0.5\n1)</p>
</dd>
<dt class="hdlist1"><strong>--af-tag</strong> <em>TAG</em></dt>
<dd>
<p>allele frequency INFO tag to use for binning. By default the allele frequency is
estimated from AC/AN, if available, or directly from the genotypes (GT) if not.</p>
</dd>
<dt class="hdlist1"><strong>-1, --1st-allele-only</strong></dt>
<dd>
<p>consider only the 1st alternate allele at multiallelic sites</p>
</dd>
<dt class="hdlist1"><strong>-c, --collapse</strong> <em>snps</em>|<em>indels</em>|<em>both</em>|<em>all</em>|<em>some</em>|<em>none</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-d, --depth</strong> <em>INT</em>,<em>INT</em>,<em>INT</em></dt>
<dd>
<p>ranges of depth distribution: min, max, and size of the bin</p>
</dd>
<dt class="hdlist1"><strong>--debug</strong></dt>
<dd>
<p>produce verbose per-site and per-sample output</p>
</dd>
<dt class="hdlist1"><strong>-e, --exclude</strong> <em>EXPRESSION</em></dt>
<dd>
<p>exclude sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-E, --exons</strong> <em>file.gz</em></dt>
<dd>
<p>tab-delimited file with exons for indel frameshifts statistics. The columns
of the file are CHR, FROM, TO, with 1-based, inclusive, positions. The file
is BGZF-compressed and indexed with tabix</p>
</dd>
</dl>
</div>
<div class="listingblock">
<div class="content">
<pre>    tabix -s1 -b2 -e3 file.gz</pre>
</div>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-f, --apply-filters</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-F, --fasta-ref</strong> <em>ref.fa</em></dt>
<dd>
<p>faidx indexed reference sequence file to determine INDEL context</p>
</dd>
<dt class="hdlist1"><strong>-i, --include</strong> <em>EXPRESSION</em></dt>
<dd>
<p>include only sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-I, --split-by-ID</strong></dt>
<dd>
<p>collect stats separately for sites which have the ID column set ("known
sites") or which do not have the ID column set ("novel sites").</p>
</dd>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-s, --samples</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-S, --samples-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-t, --targets</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-T, --targets-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--targets-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-u, --user-tstv</strong> <em>&lt;TAG[:min:max:n]&gt;</em></dt>
<dd>
<p>collect Ts/Tv stats for any tag using the given binning [0:1:100]</p>
</dd>
<dt class="hdlist1"><strong>-v, --verbose</strong></dt>
<dd>
<p>produce verbose per-site and per-sample output</p>
</dd>
</dl>
</div>
</div>
<div class="sect2">
<h3 id="view">bcftools view [<em>OPTIONS</em>] <em>file.vcf.gz</em> [<em>REGION</em> [&#8230;&#8203;]]</h3>
<div class="paragraph">
<p>View, subset and filter VCF or BCF files by position and filtering expression.
Convert between VCF and BCF. Former <strong>bcftools subset</strong>.</p>
</div>
<div class="sect3">
<h4 id="_output_options_2">Output options</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-G, --drop-genotypes</strong></dt>
<dd>
<p>drop individual genotype information (after subsetting if <strong>-s</strong> option is set)</p>
</dd>
<dt class="hdlist1"><strong>-h, --header-only</strong></dt>
<dd>
<p>output the VCF header only (see also <strong><a href="#head">bcftools head</a></strong>)</p>
</dd>
<dt class="hdlist1"><strong>-H, --no-header</strong></dt>
<dd>
<p>suppress the header in VCF output</p>
</dd>
<dt class="hdlist1"><strong>--with-header</strong></dt>
<dd>
<p>output both VCF header and records (this is the default, but the option
is useful for explicitness or to reset the effects of <strong>-h</strong> or <strong>-H</strong>)</p>
</dd>
<dt class="hdlist1"><strong>-l, --compression-level</strong> [<em>0-9</em>]</dt>
<dd>
<p>compression level. 0 stands for uncompressed, 1 for best speed and 9 for
best compression.</p>
</dd>
<dt class="hdlist1"><strong>--no-version</strong></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-O, --output-type</strong> <em>b</em>|<em>u</em>|<em>z</em>|<em>v</em>[0-9]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
</dl>
</div>
<div class="paragraph">
<p><strong>-o, --output</strong> <em>FILE</em>:
    output file name. If not present, the default is to print to standard output (stdout).</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-r, --regions</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-R, --regions-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--regions-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-t, --targets</strong> <em>chr</em>|<em>chr:pos</em>|<em>chr:from-to</em>|<em>chr:from-</em>[,&#8230;&#8203;]</dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-T, --targets-file</strong> <em>file</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--targets-overlap</strong> <em>0</em>|<em>1</em>|<em>2</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>--threads</strong> <em>INT</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-W</strong>[<em>FMT</em>]<strong>, -W</strong>[=<em>FMT</em>]<strong>, --write-index</strong>[=<em>FMT</em>]</dt>
<dd>
<p>Automatically index the output file.  <em>FMT</em> is optional and can be
one of "tbi" or "csi" depending on output file format.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_subset_options">Subset options:</h4>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-A, --trim-unseen-alleles</strong></dt>
<dd>
<p>remove the unseen allele <em>&lt;*&gt;</em> or <em>&lt;NON_REF&gt;</em> at variant sites when the option is given once (-A) or
at all sites when the options is given twice (<em>-AA</em>).</p>
</dd>
<dt class="hdlist1"><strong>-a, --trim-alt-alleles</strong></dt>
<dd>
<p>remove alleles not seen in the genotype fields from the ALT column. Note that if no alternate allele
remains after trimming, the record itself is not removed but ALT is set to ".".
If the option <strong>-s</strong> or <strong>-S</strong> is given, removes alleles not seen in the subset.
INFO and FORMAT tags declared as Type=A, G or R will be trimmed as well.</p>
</dd>
<dt class="hdlist1"><strong>--force-samples</strong></dt>
<dd>
<p>only warn about unknown subset samples</p>
</dd>
<dt class="hdlist1"><strong>-I, --no-update</strong></dt>
<dd>
<p>do not (re)calculate INFO fields for the subset (currently INFO/AC and INFO/AN)</p>
</dd>
<dt class="hdlist1"><strong>-s, --samples</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong>. Note that it is possible to create
multiple subsets simultaneously using the <strong>split</strong> plugin.</p>
</dd>
<dt class="hdlist1"><strong>-S, --samples-file</strong> <em>FILE</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong>. Note that it is possible to create
multiple subsets simultaneously using the <strong>split</strong> plugin.</p>
</dd>
</dl>
</div>
</div>
<div class="sect3">
<h4 id="_filter_options">Filter options:</h4>
<div class="paragraph">
<p>Note that filter options below dealing with counting the number of alleles
will, for speed, first check for the values of AC and AN in the INFO column to
avoid parsing all the genotype (FORMAT/GT) fields in the VCF. This means
that filters like <em>--uncalled</em>, --exclude-uncalled', or <em>--min-af 0.1</em> will be calculated from INFO/AC and
INFO/AN when available or FORMAT/GT otherwise. However, it will not attempt to use any other existing field,
like INFO/AF for example. For that, use <em>--exclude AF&lt;0.1</em> instead.</p>
</div>
<div class="paragraph">
<p>Also note that one must be careful when sample subsetting and filtering is performed in a single command
because the order of internal operations can influence the result. For example, the <strong>-i/-e</strong> filtering
is performed before sample removal, but the <strong>-P</strong> filtering is performed after,
and some are inherently ambiguous, for example allele counts can be taken from the INFO
column when present but calculated on the fly when absent. Therefore it is strongly recommended to spell out the
required order explicitly by separating such commands into two steps. (Make sure to use the <strong>-O u</strong> option
when piping!)</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-c, --min-ac</strong> <em>INT</em>[<em>:nref</em>|<em>:alt1</em>|<em>:minor</em>|<em>:major</em>|:'nonmajor']</dt>
<dd>
<p>minimum allele count (INFO/AC) of sites to be printed.
Specifying the type of allele is optional and can be set to
non-reference (<em>nref</em>, the default), 1st alternate  (<em>alt1</em>), the least
frequent (<em>minor</em>), the most frequent (<em>major</em>) or sum of all but the
most frequent (<em>nonmajor</em>) alleles.</p>
</dd>
<dt class="hdlist1"><strong>-C, --max-ac</strong> <em>INT</em>[<em>:nref</em>|<em>:alt1</em>|<em>:minor</em>|:'major'|:'nonmajor']</dt>
<dd>
<p>maximum allele count (INFO/AC) of sites to be printed.
Specifying the type of allele is optional and can be set to
non-reference (<em>nref</em>, the default), 1st alternate  (<em>alt1</em>), the least
frequent (<em>minor</em>), the most frequent (<em>major</em>) or sum of all but the
most frequent (<em>nonmajor</em>) alleles.</p>
</dd>
<dt class="hdlist1"><strong>-e, --exclude</strong> <em>EXPRESSION</em></dt>
<dd>
<p>exclude sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-f, --apply-filters</strong> <em>LIST</em></dt>
<dd>
<p>see <strong><a href="#common_options">Common Options</a></strong></p>
</dd>
<dt class="hdlist1"><strong>-g, --genotype</strong> [^][<em>hom</em>|<em>het</em>|<em>miss</em>]</dt>
<dd>
<p>include only sites with one or more homozygous (<em>hom</em>), heterozygous
(<em>het</em>) or missing (<em>miss</em>) genotypes. When prefixed with <em>^</em>, the logic
is reversed; thus <em>^het</em> excludes sites with heterozygous genotypes.</p>
</dd>
<dt class="hdlist1"><strong>-i, --include</strong> <em>EXPRESSION</em></dt>
<dd>
<p>include sites for which <em>EXPRESSION</em> is true. For valid expressions see
<strong><a href="#expressions">EXPRESSIONS</a></strong>.</p>
</dd>
<dt class="hdlist1"><strong>-k, --known</strong></dt>
<dd>
<p>print known sites only (ID column is not ".")</p>
</dd>
<dt class="hdlist1"><strong>-m, --min-alleles</strong> <em>INT</em></dt>
<dd>
<p>print sites with at least <em>INT</em> alleles listed in REF and ALT columns</p>
</dd>
<dt class="hdlist1"><strong>-M, --max-alleles</strong> <em>INT</em></dt>
<dd>
<p>print sites with at most <em>INT</em> alleles listed in REF and ALT columns.
Use <strong>-m2 -M2 -v snps</strong> to only view biallelic SNPs.</p>
</dd>
<dt class="hdlist1"><strong>-n, --novel</strong></dt>
<dd>
<p>print novel sites only (ID column is ".")</p>
</dd>
<dt class="hdlist1"><strong>-p, --phased</strong></dt>
<dd>
<p>print sites where all samples are phased. Haploid genotypes are
considered phased. Missing genotypes considered unphased unless the
phased bit is set.</p>
</dd>
<dt class="hdlist1"><strong>-P, --exclude-phased</strong></dt>
<dd>
<p>exclude sites where all samples are phased</p>
</dd>
<dt class="hdlist1"><strong>-q, --min-af</strong> <em>FLOAT</em>[<em>:nref</em>|<em>:alt1</em>|<em>:minor</em>|<em>:major</em>|<em>:nonmajor</em>]</dt>
<dd>
<p>minimum allele frequency (INFO/AC / INFO/AN) of sites to be printed.
Specifying the type of allele is optional and can be set to
non-reference (<em>nref</em>, the default), 1st alternate  (<em>alt1</em>), the least
frequent (<em>minor</em>), the most frequent (<em>major</em>) or sum of all but the
most frequent (<em>nonmajor</em>) alleles.</p>
</dd>
<dt class="hdlist1"><strong>-Q, --max-af</strong> <em>FLOAT</em>[<em>:nref</em>|<em>:alt1</em>|<em>:minor</em>|<em>:major</em>|<em>:nonmajor</em>]</dt>
<dd>
<p>maximum allele frequency (INFO/AC / INFO/AN) of sites to be printed.
Specifying the type of allele is optional and can be set to
non-reference (<em>nref</em>, the default), 1st alternate  (<em>alt1</em>), the least
frequent (<em>minor</em>), the most frequent (<em>major</em>) or sum of all but the
most frequent (<em>nonmajor</em>) alleles.</p>
</dd>
<dt class="hdlist1"><strong>-u, --uncalled</strong></dt>
<dd>
<p>print sites without a called genotype, i.e. print sites with all genotypes
missing. Note that the missingness is determined from INFO/AN and AC tags
when available to avoid parsing sample fields.</p>
</dd>
<dt class="hdlist1"><strong>-U, --exclude-uncalled</strong></dt>
<dd>
<p>exclude sites without a called genotype, i.e. print sites with at least one
non-missing genotype. Note that the missingness is determined from INFO/AN
and AC tags when available to avoid parsing sample fields.</p>
</dd>
<dt class="hdlist1"><strong>-v, --types</strong> <em>snps</em>|<em>indels</em>|<em>mnps</em>|<em>other</em></dt>
<dd>
<p>comma-separated list of variant types to select. Site is selected if
any of the ALT alleles is of the type requested. Types are determined
by comparing the REF and ALT alleles in the VCF record not INFO tags
like INFO/INDEL or INFO/VT. Use <strong>--include</strong> to select based on INFO
tags.</p>
</dd>
<dt class="hdlist1"><strong>-V, --exclude-types</strong> <em>snps</em>|<em>indels</em>|<em>mnps</em>|<em>ref</em>|<em>bnd</em>|<em>other</em></dt>
<dd>
<p>comma-separated list of variant types to exclude. Site is excluded if
any of the ALT alleles is of the type requested. Types are determined
by comparing the REF and ALT alleles in the VCF record not INFO tags
like INFO/INDEL or INFO/VT. Use <strong>--exclude</strong> to exclude based on INFO tags.</p>
</dd>
<dt class="hdlist1"><strong>-x, --private</strong></dt>
<dd>
<p>print sites where only the subset samples carry an non-reference allele.
Requires <strong>--samples</strong> or <strong>--samples-file</strong>.</p>
</dd>
<dt class="hdlist1"><strong>-X, --exclude-private</strong></dt>
<dd>
<p>exclude sites where only the subset samples carry an non-reference allele</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="sect2">
<h3 id="help">bcftools help [<em>COMMAND</em>] | bcftools --help [<em>COMMAND</em>]</h3>
<div class="paragraph">
<p>Display  a  brief usage message listing the bcftools commands available.
If the name of a command is also given, e.g., bcftools help view, the detailed
usage message for that particular command is displayed.</p>
</div>
</div>
<div class="sect2">
<h3 id="version">bcftools [<em>--version</em>|<em>-v</em>]</h3>
<div class="paragraph">
<p>Display the version numbers and copyright information for bcftools and the
important libraries used by bcftools.</p>
</div>
</div>
<div class="sect2">
<h3 id="version-only">bcftools [<em>--version-only</em>]</h3>
<div class="paragraph">
<p>Display the full bcftools version number in a machine-readable format.</p>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_scripts">SCRIPTS</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="gff2gff">gff2gff</h3>
<div class="paragraph">
<p>Attempts to fix a GFF file to be correctly parsed by <strong><a href="#csq">csq</a></strong>.</p>
</div>
<div class="openblock">
<div class="content">
<div class="literalblock">
<div class="content">
<pre>zcat in.gff.gz | gff2gff | gzip -c &gt; out.gff.gz</pre>
</div>
</div>
</div>
</div>
</div>
<div class="sect2">
<h3 id="plot-vcfstats">plot-vcfstats [<em>OPTIONS</em>] <em>file.vchk</em> [&#8230;&#8203;]</h3>
<div class="paragraph">
<p>Script for processing output of <strong><a href="#stats">bcftools stats</a></strong>. It can merge
results from multiple outputs (useful when running the stats for each
chromosome separately), plots graphs and creates a PDF presentation.</p>
</div>
<div class="dlist">
<dl>
<dt class="hdlist1"><strong>-m, --merge</strong></dt>
<dd>
<p>Merge vcfstats files to STDOUT, skip plotting.</p>
</dd>
<dt class="hdlist1"><strong>-p, --prefix</strong> <em>DIR</em></dt>
<dd>
<p>The output directory. This directory will be created if it does not exist.</p>
</dd>
<dt class="hdlist1"><strong>-P, --no-PDF</strong></dt>
<dd>
<p>Skip the PDF creation step.</p>
</dd>
<dt class="hdlist1"><strong>-r, --rasterize</strong></dt>
<dd>
<p>Rasterize PDF images for faster rendering. This is the default and the opposite of <strong>-v, --vectors</strong>.</p>
</dd>
<dt class="hdlist1"><strong>-s, --sample-names</strong></dt>
<dd>
<p>Use sample names for xticks rather than numeric IDs.</p>
</dd>
<dt class="hdlist1"><strong>-t, --title</strong> <em>STRING</em></dt>
<dd>
<p>Identify files by these titles in plots. The option can be given multiple
times, for each ID in the <strong><a href="#stats">bcftools stats</a></strong> output. If not
present, the script will use abbreviated source file names for the titles.</p>
</dd>
<dt class="hdlist1"><strong>-v, --vectors</strong></dt>
<dd>
<p>Generate vector graphics for PDF images, the opposite of <strong>-r, --rasterize</strong>.</p>
</dd>
<dt class="hdlist1"><strong>-T, --main-title</strong> <em>STRING</em></dt>
<dd>
<p>Main title for the PDF.</p>
</dd>
</dl>
</div>
<div class="paragraph">
<p><strong>Example:</strong></p>
</div>
<div class="openblock">
<div class="content">
<div class="literalblock">
<div class="content">
<pre># Generate the stats
bcftools stats -s - &gt; file.vchk</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre># Plot the stats
plot-vcfstats -p outdir file.vchk</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre># The final looks can be customized by editing the generated
# 'outdir/plot.py' script and re-running manually
cd outdir &amp;&amp; python plot.py &amp;&amp; pdflatex summary.tex</pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="expressions">FILTERING EXPRESSIONS</h2>
<div class="sectionbody">
<div class="paragraph">
<p>These filtering expressions are accepted by most of the commands.</p>
</div>
<div class="ulist">
<div class="title">Valid expressions may contain:</div>
<ul>
<li>
<p>numerical constants, string constants, file names (indicated by the prefix <em>@</em>)</p>
<div class="literalblock">
<div class="content">
<pre>1, 1.0, 1e-4
"String"
@file_name</pre>
</div>
</div>
</li>
<li>
<p>arithmetic operators (addition, multiplication, subtraction, division, modulo)</p>
<div class="literalblock">
<div class="content">
<pre>+, *, -, /, %</pre>
</div>
</div>
</li>
<li>
<p>comparison operators</p>
<div class="literalblock">
<div class="content">
<pre>== (same as =), &gt;, &gt;=, &lt;=, &lt;, !=</pre>
</div>
</div>
</li>
<li>
<p>regex operators "\~" and its negation "!~". The expressions are case sensitive unless "/i" is added.</p>
<div class="literalblock">
<div class="content">
<pre>INFO/HAYSTACK ~ "needle"
INFO/HAYSTACK ~ "NEEDless/i"</pre>
</div>
</div>
</li>
<li>
<p>parentheses</p>
<div class="literalblock">
<div class="content">
<pre>(, )</pre>
</div>
</div>
</li>
<li>
<p>logical operators. See also the examples below and the <a href="http://samtools.github.io/bcftools/howtos/filtering.html">filtering tutorial</a>
about the distinction between "&amp;&amp;" vs "&amp;" and "||" vs "|".</p>
<div class="literalblock">
<div class="content">
<pre>&amp;&amp;,  &amp;, ||,  |</pre>
</div>
</div>
</li>
<li>
<p>INFO tags, FORMAT tags, column names</p>
<div class="literalblock">
<div class="content">
<pre>INFO/DP or DP
FORMAT/DV, FMT/DV, or DV
FILTER, QUAL, ID, CHROM, POS, REF, ALT[0]</pre>
</div>
</div>
</li>
<li>
<p>the FILTER column can be queried as follows:</p>
<div class="literalblock">
<div class="content">
<pre>FILTER="PASS"
FILTER="."
FILTER="A"          .. exact match, for example "A;B" does not pass
FILTER="A;B"        .. exact match, "A;B" and "B;A" pass, everything else fails
FILTER!="A"         .. exact match, for example "A;B" does pass
FILTER~"A"          .. subset match, for example both "A" and "A;B" pass
FILTER~"A;B"        .. subset match, pass only if both "A" and "B" are present
FILTER!~"A"         .. complement match, for example both "A" and "A;B" fail
FILTER!~"A;B"       .. complement match, fail if both "A" and "B" are present</pre>
</div>
</div>
</li>
<li>
<p>1 (or 0) to test the presence (or absence) of a flag</p>
<div class="literalblock">
<div class="content">
<pre>FlagA=1 &amp;&amp; FlagB=0</pre>
</div>
</div>
</li>
<li>
<p>"." to test missing values</p>
<div class="literalblock">
<div class="content">
<pre>DP=".", DP!=".", ALT="."</pre>
</div>
</div>
</li>
<li>
<p>missing genotypes can be matched regardless of phase and ploidy (".|.", "./.", ".", "0|.")
using these expressions</p>
<div class="literalblock">
<div class="content">
<pre>GT="mis", GT~"\.", GT!~"\."</pre>
</div>
</div>
</li>
<li>
<p>missing genotypes can be matched including the phase and ploidy (".|.", "./.", ".")
using these expressions</p>
<div class="literalblock">
<div class="content">
<pre>GT=".|.", GT="./.", GT="."</pre>
</div>
</div>
</li>
<li>
<p>sample genotype: reference (haploid or diploid), alternate (hom or het,
haploid or diploid), missing genotype, homozygous, heterozygous, haploid,
ref-ref hom, alt-alt hom, ref-alt het, alt-alt het, haploid ref, haploid alt
(case-insensitive)</p>
<div class="literalblock">
<div class="content">
<pre>GT="ref"
GT="alt"
GT="mis"
GT="hom"
GT="het"
GT="hap"
GT="RR"
GT="AA"
GT="RA" or GT="AR"
GT="Aa" or GT="aA"
GT="R"
GT="A"</pre>
</div>
</div>
</li>
<li>
<p>TYPE for variant type in REF,ALT columns (indel,snp,mnp,ref,bnd,other,overlap, see <strong><a href="#terminology">TERMINOLOGY</a></strong>). Use the regex
operator "\~" to require at least one allele of the given type or the equal sign "="
to require that all alleles are of the given type. Compare</p>
<div class="literalblock">
<div class="content">
<pre>TYPE="snp"
TYPE~"snp"
TYPE!="snp"
TYPE!~"snp"</pre>
</div>
</div>
</li>
<li>
<p>array subscripts (0-based), "*" for any element, "-" to indicate a range. Note that
for querying FORMAT vectors, the colon ":" can be used to select a sample and an
element of the vector, as shown in the examples below.</p>
<div class="literalblock">
<div class="content">
<pre>INFO/AF[0] &gt; 0.3             .. first AF value bigger than 0.3
FORMAT/AD[0:0] &gt; 30          .. first AD value of the first sample bigger than 30
FORMAT/AD[0:1]               .. first sample, second AD value
FORMAT/AD[1:0]               .. second sample, first AD value
DP4[*] == 0                  .. any DP4 value
FORMAT/DP[0]   &gt; 30          .. DP of the first sample bigger than 30
FORMAT/DP[1-3] &gt; 10          .. samples 2-4
FORMAT/DP[1-]  &lt; 7           .. all samples but the first
FORMAT/DP[0,2-4] &gt; 20        .. samples 1, 3-5
FORMAT/AD[0:1]               .. first sample, second AD field
FORMAT/AD[0:*], AD[0:] or AD[0] .. first sample, any AD field
FORMAT/AD[*:1] or AD[:1]        .. any sample, second AD field
(DP4[0]+DP4[1])/(DP4[2]+DP4[3]) &gt; 0.3
CSQ[*] ~ "missense_variant.*deleterious"</pre>
</div>
</div>
</li>
<li>
<p>in addition to array subscripts shown above, it is possible to subscript arrays
of Number=R tags by alleles found in FORMAT/GT (starting with version 1.17).
For example</p>
<div class="literalblock">
<div class="content">
<pre>FORMAT/AD[GT] &gt; 10        .. require support of more than 10 reads for each allele
FORMAT/AD[0:GT] &gt; 10      .. same as above, but in the first sample
sSUM(FORMAT/AD[GT]) &gt; 20  .. require total sample depth bigger than 20</pre>
</div>
</div>
</li>
<li>
<p>with many samples it can be more practical to provide a file with sample names,
one sample name per line</p>
<div class="literalblock">
<div class="content">
<pre>GT[@samples.txt]="het" &amp; binom(AD)&lt;0.01</pre>
</div>
</div>
</li>
<li>
<p>function on FORMAT tags (over samples) and INFO tags (over vector fields):
maximum; minimum; arithmetic mean (AVG is synonymous with MEAN); median;
standard deviation from mean; sum; string length; absolute value; number of
elements:</p>
<div class="literalblock">
<div class="content">
<pre>MAX, MIN, AVG, MEAN, MEDIAN, STDEV, SUM, STRLEN, ABS, COUNT</pre>
</div>
</div>
<div class="paragraph">
<p>Note that functions above evaluate to a single value across all samples and
are intended to select sites, not samples, even when applied on FORMAT tags.
However, when prefixed with SMPL_ (or "s" for brevity, e.g. SMPL_MAX or sMAX),
they will evaluate to a vector of per-sample values when applied on FORMAT tags:</p>
</div>
<div class="literalblock">
<div class="content">
<pre>SMPL_MAX, SMPL_MIN, SMPL_AVG, SMPL_MEAN, SMPL_MEDIAN, SMPL_STDEV, SMPL_SUM,
sMAX, sMIN, sAVG, sMEAN, sMEDIAN, sSTDEV, sSUM</pre>
</div>
</div>
</li>
<li>
<p>two-tailed binomial test. Note that for N=0 the test evaluates to a missing value
and when FORMAT/GT is used to determine the vector indices, it evaluates to 1 for
homozygous genotypes.</p>
<div class="literalblock">
<div class="content">
<pre>binom(FMT/AD)                .. GT can be used to determine the correct index
binom(AD[0],AD[1])           .. or the fields can be given explicitly
phred(binom())               .. the same as binom but phred-scaled</pre>
</div>
</div>
</li>
<li>
<p>variables calculated on the fly if not present: number of alternate alleles;
number of samples; count of alternate alleles; minor allele count (similar to
AC but is always smaller than 0.5); frequency of alternate alleles (AF=AC/AN);
frequency of minor alleles (MAF=MAC/AN); number of alleles in called genotypes;
number of samples with missing genotype; fraction of samples with missing genotype;
indel length (deletions negative, insertions positive, balanced substitutions zero)</p>
<div class="literalblock">
<div class="content">
<pre>N_ALT, N_SAMPLES, AC, MAC, AF, MAF, AN, N_MISSING, F_MISSING, ILEN</pre>
</div>
</div>
</li>
<li>
<p>the number (N_PASS) or fraction (F_PASS) of samples which pass the expression</p>
<div class="literalblock">
<div class="content">
<pre>N_PASS(GQ&gt;90 &amp; GT!="mis") &gt; 90
F_PASS(GQ&gt;90 &amp; GT!="mis") &gt; 0.9</pre>
</div>
</div>
</li>
<li>
<p>custom perl filtering. Note that this command is not compiled in by default, see
the section <strong>Optional Compilation with Perl</strong> in the INSTALL file for help
and misc/demo-flt.pl for a working example. The demo defined the perl subroutine
"severity" which can be invoked from the command line as follows:</p>
<div class="literalblock">
<div class="content">
<pre>perl:path/to/script.pl; perl.severity(INFO/CSQ) &gt; 3</pre>
</div>
</div>
</li>
</ul>
</div>
<div class="ulist">
<div class="title">Notes:</div>
<ul>
<li>
<p>String comparisons and regular expressions are case-insensitive</p>
</li>
<li>
<p>Comma in strings is interpreted as a separator and when multiple values are compared, the OR logic is used.
Consequently, the following two expressions are equivalent but not the third:</p>
<div class="literalblock">
<div class="content">
<pre>-i 'TAG="hello,world"'
-i 'TAG="hello" || TAG="world"'
-i 'TAG="hello" &amp;&amp; TAG="world"'</pre>
</div>
</div>
</li>
<li>
<p>Variables and function names are case-insensitive, but not tag names. For example,
"qual" can be used instead of "QUAL", "strlen()" instead of "STRLEN()" , but
not "dp" instead of "DP".</p>
</li>
<li>
<p>When querying multiple values, all elements are tested and the OR logic is
used on the result. For example, when querying "TAG=1,2,3,4", it will be evaluated as follows:</p>
<div class="literalblock">
<div class="content">
<pre>-i 'TAG[*]=1'   .. true, the record will be printed
-i 'TAG[*]!=1'  .. true
-e 'TAG[*]=1'   .. false, the record will be discarded
-e 'TAG[*]!=1'  .. false
-i 'TAG[0]=1'   .. true
-i 'TAG[0]!=1'  .. false
-e 'TAG[0]=1'   .. false
-e 'TAG[0]!=1'  .. true</pre>
</div>
</div>
</li>
<li>
<p>When arithmetic operators are used on vectors A and B, the following logic is used to
compute the resulting vector C:</p>
<div class="ulist">
<ul>
<li>
<p>C_i = A_i + B_i when length(A)==B(A) and sets length&#169;=length(A)</p>
</li>
<li>
<p>C_i = A_i + B_0 when length(B)=1 and sets length&#169;=length(A)</p>
</li>
<li>
<p>C_i = A_0 + B_i when length(A)=1 and sets length&#169;=length(B)</p>
</li>
<li>
<p>throw an error when length(A)!=length(B) AND length(A)!=1 AND length(B)!=1</p>
</li>
</ul>
</div>
</li>
</ul>
</div>
<div class="paragraph">
<p><strong>Examples:</strong></p>
</div>
<div class="openblock">
<div class="content">
<div class="literalblock">
<div class="content">
<pre>MIN(DV)&gt;5       .. selects the whole site, evaluates min across all values and samples</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>SMPL_MIN(DV)&gt;5  .. selects matching samples, evaluates within samples</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>MIN(DV/DP)&gt;0.3</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>MIN(DP)&gt;10 &amp; MIN(DV)&gt;3</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>FMT/DP&gt;10  &amp; FMT/GQ&gt;10 .. both conditions must be satisfied within one sample</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>FMT/DP&gt;10 &amp;&amp; FMT/GQ&gt;10 .. the conditions can be satisfied in different samples</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>QUAL&gt;10 |  FMT/GQ&gt;10   .. true for sites with QUAL&gt;10 or a sample with GQ&gt;10, but selects only samples with GQ&gt;10</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>QUAL&gt;10 || FMT/GQ&gt;10   .. true for sites with QUAL&gt;10 or a sample with GQ&gt;10, plus selects all samples at such sites</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>TYPE="snp" &amp;&amp; QUAL&gt;=10 &amp;&amp; (DP4[2]+DP4[3] &gt; 2)</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>COUNT(GT="hom")=0      .. no homozygous genotypes at the site</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>AVG(GQ)&gt;50             .. average (arithmetic mean) of genotype qualities bigger than 50</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>ID=@file               .. selects lines with ID present in the file</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>ID!=@~/file            .. skip lines with ID present in the ~/file</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>INFO/TAG=@file         .. selects lines with INFO/TAG value present in the file</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>MAF[0]&lt;0.05    .. select rare variants at 5% cutoff</pre>
</div>
</div>
<div class="literalblock">
<div class="content">
<pre>POS&gt;=100   .. restrict your range query, e.g. 20:100-200 to strictly sites with POS in that range.</pre>
</div>
</div>
</div>
</div>
<div class="paragraph">
<p><strong>Shell expansion:</strong></p>
</div>
<div class="paragraph">
<p>Note that expressions must often be quoted because some characters
have special meaning in the shell.
An example of expression enclosed in single quotes which cause
that the whole expression is passed to the program as intended:</p>
</div>
<div class="openblock">
<div class="content">
<div class="literalblock">
<div class="content">
<pre>bcftools view -i '%ID!="." &amp; MAF[0]&lt;0.01'</pre>
</div>
</div>
</div>
</div>
<div class="paragraph">
<p>Please refer to the documentation of your shell for details.</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="terminology">TERMINOLOGY</h2>
<div class="sectionbody">
<div class="paragraph">
<p>The program and the documentation uses the following terminology, multiple terms can be used
interchangeably for the same VCF record type</p>
</div>
<div class="openblock">
<div class="content">
<div class="literalblock">
<div class="content">
<pre>REF   ALT
---------
C     .         .. reference allele / non-variant site / ref-only site
C     T         .. SNP or SNV (single-nucleotide polymorphism or variant), used interchangeably
CC    TT        .. MNP (multi-nucleotide polymorphism)
CAAA  C         .. indel, deletion (regardless of length)
C     CAAA      .. indel, insertion (regardless of length)
C     &lt;*&gt;       .. gVCF block, the allele &lt;*&gt; is a placeholder for alternate allele possibly missed because of low coverage
C     &lt;NON_REF&gt; .. synonymous to &lt;*&gt;
C     *         .. overlapping deletion
C     &lt;INS&gt;     .. symbolic allele, known also as 'other [than above]'</pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_performance">PERFORMANCE</h2>
<div class="sectionbody">
<div class="paragraph">
<p>HTSlib was designed with BCF format in mind. When parsing VCF files, all records
are internally converted into BCF representation. Simple operations, like removing
a single column from a VCF file, can be therefore done much faster with standard
UNIX commands, such as <strong>awk</strong> or <strong>cut</strong>.
Therefore it is recommended to use BCF as input/output format whenever possible to avoid
large overhead of the VCF &#8594; BCF &#8594; VCF conversion.</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_bugs">BUGS</h2>
<div class="sectionbody">
<div class="paragraph">
<p>Please report any bugs you encounter on the github website: <a href="http://github.com/samtools/bcftools" class="bare">http://github.com/samtools/bcftools</a></p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_authors">AUTHORS</h2>
<div class="sectionbody">
<div class="paragraph">
<p>Heng Li from the Sanger Institute wrote the original C version of htslib,
samtools and bcftools. Bob Handsaker from the Broad Institute implemented the
BGZF library. Petr Danecek is maintaining and further developing bcftools, together
with the rest of the <a href="https://www.sanger.ac.uk/tool/samtools-bcftools-htslib">samtools team</a>.
Many other people contributed to the program
and to the file format specifications, both directly and indirectly by
providing patches, testing and reporting bugs. We thank them all.</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_resources">RESOURCES</h2>
<div class="sectionbody">
<div class="paragraph">
<p>BCFtools GitHub website: <a href="http://github.com/samtools/bcftools" class="bare">http://github.com/samtools/bcftools</a></p>
</div>
<div class="paragraph">
<p>Samtools GitHub website: <a href="http://github.com/samtools/samtools" class="bare">http://github.com/samtools/samtools</a></p>
</div>
<div class="paragraph">
<p>HTSlib GitHub website: <a href="http://github.com/samtools/htslib" class="bare">http://github.com/samtools/htslib</a></p>
</div>
<div class="paragraph">
<p>File format specifications: <a href="http://samtools.github.io/hts-specs" class="bare">http://samtools.github.io/hts-specs</a></p>
</div>
<div class="paragraph">
<p>BCFtools documentation: <a href="http://samtools.github.io/bcftools" class="bare">http://samtools.github.io/bcftools</a></p>
</div>
<div class="paragraph">
<p>BCFtools wiki page: <a href="https://github.com/samtools/bcftools/wiki" class="bare">https://github.com/samtools/bcftools/wiki</a></p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_copying">COPYING</h2>
<div class="sectionbody">
<div class="paragraph">
<p>The MIT/Expat License or GPL License, see the LICENSE document for details.
Copyright (c) Genome Research Ltd.</p>
</div>
</div>
</div>
</div>
<div id="footer">
<div id="footer-text">
Last updated 2024-09-12 16:06:32 +0100
</div>
</div>
</body>
</html>